{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Identification of DNA-binding motifs case study "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "This section will present a comparative analysis to demonstrate the application and performance of proPythia for addressing sequence-based prediction problems.\n",
        "\n",
        "We'll try to replicate the [Deep Learning in Genomics Primer Notebook](https://github.com/abidlabs/deep-learning-genomics-primer/blob/master/A_Primer_on_Deep_Learning_in_Genomics_Public.ipynb) which is a supplement to the manuscript [A Primer on Deep Learning in Genomics](https://www.nature.com/articles/s41588-018-0295-5)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "pycharm": {
          "name": "#%%\n"
        }
      },
      "outputs": [],
      "source": [
        "%load_ext autoreload\n",
        "\n",
        "import sys\n",
        "import pandas as pd\n",
        "\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.metrics import make_scorer, matthews_corrcoef\n",
        "from sklearn.preprocessing import StandardScaler\n",
        "\n",
        "sys.path.append('../../../../src/')\n",
        "from propythia.shallow_ml import ShallowML\n",
        "\n",
        "from descriptors import DNADescriptor"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "(2000, 2)\n",
            "                                            sequence  label\n",
            "0  CCGAGGGCTATGGTTTGGAAGTTAGAACCCTGGGGCTTCTCGCGGA...      0\n",
            "1  GAGTTTATATGGCGCGAGCCTAGTGGTTTTTGTACTTGTTTGTCGC...      0\n",
            "2  GATCAGTAGGGAAACAAACAGAGGGCCCAGCCACATCTAGCAGGTA...      0\n",
            "3  GTCCACGACCGAACTCCCACCTTGACCGCAGAGGTACCACCAGAGC...      1\n",
            "4  GGCGACCGAACTCCAACTAGAACCTGCATAACTGGCCTGGGAGATA...      1\n"
          ]
        }
      ],
      "source": [
        "dataset = pd.read_csv(\"dataset.csv\")\n",
        "print(dataset.shape)\n",
        "print(dataset.head())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "<AxesSubplot:xlabel='label'>"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        },
        {
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEJCAYAAACZjSCSAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAQ40lEQVR4nO3df2iV9d/H8dc510hLPWzntDOPP0gcZseCBAshgkDJoZ1p3wqOnMJ/rJAIpB/kjNttZlRHixJ0aSCVgRomJTsRszCESqWi30fSbDPR446ezZyju/Dsuv+Q77glz6ae47k238/HX3be59r5nPzg85zr/JjPdV1XAACz/F4vAADgLUIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjKrxewJXq7u5VXx8fgShWKDRaudxZr5cBXBT7s3T8fp+qqkZddDZsQ9DX5xKCEuH/I4Yy9ufVx6khADCOEACAcYQAAIwjBABgHCEAAOMIAQAYN2gIksmkZs2apalTp+rgwYP9l7e3tysej6uurk7xeFwdHR1FzwAA5ecb7BfTfPPNNxo/frwefvhhbdiwQTfffLMkadGiRXrwwQe1YMEC7dy5Uzt27NDmzZuLml2OXO7skH9/8ZjA9Ro5Yth+VGNI+d+/z6nnzF9eLwNlVl09RidP9ni9jGuC3+9TKDT6orNBQ/Bfs2bN6g9BLpdTXV2d9u/fL8dxlM/nNXPmTO3atUuu617RLBgMXtadGg4hqK4eo/pndnq9jGtC62sL+AfBIEJQOgOF4IoermYyGdXU1MhxHEmS4zgKh8PKZDJyXfeKZpcbAgBAaQzb8xaFyoZrV3X1GK+XAA/w9371XVEIIpGIOjs7lc/n+0/xZLNZRSIRua57RbPLNVxODaF0OEVgD6eGSmegU0NX9PbRUCikaDSqVColSUqlUopGowoGg1c8AwB4Y9AXi1988UXt2rVLp06dUlVVlSorK/Xxxx/r8OHDamho0JkzZxQIBJRMJjV58mRJuuLZ5Rguzwh4sbg0eLG4tHhHW2kNh3e1leRdQ0MNIbCFEJQWe7O0hsP+LPmpIQDAtYMQAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYVHYLPP/9c999/vxYsWKD6+nrt2rVLktTe3q54PK66ujrF43F1dHT0HzPQDABQXkWFwHVdPffcc1q9erV27typNWvWaNmyZerr61NTU5MSiYTa2tqUSCTU2NjYf9xAMwBAeRX9jMDv96unp0eS1NPTo3A4rO7ubqXTacViMUlSLBZTOp1WV1eXcrlcwRkAoPwqijnY5/PpjTfe0BNPPKEbbrhBvb292rhxozKZjGpqauQ4jiTJcRyFw2FlMhm5rltwFgwGi79HAIDLUlQIzp07p40bN6qlpUUzZszQt99+q6eeekqrV68u1foKCoVGX/XbwNBSXT3G6yUABQ3n/VlUCA4cOKBsNqsZM2ZIkmbMmKHrr79eI0aMUGdnp/L5vBzHUT6fVzabVSQSkeu6BWeXI5c7q74+t5jlX3XDeWMMRSdP9ni9hGsGe7P0hvr+9Pt9BR9AF/UawdixY3XixAn9/vvvkqTDhw/r1KlTuummmxSNRpVKpSRJqVRK0WhUwWBQoVCo4AwAUH5FPSOorq5Wc3Ozli5dKp/PJ0l6+eWXVVlZqebmZjU0NKilpUWBQEDJZLL/uIFmAIDyKioEkjR//nzNnz//X5fX1tZq+/btFz1moBkAoLz4ZDEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOOKDsHff/+tpqYmzZkzR/X19VqxYoUkqb29XfF4XHV1dYrH4+ro6Og/ZqAZAKC8ig7BmjVrNGLECLW1tam1tVVLly6VJDU1NSmRSKitrU2JREKNjY39xww0AwCUV1Eh6O3t1UcffaSlS5fK5/NJkm688Ublcjml02nFYjFJUiwWUzqdVldX14AzAED5VRRz8NGjR1VZWal169Zp//79GjVqlJYuXaqRI0eqpqZGjuNIkhzHUTgcViaTkeu6BWfBYLD4ewQAuCxFheDcuXM6evSopk2bpmXLlumHH37QkiVLtHbt2lKtr6BQaPRVvw0MLdXVY7xeAlDQcN6fRYVg3Lhxqqio6D/Nc/vtt6uqqkojR45UZ2en8vm8HMdRPp9XNptVJBKR67oFZ5cjlzurvj63mOVfdcN5YwxFJ0/2eL2EawZ7s/SG+v70+30FH0AX9RpBMBjUzJkz9eWXX0o6/26gXC6nSZMmKRqNKpVKSZJSqZSi0aiCwaBCoVDBGQCg/Ip6RiBJK1eu1PPPP69kMqmKigqtXr1agUBAzc3NamhoUEtLiwKBgJLJZP8xA80AAOVVdAgmTpyo995771+X19bWavv27Rc9ZqAZAKC8+GQxABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjShaCdevWaerUqTp48KAkqb29XfF4XHV1dYrH4+ro6Oi/7kAzAEB5lSQEv/zyi77//nuNGzeu/7KmpiYlEgm1tbUpkUiosbHxkmYAgPIqOgT//POPXnjhBTU1Ncnn80mScrmc0um0YrGYJCkWiymdTqurq2vAGQCg/CqK/QFr167V/PnzNXHixP7LMpmMampq5DiOJMlxHIXDYWUyGbmuW3AWDAaLXQ4A4DIVFYLvvvtOP/30k5599tlSreeShUKjy36b8FZ19RivlwAUNJz3Z1Eh+Prrr/X7779r9uzZkqQTJ05o8eLFWr58uTo7O5XP5+U4jvL5vLLZrCKRiFzXLTi7HLncWfX1ucUs/6obzhtjKDp5ssfrJVwz2JulN9T3p9/vK/gAuqjXCB5//HF98cUX2r17t3bv3q2xY8dq06ZNmjdvnqLRqFKplCQplUopGo0qGAwqFAoVnAEAyq/o1wgKaW5uVkNDg1paWhQIBJRMJi9pBgAor5KGYPfu3f1/rq2t1fbt2y96vYFmAIDy4pPFAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMKyoE3d3deuyxx1RXV6f6+no9+eST6urqkiS1t7crHo+rrq5O8XhcHR0d/ccNNAMAlFdRIfD5fHr00UfV1tam1tZWTZw4Ua+++qokqampSYlEQm1tbUokEmpsbOw/bqAZAKC8igpBZWWlZs6c2f/f06dP1/Hjx5XL5ZROpxWLxSRJsVhM6XRaXV1dA84AAOVXUaof1NfXp61bt2rWrFnKZDKqqamR4ziSJMdxFA6Hlclk5LpuwVkwGLzk2wuFRpdq6RgmqqvHeL0EoKDhvD9LFoJVq1bphhtu0COPPKJ0Ol2qH1tQLndWfX3uVb+dYgznjTEUnTzZ4/USrhnszdIb6vvT7/cVfABdkhAkk0kdOXJEGzZskN/vVyQSUWdnp/L5vBzHUT6fVzabVSQSkeu6BWcAgPIr+u2jr7/+un7++WetX79e1113nSQpFAopGo0qlUpJklKplKLRqILB4IAzAED5FfWM4NChQ9qwYYMmTZqkhQsXSpImTJig9evXq7m5WQ0NDWppaVEgEFAymew/bqAZAKC8igrBlClT9Ouvv150Vltbq+3bt1/2DABQXnyyGACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcZ6FoL29XfF4XHV1dYrH4+ro6PBqKQBgmmchaGpqUiKRUFtbmxKJhBobG71aCgCYVuHFjeZyOaXTab399tuSpFgsplWrVqmrq0vBYPCSfobf77uaSyyZcNX1Xi/hmjFc/s6HC/ZmaQ31/TnQ+jwJQSaTUU1NjRzHkSQ5jqNwOKxMJnPJIaiqGnU1l1gym/5njtdLuGaEQqO9XsI1hb1ZWsN5f/JiMQAY50kIIpGIOjs7lc/nJUn5fF7ZbFaRSMSL5QCAaZ6EIBQKKRqNKpVKSZJSqZSi0eglnxYCAJSOz3Vd14sbPnz4sBoaGnTmzBkFAgElk0lNnjzZi6UAgGmehQAAMDTwYjEAGEcIAMA4QgAAxhECADCOEACAcZ58xQS81d3drRMnTkiSxo4dq6qqKo9XBMBLhMCQP/74QytWrFA6nVY4HJYkZbNZTZs2TStXrtSkSZO8XSAAT/A5AkMWLlyoRCKhWCwmv//8WcG+vj61trZqy5Ytev/99z1eIXBx9fX1am1t9XoZ1yyeERhy+vRpzZ8//4LL/H6/FixYoDfffNOjVQHn/fbbbwVn3d3dZVyJPYTAkMrKSqVSKd13333y+c5/N7nrumptbVUgEPB4dbAuFotp/PjxuthJitOnT5d/QYZwasiQjo4ONTU16cCBA6qpqZEkdXZ26pZbblFzczPf9QRPzZ49W1u2bOnfm//fPffcoz179niwKht4RmDIpEmT9O6776qrq0uZTEbS+a8E51tfMRTMmTNHx44du2gI7r33Xg9WZAfPCADAOD5QBgDGEQIAMI4QAAXMmjVLX3311aDXmzp1qo4cOXJFt1HMsUCpEAIAMI4QAIBxhAAYxI8//qh4PK477rhDd999t1544QX9888/F1xnz549mj17tmbOnKlkMqm+vr7+2QcffKC5c+fqzjvv1OLFi3Xs2LFy3wVgQIQAGITf79fy5cu1b98+bdu2TXv37tWWLVsuuM6nn36qHTt26MMPP9Tu3bu1Y8cOSdJnn32mjRs3at26ddq7d69mzJihZ555xou7ARRECIBB3HbbbZo+fboqKio0YcIExeNxff311xdc57HHHlNlZaXGjRunRYsWKZVKSZK2bdumxx9/XLW1taqoqNCSJUt04MABnhVgSOGTxcAg2tvb9corr+jnn3/WX3/9pXw+r1tvvfWC60Qikf4/jx8/XtlsVpJ0/PhxvfTSS0omk/1z13XV2dmp8ePHl+cOAIMgBMAgmpubNW3aNL322msaPXq03nnnHbW1tV1wnUwmoylTpkg6/4//f3/fQyQS0ZIlS/71ra/AUMKpIWAQvb29GjVqlEaNGqXDhw9r69at/7rOpk2b9OeffyqTyWjz5s2aN2+epPO/A+Ktt97SoUOHJEk9PT365JNPyrp+YDA8IwAGsWzZMq1YsUKbNm1SNBrVvHnztG/fvguuM3v2bD3wwAM6e/as/vOf/+ihhx6SdP7L0np7e/X000/r2LFjGjNmjO666y7NnTvXi7sCXBRfOgcAxnFqCACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIz7P14CyJG+yc4+AAAAAElFTkSuQmCC",
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "# plot the distribution of each class\n",
        "dataset.groupby('label').size().plot(kind='bar')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [],
      "source": [
        "# List of descriptors to be calculed. If empty, all descriptors will be calculated.\n",
        "specifics = []"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "To calculate features, and to be more easy, we create a function to calculate features, calculating all available DNA features.\n",
        "There are a total of 20 descriptors. They are: \n",
        "- length\n",
        "- gc_content\n",
        "- at_content\n",
        "- nucleic_acid_composition\n",
        "- enhanced_nucleic_acid_composition\n",
        "- dinucleotide_composition\n",
        "- trinucleotide_composition\n",
        "- k_spaced_nucleic_acid_pairs\n",
        "- kmer\n",
        "- nucleotide_chemical_property\n",
        "- accumulated_nucleotide_frequency\n",
        "- DAC\n",
        "- DCC\n",
        "- DACC\n",
        "- TAC\n",
        "- TCC\n",
        "- TACC\n",
        "- PseDNC\n",
        "- PseKNC\n",
        "- binary"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "scrolled": false
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "0 / 2000\n",
            "100 / 2000\n",
            "200 / 2000\n",
            "300 / 2000\n",
            "400 / 2000\n",
            "500 / 2000\n",
            "600 / 2000\n",
            "700 / 2000\n",
            "800 / 2000\n",
            "900 / 2000\n",
            "1000 / 2000\n",
            "1100 / 2000\n",
            "1200 / 2000\n",
            "1300 / 2000\n",
            "1400 / 2000\n",
            "1500 / 2000\n",
            "1600 / 2000\n",
            "1700 / 2000\n",
            "1800 / 2000\n",
            "1900 / 2000\n",
            "Done!\n"
          ]
        }
      ],
      "source": [
        "def calculate_feature(data):\n",
        "    list_feature = []\n",
        "    count = 0\n",
        "    for seq in data['sequence']:\n",
        "        res = {'sequence': seq}\n",
        "        dna = DNADescriptor(seq)\n",
        "        feature = dna.get_descriptors(specifics=specifics)\n",
        "        res.update(feature)\n",
        "        list_feature.append(res)\n",
        "        \n",
        "        # print progress every 100 sequences\n",
        "        if count % 100 == 0:\n",
        "            print(count, '/', len(data))\n",
        "\n",
        "        count += 1\n",
        "    print(\"Done!\")\n",
        "    df = pd.DataFrame(list_feature)\n",
        "    return df\n",
        "\n",
        "features = calculate_feature(dataset)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [],
      "source": [
        "fps_y = dataset['label']\n",
        "fps_x = features.loc[:, features.columns != 'label']\n",
        "fps_x = fps_x.loc[:, fps_x.columns != 'sequence']"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>length</th>\n",
              "      <th>gc_content</th>\n",
              "      <th>at_content</th>\n",
              "      <th>nucleic_acid_composition_A</th>\n",
              "      <th>nucleic_acid_composition_C</th>\n",
              "      <th>nucleic_acid_composition_G</th>\n",
              "      <th>nucleic_acid_composition_T</th>\n",
              "      <th>enhanced_nucleic_acid_composition_0_A</th>\n",
              "      <th>enhanced_nucleic_acid_composition_0_C</th>\n",
              "      <th>enhanced_nucleic_acid_composition_0_G</th>\n",
              "      <th>...</th>\n",
              "      <th>binary_47_2</th>\n",
              "      <th>binary_47_3</th>\n",
              "      <th>binary_48_0</th>\n",
              "      <th>binary_48_1</th>\n",
              "      <th>binary_48_2</th>\n",
              "      <th>binary_48_3</th>\n",
              "      <th>binary_49_0</th>\n",
              "      <th>binary_49_1</th>\n",
              "      <th>binary_49_2</th>\n",
              "      <th>binary_49_3</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>50</td>\n",
              "      <td>0.60</td>\n",
              "      <td>0.40</td>\n",
              "      <td>0.18</td>\n",
              "      <td>0.26</td>\n",
              "      <td>0.34</td>\n",
              "      <td>0.22</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.4</td>\n",
              "      <td>0.4</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>50</td>\n",
              "      <td>0.48</td>\n",
              "      <td>0.52</td>\n",
              "      <td>0.12</td>\n",
              "      <td>0.16</td>\n",
              "      <td>0.32</td>\n",
              "      <td>0.40</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.4</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>50</td>\n",
              "      <td>0.54</td>\n",
              "      <td>0.46</td>\n",
              "      <td>0.34</td>\n",
              "      <td>0.26</td>\n",
              "      <td>0.28</td>\n",
              "      <td>0.12</td>\n",
              "      <td>0.4</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.2</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>50</td>\n",
              "      <td>0.64</td>\n",
              "      <td>0.36</td>\n",
              "      <td>0.24</td>\n",
              "      <td>0.42</td>\n",
              "      <td>0.22</td>\n",
              "      <td>0.12</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.4</td>\n",
              "      <td>0.2</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>50</td>\n",
              "      <td>0.54</td>\n",
              "      <td>0.46</td>\n",
              "      <td>0.28</td>\n",
              "      <td>0.26</td>\n",
              "      <td>0.28</td>\n",
              "      <td>0.18</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.6</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1995</th>\n",
              "      <td>50</td>\n",
              "      <td>0.66</td>\n",
              "      <td>0.34</td>\n",
              "      <td>0.16</td>\n",
              "      <td>0.26</td>\n",
              "      <td>0.40</td>\n",
              "      <td>0.18</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.4</td>\n",
              "      <td>0.4</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1996</th>\n",
              "      <td>50</td>\n",
              "      <td>0.44</td>\n",
              "      <td>0.56</td>\n",
              "      <td>0.22</td>\n",
              "      <td>0.22</td>\n",
              "      <td>0.22</td>\n",
              "      <td>0.34</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.4</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1997</th>\n",
              "      <td>50</td>\n",
              "      <td>0.46</td>\n",
              "      <td>0.54</td>\n",
              "      <td>0.18</td>\n",
              "      <td>0.24</td>\n",
              "      <td>0.22</td>\n",
              "      <td>0.36</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.4</td>\n",
              "      <td>0.2</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1998</th>\n",
              "      <td>50</td>\n",
              "      <td>0.48</td>\n",
              "      <td>0.52</td>\n",
              "      <td>0.28</td>\n",
              "      <td>0.24</td>\n",
              "      <td>0.24</td>\n",
              "      <td>0.24</td>\n",
              "      <td>0.4</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.2</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1999</th>\n",
              "      <td>50</td>\n",
              "      <td>0.60</td>\n",
              "      <td>0.40</td>\n",
              "      <td>0.22</td>\n",
              "      <td>0.32</td>\n",
              "      <td>0.28</td>\n",
              "      <td>0.18</td>\n",
              "      <td>0.4</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.2</td>\n",
              "      <td>...</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>2000 rows Ã— 819 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      length  gc_content  at_content  nucleic_acid_composition_A  \\\n",
              "0         50        0.60        0.40                        0.18   \n",
              "1         50        0.48        0.52                        0.12   \n",
              "2         50        0.54        0.46                        0.34   \n",
              "3         50        0.64        0.36                        0.24   \n",
              "4         50        0.54        0.46                        0.28   \n",
              "...      ...         ...         ...                         ...   \n",
              "1995      50        0.66        0.34                        0.16   \n",
              "1996      50        0.44        0.56                        0.22   \n",
              "1997      50        0.46        0.54                        0.18   \n",
              "1998      50        0.48        0.52                        0.28   \n",
              "1999      50        0.60        0.40                        0.22   \n",
              "\n",
              "      nucleic_acid_composition_C  nucleic_acid_composition_G  \\\n",
              "0                           0.26                        0.34   \n",
              "1                           0.16                        0.32   \n",
              "2                           0.26                        0.28   \n",
              "3                           0.42                        0.22   \n",
              "4                           0.26                        0.28   \n",
              "...                          ...                         ...   \n",
              "1995                        0.26                        0.40   \n",
              "1996                        0.22                        0.22   \n",
              "1997                        0.24                        0.22   \n",
              "1998                        0.24                        0.24   \n",
              "1999                        0.32                        0.28   \n",
              "\n",
              "      nucleic_acid_composition_T  enhanced_nucleic_acid_composition_0_A  \\\n",
              "0                           0.22                                    0.2   \n",
              "1                           0.40                                    0.2   \n",
              "2                           0.12                                    0.4   \n",
              "3                           0.12                                    0.2   \n",
              "4                           0.18                                    0.2   \n",
              "...                          ...                                    ...   \n",
              "1995                        0.18                                    0.0   \n",
              "1996                        0.34                                    0.0   \n",
              "1997                        0.36                                    0.2   \n",
              "1998                        0.24                                    0.4   \n",
              "1999                        0.18                                    0.4   \n",
              "\n",
              "      enhanced_nucleic_acid_composition_0_C  \\\n",
              "0                                       0.4   \n",
              "1                                       0.0   \n",
              "2                                       0.2   \n",
              "3                                       0.4   \n",
              "4                                       0.2   \n",
              "...                                     ...   \n",
              "1995                                    0.4   \n",
              "1996                                    0.2   \n",
              "1997                                    0.4   \n",
              "1998                                    0.2   \n",
              "1999                                    0.0   \n",
              "\n",
              "      enhanced_nucleic_acid_composition_0_G  ...  binary_47_2  binary_47_3  \\\n",
              "0                                       0.4  ...            0            0   \n",
              "1                                       0.4  ...            0            1   \n",
              "2                                       0.2  ...            0            0   \n",
              "3                                       0.2  ...            0            0   \n",
              "4                                       0.6  ...            1            0   \n",
              "...                                     ...  ...          ...          ...   \n",
              "1995                                    0.4  ...            1            0   \n",
              "1996                                    0.4  ...            0            1   \n",
              "1997                                    0.2  ...            0            0   \n",
              "1998                                    0.2  ...            1            0   \n",
              "1999                                    0.2  ...            1            0   \n",
              "\n",
              "      binary_48_0  binary_48_1  binary_48_2  binary_48_3  binary_49_0  \\\n",
              "0               0            1            0            0            0   \n",
              "1               0            1            0            0            0   \n",
              "2               0            1            0            0            0   \n",
              "3               0            0            0            1            0   \n",
              "4               0            0            1            0            0   \n",
              "...           ...          ...          ...          ...          ...   \n",
              "1995            1            0            0            0            0   \n",
              "1996            0            1            0            0            0   \n",
              "1997            0            1            0            0            0   \n",
              "1998            1            0            0            0            1   \n",
              "1999            1            0            0            0            0   \n",
              "\n",
              "      binary_49_1  binary_49_2  binary_49_3  \n",
              "0               1            0            0  \n",
              "1               0            1            0  \n",
              "2               0            0            1  \n",
              "3               0            1            0  \n",
              "4               0            0            1  \n",
              "...           ...          ...          ...  \n",
              "1995            0            1            0  \n",
              "1996            0            0            1  \n",
              "1997            0            0            1  \n",
              "1998            0            0            0  \n",
              "1999            1            0            0  \n",
              "\n",
              "[2000 rows x 819 columns]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "no_need_normalization = [\"length\", \"at_content\", \"gc_content\"]\n",
        "\n",
        "need_dict_normalization = [\"nucleic_acid_composition\", \"enhanced_nucleic_acid_composition\",\"dinucleotide_composition\",\"trinucleotide_composition\",\"k_spaced_nucleic_acid_pairs\",\"kmer\",\"PseDNC\", \"PseKNC\"]\n",
        "\n",
        "need_list_normalization = [\"nucleotide_chemical_property\", \"accumulated_nucleotide_frequency\", \"DAC\", \"DCC\", \"DACC\", \"TAC\",\"TCC\",\"TACC\", \"binary\"]\n",
        "\n",
        "def normalize_dict(d, field):\n",
        "    df = pd.json_normalize(d)\n",
        "    df.columns = [str(field) + \"_\" + str(i) for i in df.columns]\n",
        "    \n",
        "    for f in df.columns:\n",
        "        if isinstance(df[f][0], dict):\n",
        "            df = pd.concat([df, normalize_dict(df[f], f)], axis=1)\n",
        "            df.drop(f, axis=1, inplace=True)\n",
        "    return df\n",
        "\n",
        "def normalize_list(l, field):\n",
        "    df = pd.DataFrame(l.to_list())\n",
        "    df.columns = [str(field) + \"_\" + str(i) for i in df.columns]\n",
        "    \n",
        "    for f in df.columns:\n",
        "        if isinstance(df[f][0], list):\n",
        "            df = pd.concat([df, normalize_list(df[f], f)], axis=1)\n",
        "            df.drop(f, axis=1, inplace=True)\n",
        "    return df\n",
        "\n",
        "new_fps_x = pd.DataFrame()\n",
        "\n",
        "for col in fps_x.columns:\n",
        "    if col in need_dict_normalization:\n",
        "        new_fps_x = pd.concat([new_fps_x, normalize_dict(fps_x[col], col)], axis=1)\n",
        "    elif col in need_list_normalization:\n",
        "        new_fps_x = pd.concat([new_fps_x, normalize_list(fps_x[col], col)], axis=1)\n",
        "    else:\n",
        "        new_fps_x[col] = fps_x[col].to_numpy()\n",
        "        \n",
        "new_fps_x"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "performing gridSearch...\n",
            "GridSearchCV took 10.48 seconds for 4 candidate parameter settings.\n",
            "GridSearchCV(cv=10,\n",
            "             estimator=Pipeline(steps=[('scl', None),\n",
            "                                       ('clf',\n",
            "                                        RandomForestClassifier(random_state=1))]),\n",
            "             n_jobs=10,\n",
            "             param_grid=[{'clf__max_features': ['sqrt'],\n",
            "                          'clf__n_estimators': [100, 250, 500, 750]}],\n",
            "             scoring=make_scorer(matthews_corrcoef))\n",
            "Model with rank: 1\n",
            " Mean validation score: 0.988 (std: 0.011)\n",
            " Parameters: {'clf__max_features': 'sqrt', 'clf__n_estimators': 500}\n",
            " \n",
            "\n",
            "Model with rank: 2\n",
            " Mean validation score: 0.988 (std: 0.013)\n",
            " Parameters: {'clf__max_features': 'sqrt', 'clf__n_estimators': 750}\n",
            " \n",
            "\n",
            "Model with rank: 3\n",
            " Mean validation score: 0.987 (std: 0.013)\n",
            " Parameters: {'clf__max_features': 'sqrt', 'clf__n_estimators': 250}\n",
            " \n",
            "\n",
            "make_scorer(matthews_corrcoef)\n",
            "10\n",
            "Best score (scorer: make_scorer(matthews_corrcoef)) and parameters from a 10-fold cross validation:\n",
            " MCC score:\t0.988\n",
            " Parameters:\t{'clf__max_features': 'sqrt', 'clf__n_estimators': 500}\n",
            "\n",
            "0.986700 (0.015748) with: {'clf__max_features': 'sqrt', 'clf__n_estimators': 100}\n",
            "0.986736 (0.013282) with: {'clf__max_features': 'sqrt', 'clf__n_estimators': 250}\n",
            "0.988096 (0.010980) with: {'clf__max_features': 'sqrt', 'clf__n_estimators': 500}\n",
            "0.988079 (0.012515) with: {'clf__max_features': 'sqrt', 'clf__n_estimators': 750}\n",
            "  clf__max_features  clf__n_estimators     means      stds\n",
            "0              sqrt                100  0.986700  0.015748\n",
            "1              sqrt                250  0.986736  0.013282\n",
            "2              sqrt                500  0.988096  0.010980\n",
            "3              sqrt                750  0.988079  0.012515\n"
          ]
        }
      ],
      "source": [
        "X_train, X_test, y_train, y_test = train_test_split(new_fps_x, fps_y, stratify=fps_y)\n",
        "\n",
        "scaler = StandardScaler().fit(X_train)\n",
        "X_train = scaler.transform(X_train)\n",
        "X_test = scaler.transform(X_test)\n",
        "\n",
        "ml = ShallowML(X_train, X_test, y_train, y_test, report_name=None, columns_names=new_fps_x.columns)\n",
        "\n",
        "param_grid = [{'clf__n_estimators': [100, 250, 500, 750], 'clf__max_features': ['sqrt']}]\n",
        "\n",
        "best_rf_model_enhancers = ml.train_best_model(\n",
        "    model_name=None,\n",
        "    model='rf',\n",
        "    score=make_scorer(matthews_corrcoef),\n",
        "    param_grid=param_grid,\n",
        "    cv=10\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0       0.99      0.99      0.99       253\n",
            "           1       0.99      0.99      0.99       247\n",
            "\n",
            "    accuracy                           0.99       500\n",
            "   macro avg       0.99      0.99      0.99       500\n",
            "weighted avg       0.99      0.99      0.99       500\n",
            "\n",
            "[[250   3]\n",
            " [  2 245]]\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "{'Accuracy': 0.99,\n",
              " 'MCC': 0.9800059213999118,\n",
              " 'log_loss': 0.08747214823136405,\n",
              " 'f1 score': 0.98989898989899,\n",
              " 'roc_auc': 0.9900225632491079,\n",
              " 'Precision': array([0.494     , 0.98790323, 1.        ]),\n",
              " 'Recall': array([1.        , 0.99190283, 0.        ]),\n",
              " 'fdr': 0.012096774193548387,\n",
              " 'sn': 0.9919028340080972,\n",
              " 'sp': 0.9881422924901185}"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "scores, report, cm, cm2 = ml.score_testset(best_rf_model_enhancers)\n",
        "print(report)\n",
        "print(cm)  \n",
        "scores"
      ]
    }
  ],
  "metadata": {
    "interpreter": {
      "hash": "eccc737e2ce2db9aa50ea00a61eebd654f595be3688de5a9cd756821a0256d89"
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.13"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 1
}
