

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Adjuv functions: Features &mdash; ProPythia 0.04 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Adjuv functions: Scoring" href="propythia.adjuv_functions.scoring.html" />
    <link rel="prev" title="Adjuv functions: Sequence" href="propythia.adjuv_functions.sequence.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> ProPythia
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">here will be a mega pretty file explaining everything</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="src.html">Source Code</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="propythia.html">ProPythia modules</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="propythia.adjuv_functions.html">Adjuv Functions</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="propythia.adjuv_functions.sequence.html">Adjuv functions: Sequence</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Adjuv functions: Features</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.aa_composition">aa_composition</a></li>
<li class="toctree-l4"><a class="reference internal" href="#aa-index">aa_index</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.autocorrelation">autocorrelation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.binary">binary</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.binary_aa_properties">binary_aa_properties</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.bondcomp">bondcomp</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.conjoint_triad">conjoint_triad</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.ctd">ctd</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.descriptors_modlamp">descriptors_modlamp</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.modlamp_core">modlamp_core</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.pcp">pcp</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.pseudo_aac">pseudo_aac</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions.quasi_sequence_order">quasi_sequence_order</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-propythia.adjuv_functions.features_functions">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="propythia.adjuv_functions.scoring.html">Adjuv functions: Scoring</a></li>
<li class="toctree-l3"><a class="reference internal" href="propythia.adjuv_functions.html#module-propythia.adjuv_functions">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="src.html#module-propythia">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="guides.html">Guides</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ProPythia</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="src.html">Source Code</a> &raquo;</li>
        
          <li><a href="propythia.adjuv_functions.html">Adjuv Functions</a> &raquo;</li>
        
      <li>Adjuv functions: Features</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/propythia.adjuv_functions.features_functions.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="adjuv-functions-features">
<h1>Adjuv functions: Features<a class="headerlink" href="#adjuv-functions-features" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-propythia.adjuv_functions.features_functions.aa_composition">
<span id="aa-composition"></span><h2>aa_composition<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.aa_composition" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>The module is used for computing the composition of amino acids, dipetide and</p>
<p>3-mers (tri-peptide) for a given protein sequence. You can get 8420 descriptors</p>
<p>for a given protein sequence. You can freely use and distribute it. If you hava</p>
<p>any problem, you could contact with us timely!</p>
<p>References:</p>
<p>[1]: Reczko, M. and Bohr, H. (1994) The DEF data base of sequence based protein</p>
<p>fold class predictions. Nucleic Acids Res, 22, 3616-3619.</p>
<p>[2]: Hua, S. and Sun, Z. (2001) Support vector machine approach for protein</p>
<p>subcellular localization prediction. Bioinformatics, 17, 721-728.</p>
<p>[3]:Grassmann, J., Reczko, M., Suhai, S. and Edler, L. (1999) Protein fold class</p>
<p>prediction: new methods of statistical classification. Proc Int Conf Intell Syst Mol</p>
<p>Biol, 106-112.</p>
<p>Authors: Dongsheng Cao and Yizeng Liang.</p>
<p>Date: 2012.3.27</p>
<p>Email: <a class="reference external" href="mailto:oriental-cds&#37;&#52;&#48;163&#46;com">oriental-cds<span>&#64;</span>163<span>&#46;</span>com</a></p>
<p>Altered and converted to python 3.6 for Ana Marta Sequeira 05/2019</p>
<hr class="docutils" />
<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_composition.calculate_aa_composition">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_composition.</code><code class="sig-name descname">calculate_aa_composition</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_composition.html#calculate_aa_composition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_composition.calculate_aa_composition" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the composition of Amino acids for a given protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing the composition of 20 amino acids.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_composition.calculate_aa_tripeptide_composition">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_composition.</code><code class="sig-name descname">calculate_aa_tripeptide_composition</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_composition.html#calculate_aa_tripeptide_composition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_composition.calculate_aa_tripeptide_composition" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the composition of AADs, dipeptide and 3-mers for a given protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing all composition values of AADs, dipeptide and 3-mers (8420).</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_composition.calculate_dipeptide_composition">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_composition.</code><code class="sig-name descname">calculate_dipeptide_composition</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_composition.html#calculate_dipeptide_composition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_composition.calculate_dipeptide_composition" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the composition of dipeptidefor a given protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing the composition of 400 dipeptides.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_composition.get_spectrum_dict">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_composition.</code><code class="sig-name descname">get_spectrum_dict</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_composition.html#get_spectrum_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_composition.get_spectrum_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the spectrum of 3-mers for a given protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing the composition values of 8000</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_composition.getkmers">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_composition.</code><code class="sig-name descname">getkmers</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_composition.html#getkmers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_composition.getkmers" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the amino acid list of 3-mers.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>result is a list form containing 8000 tri-peptides.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="aa-index">
<h2>aa_index<a class="headerlink" href="#aa-index" title="Permalink to this headline">¶</a></h2>
<p>from the aaindex database. You can freely use and distribute it. If you hava</p>
<p>any problem, you could contact with us timely!</p>
<p>Authors: Dongsheng Cao and Yizeng Liang.</p>
<p>Date: 2012.09.10</p>
<p>Email: <a class="reference external" href="mailto:oriental-cds&#37;&#52;&#48;163&#46;com">oriental-cds<span>&#64;</span>163<span>&#46;</span>com</a>
Altered and converted to python 3.6 by Ana Marta Sequeira 05/2019</p>
<dl class="class">
<dt id="propythia.adjuv_functions.features_functions.aa_index.MatrixRecord">
<em class="property">class </em><code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_index.</code><code class="sig-name descname">MatrixRecord</code><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#MatrixRecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.MatrixRecord" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#propythia.adjuv_functions.features_functions.aa_index.Record" title="propythia.adjuv_functions.features_functions.aa_index.Record"><code class="xref py py-class docutils literal notranslate"><span class="pre">propythia.adjuv_functions.features_functions.aa_index.Record</span></code></a></p>
<p>Matrix record for mutation matrices or pair-wise contact potentials</p>
<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.aa_index.MatrixRecord.extend">
<code class="sig-name descname">extend</code><span class="sig-paren">(</span><em class="sig-param">row</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#MatrixRecord.extend"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.MatrixRecord.extend" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.aa_index.MatrixRecord.get">
<code class="sig-name descname">get</code><span class="sig-paren">(</span><em class="sig-param">aai</em>, <em class="sig-param">aaj</em>, <em class="sig-param">d=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#MatrixRecord.get"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.MatrixRecord.get" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.aa_index.MatrixRecord.median">
<code class="sig-name descname">median</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#MatrixRecord.median"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.MatrixRecord.median" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="propythia.adjuv_functions.features_functions.aa_index.Record">
<em class="property">class </em><code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_index.</code><code class="sig-name descname">Record</code><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#Record"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.Record" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Amino acid index (AAindex) Record</p>
<dl class="attribute">
<dt id="propythia.adjuv_functions.features_functions.aa_index.Record.aakeys">
<code class="sig-name descname">aakeys</code><em class="property"> = 'ARNDCQEGHILKMFPSTWYV'</em><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.Record.aakeys" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.aa_index.Record.extend">
<code class="sig-name descname">extend</code><span class="sig-paren">(</span><em class="sig-param">row</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#Record.extend"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.Record.extend" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.aa_index.Record.get">
<code class="sig-name descname">get</code><span class="sig-paren">(</span><em class="sig-param">aai</em>, <em class="sig-param">aaj=None</em>, <em class="sig-param">d=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#Record.get"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.Record.get" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.aa_index.Record.median">
<code class="sig-name descname">median</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#Record.median"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.Record.median" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_index.get">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_index.</code><code class="sig-name descname">get</code><span class="sig-paren">(</span><em class="sig-param">key</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#get"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.get" title="Permalink to this definition">¶</a></dt>
<dd><p>Get record for key</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>key</strong> – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_index.get_aa_index1">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_index.</code><code class="sig-name descname">get_aa_index1</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">path='.'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#get_aa_index1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.get_aa_index1" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the amino acid property values from aaindex1</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> – name is the name of amino acid property (e.g., KRIW790103)</p></li>
<li><p><strong>path</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>get_aa_index1(name). result is a dict form containing the properties of 20 amino acids</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_index.get_aa_index23">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_index.</code><code class="sig-name descname">get_aa_index23</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">path='.'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#get_aa_index23"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.get_aa_index23" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the amino acid property values from aaindex2 and aaindex3</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> – name is the name of amino acid property (e.g.,TANS760101,GRAR740104)</p></li>
<li><p><strong>path</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>get_aa_index23(name). result is a dict form containing the properties of 400 amino acid pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_index.grep">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_index.</code><code class="sig-name descname">grep</code><span class="sig-paren">(</span><em class="sig-param">pattern</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#grep"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.grep" title="Permalink to this definition">¶</a></dt>
<dd><p>Search for pattern in title and description of all records (case insensitive) and print results on standard output.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pattern</strong> – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_index.init">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_index.</code><code class="sig-name descname">init</code><span class="sig-paren">(</span><em class="sig-param">path=None</em>, <em class="sig-param">index='123'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#init"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.init" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in the aaindex files. You need to run this (once) before you can
access any records. If the files are not within the current directory,
you need to specify the correct directory path. By default all three
aaindex files are read in.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> – </p></li>
<li><p><strong>index</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_index.init_from_file">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_index.</code><code class="sig-name descname">init_from_file</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">type=&lt;class 'propythia.adjuv_functions.features_functions.aa_index.Record'&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#init_from_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.init_from_file" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.aa_index.search">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.aa_index.</code><code class="sig-name descname">search</code><span class="sig-paren">(</span><em class="sig-param">pattern</em>, <em class="sig-param">searchtitle=True</em>, <em class="sig-param">casesensitive=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/aa_index.html#search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.aa_index.search" title="Permalink to this definition">¶</a></dt>
<dd><p>Search for pattern in description and title (optional) of all records and
return matched records as list. By default search case insensitive.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pattern</strong> – </p></li>
<li><p><strong>searchtitle</strong> – </p></li>
<li><p><strong>casesensitive</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.autocorrelation">
<span id="autocorrelation"></span><h2>autocorrelation<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.autocorrelation" title="Permalink to this headline">¶</a></h2>
<p>This module is used for computing the Autocorrelation based different</p>
<blockquote>
<div><p>properties of AADs.You can also input your properties of AADs, then it can help you</p>
</div></blockquote>
<p>to compute Autocorrelation based on the property of AADs. Currently, You</p>
<p>can get 720  for a given protein sequence based on our provided physicochemical</p>
<p>properties of AADs. You can freely use and distribute it. If you hava  any problem,</p>
<p>you could contact with us timely!</p>
<p>References:</p>
<p>[1]: <a class="reference external" href="http://www.genome.ad.jp/dbget/aaindex.html">http://www.genome.ad.jp/dbget/aaindex.html</a></p>
<p>[2]:Feng, Z.P. and Zhang, C.T. (2000) Prediction of membrane protein types based on</p>
<p>the hydrophobic index of amino acids. J Protein Chem, 19, 269-275.</p>
<p>[3]:Horne, D.S. (1988) Prediction of protein helix content from an autocorrelation</p>
<p>analysis of sequence hydrophobicities. Biopolymers, 27, 451-477.</p>
<p>[4]:Sokal, R.R. and Thomson, B.A. (2006) Population structure inferred by local</p>
<p>spatial autocorrelation: an Usage from an Amerindian tribal population. Am J</p>
<p>Phys Anthropol, 129, 121-131.</p>
<p>Authors: Dongsheng Cao and Yizeng Liang.</p>
<p>Date: 2010.11.22</p>
<p>Email: <a class="reference external" href="mailto:oriental-cds&#37;&#52;&#48;163&#46;com">oriental-cds<span>&#64;</span>163<span>&#46;</span>com</a></p>
<p>Altered to python 3.6 by Ana Marta Sequeira 05/2019</p>
<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.CalculateNormalizedMoreauBrotoAutoMutability">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">CalculateNormalizedMoreauBrotoAutoMutability</code><span class="sig-paren">(</span><em class="sig-param">ProteinSequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#CalculateNormalizedMoreauBrotoAutoMutability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.CalculateNormalizedMoreauBrotoAutoMutability" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the NormalizedMoreauBorto Autocorrelation based on Mutability.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ProteinSequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Normalized Moreau-Broto Autocorrelation based on Mutability.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_auto_total">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_auto_total</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_auto_total"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_auto_total" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for computing all autocorrelation  based on 8 properties of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30*8*3=720 normalized Moreau Broto, Moran, and Geary autocorrelation
based on the given properties(i.e., _AAPropert).</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_each_geary_auto">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_each_geary_auto</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">aap</em>, <em class="sig-param">aap_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_each_geary_auto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_each_geary_auto" title="Permalink to this definition">¶</a></dt>
<dd><p>you can use the function to compute GearyAuto for different properties based on AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p></li>
<li><p><strong>aap</strong> – AAP is a dict form containing the properties of 20 amino acids (e.g., _AvFlexibility).</p></li>
<li><p><strong>aap_name</strong> – AAPName is a string used for indicating the property (e.g., ‘_AvFlexibility’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Geary autocorrelation based on the given property.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_each_moran_auto">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_each_moran_auto</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">aap</em>, <em class="sig-param">aap_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_each_moran_auto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_each_moran_auto" title="Permalink to this definition">¶</a></dt>
<dd><p>you can use the function to compute MoranAuto for different properties based on AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p></li>
<li><p><strong>aap</strong> – AAP is a dict form containing the properties of 20 amino acids (e.g., _AvFlexibility).</p></li>
<li><p><strong>aap_name</strong> – AAPName is a string used for indicating the property (e.g., ‘_AvFlexibility’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Moran autocorrelation based on the given property.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_each_normalized_moreau_broto_auto">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_each_normalized_moreau_broto_auto</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">aap</em>, <em class="sig-param">aap_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_each_normalized_moreau_broto_auto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_each_normalized_moreau_broto_auto" title="Permalink to this definition">¶</a></dt>
<dd><p>you can use the function to compute MoreauBrotoAuto for different properties based on AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p></li>
<li><p><strong>aap</strong> – AAP is a dict form containing the properties of 20 amino acids (e.g., _AvFlexibility).</p></li>
<li><p><strong>aap_name</strong> – AAPName is a string used for indicating the property (e.g., ‘_AvFlexibility’)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Normalized Moreau-Broto autocorrelation based on the given property.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_geary_auto</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">aa_property</em>, <em class="sig-param">aa_property_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_geary_auto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for computing GearyAuto for all properties</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>aa_property</strong> – AAProperty is a list or tuple form containing the properties of 20 amino acids (e.g., _AAProperty).</p></li>
<li><p><strong>aa_property_name</strong> – AAPName is a list or tuple form used for indicating the property (e.g., ‘_AAPropertyName’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30*p Geary autocorrelation based on the given properties.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_av_flexibility">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_geary_auto_av_flexibility</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_geary_auto_av_flexibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_av_flexibility" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the GearyAuto Autocorrelation  based on AvFlexibility.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Geary Autocorrelation based on AvFlexibility.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_free_energy">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_geary_auto_free_energy</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_geary_auto_free_energy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_free_energy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the GearyAuto Autocorrelation based on FreeEnergy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Geary Autocorrelation based on FreeEnergy.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_hydrophobicity">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_geary_auto_hydrophobicity</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_geary_auto_hydrophobicity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_hydrophobicity" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the GearyAuto Autocorrelation  based on hydrophobicity.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Geary Autocorrelation based on hydrophobicity.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_mutability">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_geary_auto_mutability</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_geary_auto_mutability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_mutability" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the GearyAuto Autocorrelation  based on Mutability.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Geary Autocorrelation based on Mutability.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_polarizability">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_geary_auto_polarizability</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_geary_auto_polarizability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_polarizability" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the GearyAuto Autocorrelation  based on Polarizability.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Geary Autocorrelation based on Polarizability.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_residue_asa">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_geary_auto_residue_asa</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_geary_auto_residue_asa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_residue_asa" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the GearyAuto Autocorrelation  based on ResidueASA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Geary Autocorrelation based on ResidueASA.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_residue_vol">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_geary_auto_residue_vol</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_geary_auto_residue_vol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_residue_vol" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the GearyAuto Autocorrelation  based on ResidueVol.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Geary Autocorrelation based on ResidueVol.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_steric">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_geary_auto_steric</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_geary_auto_steric"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_steric" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the GearyAuto Autocorrelation  based on Steric.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Geary Autocorrelation based on Steric.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_total">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_geary_auto_total</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_geary_auto_total"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_geary_auto_total" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for computing Geary autocorrelation based on 8 properties of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30*8=240 Geary autocorrelation based on the given
properties(i.e., _AAPropert).</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_moran_auto</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">aa_property</em>, <em class="sig-param">aa_property_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_moran_auto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for computing MoranAuto for all properties</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p></li>
<li><p><strong>aa_property</strong> – AAProperty is a list or tuple form containing the properties of 20 amino acids (e.g., _AAProperty).</p></li>
<li><p><strong>aa_property_name</strong> – AAPName is a list or tuple form used for indicating the property (e.g., ‘_AAPropertyName’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30*p Moran autocorrelation based on the given properties.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_av_flexibility">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_moran_auto_av_flexibility</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_moran_auto_av_flexibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_av_flexibility" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the MoranAuto Autocorrelation based on        AvFlexibility.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Moran Autocorrelation based on AvFlexibility.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_free_energy">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_moran_auto_free_energy</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_moran_auto_free_energy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_free_energy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the MoranAuto Autocorrelation based on FreeEnergy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Moran Autocorrelation based on FreeEnergy.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_hydrophobicity">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_moran_auto_hydrophobicity</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_moran_auto_hydrophobicity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_hydrophobicity" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the MoranAuto Autocorrelation  based on hydrophobicity.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Moran Autocorrelation based on hydrophobicity.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_mutability">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_moran_auto_mutability</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_moran_auto_mutability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_mutability" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the MoranAuto Autocorrelation based on Mutability.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Moran Autocorrelation based on Mutability.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_polarizability">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_moran_auto_polarizability</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_moran_auto_polarizability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_polarizability" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the MoranAuto Autocorrelation based on Polarizability.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Moran Autocorrelation based on Polarizability.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_residue_asa">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_moran_auto_residue_asa</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_moran_auto_residue_asa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_residue_asa" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the MoranAuto Autocorrelation based on ResidueASA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Moran Autocorrelation based on ResidueASA.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_residue_vol">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_moran_auto_residue_vol</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_moran_auto_residue_vol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_residue_vol" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the MoranAuto Autocorrelation based on ResidueVol.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Moran Autocorrelation based on ResidueVol.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_steric">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_moran_auto_steric</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_moran_auto_steric"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_steric" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the MoranAuto Autocorrelation based on AutoSteric.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Moran Autocorrelation based on AutoSteric.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_total">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_moran_auto_total</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_moran_auto_total"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_moran_auto_total" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for computing Moran autocorrelation based on 8 properties of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30*8=240 Moran autocorrelation  based on the given properties(i.e., _AAPropert).</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_normalized_moreau_broto_auto</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">aa_property</em>, <em class="sig-param">aa_property_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_normalized_moreau_broto_auto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for computing MoreauBrotoAuto for all properties.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p></li>
<li><p><strong>aa_property</strong> – AAProperty is a list or tuple form containing the properties of 20 amino acids (e.g., _AAProperty).</p></li>
<li><p><strong>aa_property_name</strong> – AAPName is a list or tuple form used for indicating the property (e.g., ‘_AAPropertyName’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30*p Normalized Moreau-Broto autocorrelation based on the given properties.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_av_flexibility">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_normalized_moreau_broto_auto_av_flexibility</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_normalized_moreau_broto_auto_av_flexibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_av_flexibility" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the NormalizedMoreauBorto Autocorrelation based on AvFlexibility.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Normalized Moreau-Broto Autocorrelation based on AvFlexibility.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_free_energy">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_normalized_moreau_broto_auto_free_energy</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_normalized_moreau_broto_auto_free_energy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_free_energy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the NormalizedMoreauBorto Autocorrelation  based on FreeEnergy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Normalized Moreau-Broto Autocorrelation based on FreeEnergy.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_hydrophobicity">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_normalized_moreau_broto_auto_hydrophobicity</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_normalized_moreau_broto_auto_hydrophobicity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_hydrophobicity" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the NormalizedMoreauBorto Autocorrelation based on hydrophobicity.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Normalized Moreau-Broto Autocorrelation based on Hydrophobicity.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_polarizability">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_normalized_moreau_broto_auto_polarizability</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_normalized_moreau_broto_auto_polarizability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_polarizability" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the NormalizedMoreauBorto Autocorrelation based on Polarizability.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Normalized Moreau-Broto Autocorrelation based on Polarizability.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_residue_asa">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_normalized_moreau_broto_auto_residue_asa</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_normalized_moreau_broto_auto_residue_asa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_residue_asa" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the NormalizedMoreauBorto Autocorrelation based on ResidueASA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Normalized Moreau-Broto Autocorrelation based on ResidueASA.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_residue_vol">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_normalized_moreau_broto_auto_residue_vol</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_normalized_moreau_broto_auto_residue_vol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_residue_vol" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the NormalizedMoreauBorto Autocorrelation based on ResidueVol.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Normalized Moreau-Broto Autocorrelation based on ResidueVol.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_steric">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_normalized_moreau_broto_auto_steric</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_normalized_moreau_broto_auto_steric"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_steric" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the NormalizedMoreauBorto Autocorrelation  based on Steric.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30 Normalized Moreau-Broto Autocorrelation based on Steric.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_total">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">calculate_normalized_moreau_broto_auto_total</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#calculate_normalized_moreau_broto_auto_total"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.calculate_normalized_moreau_broto_auto_total" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for computing normalized Moreau Broto autocorrelation  based on 8 proterties of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing 30*8=240 normalized Moreau Broto autocorrelation  based on the given
properties(i.e., _AAPropert).</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.autocorrelation.normalize_each_aap">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.autocorrelation.</code><code class="sig-name descname">normalize_each_aap</code><span class="sig-paren">(</span><em class="sig-param">aap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/autocorrelation.html#normalize_each_aap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.autocorrelation.normalize_each_aap" title="Permalink to this definition">¶</a></dt>
<dd><p>All of the amino acid indices are centralized and standardized before the calculation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>aap</strong> – AAP is a dict form containing the properties of 20 amino acids</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is the a dict form containing the normalized properties of 20 amino acids.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.binary">
<span id="binary"></span><h2>binary<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.binary" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>This function gives binary profile of aminoacid composition</p>
<p>It receives a sequence and returns a dictionary containing the binary profile</p>
<p>Authors: Ana Marta Sequeira</p>
<p>Date: 05/ 2019</p>
<p>Email:</p>
<hr class="docutils" />
<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.binary.bin_aa_ct">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.binary.</code><code class="sig-name descname">bin_aa_ct</code><span class="sig-paren">(</span><em class="sig-param">seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/binary.html#bin_aa_ct"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.binary.bin_aa_ct" title="Permalink to this definition">¶</a></dt>
<dd><p>Transform sequency in binary profile</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seq</strong> – protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>binary matrix of sequence</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.binary_aa_properties">
<span id="binary-aa-properties"></span><h2>binary_aa_properties<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.binary_aa_properties" title="Permalink to this headline">¶</a></h2>
<p>It receives a sequence and a list containing feature numbers (valid number, 0-24)</p>
<blockquote>
<div><p>FEATURE NAME                    FEATURE NUMBER</p>
<p>‘Positively charged’ –                  0</p>
<p>‘Negatively charged’ –                  1</p>
<p>‘Neutral charged’ –                     2</p>
<p>‘Polarity’ –                            3</p>
<p>‘Non polarity’ –                        4</p>
<p>‘Aliphaticity’ –                        5</p>
<p>‘Cyclic’ –                              6</p>
<p>‘Aromaticity’ –                         7</p>
<p>‘Acidicity’–                            8</p>
<p>‘Basicity’–                             9</p>
<p>‘Neutral (ph)’ –                       10</p>
<p>‘Hydrophobicity’ –                     11</p>
<p>‘Hydrophilicity’ –                     12</p>
<p>‘Neutral’ –                            13</p>
<p>‘Hydroxylic’ –                         14</p>
<p>‘Sulphur content’ -                     15</p>
<p>‘Secondary Structure(Helix)’            16</p>
<p>‘Secondary Structure(Strands)’,         17</p>
<p>‘Secondary Structure(Coil)’,            18</p>
<p>‘Solvent Accessibilty (Buried)’,        19</p>
<p>‘Solvent Accesibilty(Exposed)’,         20</p>
<p>‘Solvent Accesibilty(Intermediate)’,    21</p>
<p>‘Tiny’,                                 22</p>
<p>‘Small’,                                23</p>
<p>‘Large’                                 24</p>
</div></blockquote>
<p>The code is based on the package Pfeature:
Pande, Akshara &amp; Patiyal, Sumeet &amp; Lathwal, Anjali &amp; Arora, Chakit &amp; Kaur, Dilraj &amp; Dhall, Anjali &amp; Mishra, Gaurav &amp; Kaur,
Harpreet &amp; Sharma, Neelam &amp; Jain, Shipra &amp; Usmani, Salman &amp; Agrawal, Piyush &amp; Kumar, Rajesh &amp; Kumar, Vinod &amp; Raghava, Gajendra.
(2019). Computing wide range of protein/peptide features from their sequence and structure. 10.1101/599126.</p>
<p>It returns a dictionary form with the values</p>
<p>Authors: Ana Marta Sequeira</p>
<p>Date: 05/2019</p>
<p>Email:</p>
<hr class="docutils" />
<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.binary_aa_properties.bin_pc_wp">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.binary_aa_properties.</code><code class="sig-name descname">bin_pc_wp</code><span class="sig-paren">(</span><em class="sig-param">sequence</em>, <em class="sig-param">feature_numb</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/binary_aa_properties.html#bin_pc_wp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.binary_aa_properties.bin_pc_wp" title="Permalink to this definition">¶</a></dt>
<dd><p>Binary profile of a sequence based on physicalchemical characterisitics choosed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> – protein sequence</p></li>
<li><p><strong>feature_numb</strong> – list with numbers of the features to be calculated</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary with binary profile of a sequence for a list of desired features</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.binary_aa_properties.binary_profile_all">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.binary_aa_properties.</code><code class="sig-name descname">binary_profile_all</code><span class="sig-paren">(</span><em class="sig-param">seq</em>, <em class="sig-param">feature_numb</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/binary_aa_properties.html#binary_profile_all"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.binary_aa_properties.binary_profile_all" title="Permalink to this definition">¶</a></dt>
<dd><p>binary profile of physicalchemical characterisitics choosed of a sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – protein sequence</p></li>
<li><p><strong>feature_numb</strong> – number of the feature to be calculated</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>value of property for each AA in sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.binary_aa_properties.encode">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.binary_aa_properties.</code><code class="sig-name descname">encode</code><span class="sig-paren">(</span><em class="sig-param">peptide</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/binary_aa_properties.html#encode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.binary_aa_properties.encode" title="Permalink to this definition">¶</a></dt>
<dd><p>encode AA in sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>peptide</strong> – protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>protein sequence with encoded AA</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.binary_aa_properties.init">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.binary_aa_properties.</code><code class="sig-name descname">init</code><span class="sig-paren">(</span><em class="sig-param">path=None</em>, <em class="sig-param">index=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/binary_aa_properties.html#init"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.binary_aa_properties.init" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in files. You need to run this (once) before you can
access any records. If the files are not within the current directory,
you need to specify the correct directory path.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> – </p></li>
<li><p><strong>index</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.binary_aa_properties.lookup">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.binary_aa_properties.</code><code class="sig-name descname">lookup</code><span class="sig-paren">(</span><em class="sig-param">peptide</em>, <em class="sig-param">feature_num</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/binary_aa_properties.html#lookup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.binary_aa_properties.lookup" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptide</strong> – protein sequence</p></li>
<li><p><strong>feature_num</strong> – number of the feature to be calculated</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>value of property for each AA in sequence</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.bondcomp">
<span id="bondcomp"></span><h2>bondcomp<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.bondcomp" title="Permalink to this headline">¶</a></h2>
<p>This function gives the sum of the bond composition for each type of bond
For bond composition four types of bonds are
considered total number of bonds (including aromatic), hydrogen bond, single bond and double
bond. The number of values for each kind of bond is provided as bonds.csv file</p>
<p>The code is based on the package Pfeature :
Pande, Akshara &amp; Patiyal, Sumeet &amp; Lathwal, Anjali &amp; Arora, Chakit &amp; Kaur, Dilraj &amp; Dhall, Anjali &amp; Mishra, Gaurav &amp; Kaur,
Harpreet &amp; Sharma, Neelam &amp; Jain, Shipra &amp; Usmani, Salman &amp; Agrawal, Piyush &amp; Kumar, Rajesh &amp; Kumar, Vinod &amp; Raghava, Gajendra.
(2019). Computing wide range of protein/peptide features from their sequence and structure. 10.1101/599126.</p>
<p>It returns a dictionary form with the values</p>
<p>Authors: Ana Marta Sequeira</p>
<p>Date: 05/2019</p>
<p>Email:</p>
<hr class="docutils" />
<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.bondcomp.boc_wp">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.bondcomp.</code><code class="sig-name descname">boc_wp</code><span class="sig-paren">(</span><em class="sig-param">seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/bondcomp.html#boc_wp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.bondcomp.boc_wp" title="Permalink to this definition">¶</a></dt>
<dd><p>Sum of the bond composition for each type of bond: total number of bonds (including aromatic), hydrogen bond,
single bond and double</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seq</strong> – protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary with number of total, hydrogen, single and double bonds</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.bondcomp.init">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.bondcomp.</code><code class="sig-name descname">init</code><span class="sig-paren">(</span><em class="sig-param">path=None</em>, <em class="sig-param">index=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/bondcomp.html#init"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.bondcomp.init" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in files. You need to run this (once) before you can
access any records. If the files are not within the current directory,
you need to specify the correct directory path.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> – </p></li>
<li><p><strong>index</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.conjoint_triad">
<span id="conjoint-triad"></span><h2>conjoint_triad<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.conjoint_triad" title="Permalink to this headline">¶</a></h2>
<p>protein sequence information. You can get 7*7*7=343 features.You can freely</p>
<p>use and distribute it. If you hava any problem, you could contact with us timely!</p>
<p>Reference:</p>
<p>Juwen Shen, Jian Zhang, Xiaomin Luo, Weiliang Zhu, Kunqian Yu, Kaixian Chen,</p>
<p>Yixue Li, Huanliang Jiang. Predicting proten-protein interactions based only</p>
<p>on sequences inforamtion. PNAS. 2007 (104) 4337-4341.</p>
<p>Authors: Dongsheng Cao and Yizeng Liang.</p>
<p>Date: 2012.09.18</p>
<p>Email: <a class="reference external" href="mailto:oriental-cds&#37;&#52;&#48;163&#46;com">oriental-cds<span>&#64;</span>163<span>&#46;</span>com</a></p>
<p>Altered and converted to python 3.6 by Ana Marta Sequeira 05/2019</p>
<hr class="docutils" />
<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.conjoint_triad.calculate_conjoint_triad">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.conjoint_triad.</code><code class="sig-name descname">calculate_conjoint_triad</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/conjoint_triad.html#calculate_conjoint_triad"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.conjoint_triad.calculate_conjoint_triad" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the conjoint triad features from protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing all 343 conjoint triad features</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.ctd">
<span id="ctd"></span><h2>ctd<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.ctd" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>This module is used for computing the composition, transition and distribution</p>
<p>based on the different properties of AADs. The AADs with the same</p>
<p>properties is marked as the same number. You can get 147  for a given</p>
<p>protein sequence. You can freely use and distribute it. If you hava  any problem,</p>
<p>you could contact with us timely!</p>
<p>References:</p>
<p>[1]: Inna Dubchak, Ilya Muchink, Stephen R.Holbrook and Sung-Hou Kim. Prediction</p>
<p>of protein folding class using global description of amino acid sequence. Proc.Natl.</p>
<p>Acad.Sci.USA, 1995, 92, 8700-8704.</p>
<p>[2]:Inna Dubchak, Ilya Muchink, Christopher Mayor, Igor Dralyuk and Sung-Hou Kim.</p>
<p>Recognition of a Protein Fold in the Context of the SCOP classification. Proteins:</p>
<p>Structure, Function and Genetics,1999,35,401-407.</p>
<p>Authors: Dongsheng Cao and Yizeng Liang.</p>
<p>Date: 2010.11.22</p>
<p>Email: <a class="reference external" href="mailto:oriental-cds&#37;&#52;&#48;163&#46;com">oriental-cds<span>&#64;</span>163<span>&#46;</span>com</a></p>
<p>Altered and converted to python 3.6 by Ana Marta Sequeira 05/2019</p>
<hr class="docutils" />
<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_c">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_c</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_c"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_c" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all composition seven different properties of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing all composition</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_composition">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_composition</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">aa_property</em>, <em class="sig-param">aap_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_composition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_composition" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for computing composition</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>aa_property</strong> – dict form containing classifciation of amino acids such as _Polarizability.</p></li>
<li><p><strong>aap_name</strong> – string used for indicating a AAP name.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing composition based on the given property.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_composition_charge">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_composition_charge</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_composition_charge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_composition_charge" title="Permalink to this definition">¶</a></dt>
<dd><p>alculating composition based on Charge of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Composition based on Charge</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_composition_hydrophobicity">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_composition_hydrophobicity</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_composition_hydrophobicity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_composition_hydrophobicity" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating composition based on Hydrophobicity of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Composition based on Hydrophobicity.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_composition_normalized_vdwv">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_composition_normalized_vdwv</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_composition_normalized_vdwv"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_composition_normalized_vdwv" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating composition based on NormalizedVDWV of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Composition based on NormalizedVDWV</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_composition_polarity">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_composition_polarity</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_composition_polarity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_composition_polarity" title="Permalink to this definition">¶</a></dt>
<dd><p>calculating composition based on Polarity of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Composition based on Polarity</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_composition_polarizability">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_composition_polarizability</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_composition_polarizability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_composition_polarizability" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating composition based on Polarizability of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>ict form containing Composition based on Polarizability</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_composition_secondary_str">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_composition_secondary_str</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_composition_secondary_str"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_composition_secondary_str" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div><p>method used for calculating composition based on SecondaryStr of AADs</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Composition based on SecondaryStr</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_composition_solvent_accessibility">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_composition_solvent_accessibility</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_composition_solvent_accessibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_composition_solvent_accessibility" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating composition based on SolventAccessibility of  AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Composition based on SolventAccessibility</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_ctd">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_ctd</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_ctd"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_ctd" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all CTD based seven different properties of AADs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing all C</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_d">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_d</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_d" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all distribution based seven different properties of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing all distribution</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_distribution">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_distribution</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">aa_property</em>, <em class="sig-param">aap_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_distribution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_distribution" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for computing distribution</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>aa_property</strong> – dict form containing classification of amino acids such as _Polarizability</p></li>
<li><p><strong>aap_name</strong> – string used for indicating a AAP name</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Distribution based on the given property</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_distribution_charge">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_distribution_charge</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_distribution_charge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_distribution_charge" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Distribution based on Charge of   AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Distribution based on Charge</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_distribution_hydrophobicity">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_distribution_hydrophobicity</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_distribution_hydrophobicity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_distribution_hydrophobicity" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Distribution based on Hydrophobicity of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Distribution based on Hydrophobicity.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_distribution_normalized_vdwv">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_distribution_normalized_vdwv</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_distribution_normalized_vdwv"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_distribution_normalized_vdwv" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Distribution based on NormalizedVDWV of AADs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Distribution based on NormalizedVDWV</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_distribution_polarity">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_distribution_polarity</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_distribution_polarity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_distribution_polarity" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Distribution based on Polarity of AADs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Distribution based on Polarity</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_distribution_polarizability">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_distribution_polarizability</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_distribution_polarizability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_distribution_polarizability" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Distribution based on Polarizability of   AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Distribution based on Polarizability</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_distribution_secondary_str">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_distribution_secondary_str</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_distribution_secondary_str"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_distribution_secondary_str" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Distribution based on SecondaryStr of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Distribution based on SecondaryStr.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_distribution_solvent_accessibility">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_distribution_solvent_accessibility</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_distribution_solvent_accessibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_distribution_solvent_accessibility" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Distribution based on SolventAccessibility of  AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Distribution based on SolventAccessibility</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_t">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_t</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_t"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_t" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all transition based seven different properties of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing all transition</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_transition">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_transition</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">aa_property</em>, <em class="sig-param">aap_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_transition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_transition" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for computing transition</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>aa_property</strong> – AAProperty is a dict form containing classifciation of amino acids such as _Polarizability</p></li>
<li><p><strong>aap_name</strong> – string used for indicating a AAP name</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing transition based on the given property</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_transition_charge">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_transition_charge</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_transition_charge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_transition_charge" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Transition based on Charge of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing Transition based on Charge.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_transition_hydrophobicity">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_transition_hydrophobicity</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_transition_hydrophobicity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_transition_hydrophobicity" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Transition based on Hydrophobicity of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Transition based on Hydrophobicity.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_transition_normalized_vdwv">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_transition_normalized_vdwv</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_transition_normalized_vdwv"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_transition_normalized_vdwv" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Transition based on NormalizedVDWV of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing Transition based on NormalizedVDWV.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_transition_polarity">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_transition_polarity</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_transition_polarity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_transition_polarity" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Transition based on Polarity of   AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing Transition based on Polarity.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_transition_polarizability">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_transition_polarizability</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_transition_polarizability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_transition_polarizability" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Transition based on Polarizability of AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Transition based on Polarizability.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_transition_secondary_str">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_transition_secondary_str</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_transition_secondary_str"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_transition_secondary_str" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Transition based on SecondaryStr of       AADs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a dict form containing Transition based on SecondaryStr</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.calculate_transition_solvent_accessibility">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">calculate_transition_solvent_accessibility</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#calculate_transition_solvent_accessibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.calculate_transition_solvent_accessibility" title="Permalink to this definition">¶</a></dt>
<dd><p>A method used for calculating Transition based on SolventAccessibility of  AADs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing Transition based on SolventAccessibility.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.ctd.stringto_num">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.ctd.</code><code class="sig-name descname">stringto_num</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">aa_property</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/ctd.html#stringto_num"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.ctd.stringto_num" title="Permalink to this definition">¶</a></dt>
<dd><p>Tranform the protein sequence into the string form such as 32123223132121123.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence.</p></li>
<li><p><strong>aa_property</strong> – AAProperty is a dict form containing classifciation of amino acids such as _Polarizability</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is a string such as 123321222132111123222</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.descriptors_modlamp">
<span id="descriptors-modlamp"></span><h2>descriptors_modlamp<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.descriptors_modlamp" title="Permalink to this headline">¶</a></h2>
<p>This module incorporates different classes to calculate peptide descriptor values. The following classes are available:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 28%" />
<col style="width: 72%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class</p></th>
<th class="head"><p>Characteristics</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GlobalDescriptor</span></code></p></td>
<td><p>Global one-dimensional peptide pydpi_py3 calculated from the AA sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-class docutils literal notranslate"><span class="pre">PeptideDescriptor</span></code></p></td>
<td><p>AA scale based global or convoluted pydpi_py3 (auto-/cross-correlated).</p></td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-class docutils literal notranslate"><span class="pre">modlamp.core.BaseDescriptor</span></code> from which the classes in <code class="xref py py-mod docutils literal notranslate"><span class="pre">modlamp.pydpi_py3</span></code> inherit.</p>
</div>
<dl class="class">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor">
<em class="property">class </em><code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.descriptors_modlamp.</code><code class="sig-name descname">GlobalDescriptor</code><span class="sig-paren">(</span><em class="sig-param">seqs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor" title="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor"><code class="xref py py-class docutils literal notranslate"><span class="pre">propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor</span></code></a></p>
<p>Base class for global, non-amino acid scale dependant pydpi_py3. The following pydpi_py3 can be calculated by
the <strong>methods</strong> linked below:</p>
<ul class="simple">
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.length">Sequence Length</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.formula">Molecular Formula</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.calculate_MW">Molecular Weight</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.calculate_charge">Sequence Charge</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.charge_density">Charge Density</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.isoelectric_point">Isoelectric Point</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.instability_index">Instability Index</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.aromaticity">Aromaticity</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.aliphatic_index">Aliphatic Index</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.boman_index">Boman Index</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.hydrophobic_ratio">Hydrophobic Ratio</a></p></li>
<li><p><a class="reference external" href="modlamp.html#modlamp.pydpi_py3.GlobalDescriptor.calculate_all">all of the above</a></p></li>
</ul>
<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.aliphatic_index">
<code class="sig-name descname">aliphatic_index</code><span class="sig-paren">(</span><em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.aliphatic_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.aliphatic_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the aliphatic index of every sequence in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.
According to Ikai, 1980, the aliphatic index is a measure of thermal stability of proteins and is dependant
on the relative volume occupied by aliphatic amino acids (A,I,L &amp; V).
([1] A. Ikai, J. Biochem. 1980, 88, 1895–1898.)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of descriptor values in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">(</span><span class="s1">&#39;KWLKYLKKLAKLVK&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">aliphatic_index</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 139.28571429]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.aromaticity">
<code class="sig-name descname">aromaticity</code><span class="sig-paren">(</span><em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.aromaticity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.aromaticity" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the aromaticity of every sequence in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.
According to Lobry, 1994, it is simply the relative frequency of Phe+Trp+Tyr.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of descriptor values in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">(</span><span class="s1">&#39;GLFYWRFFLQRRFLYWW&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">aromaticity</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 0.52941176]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.boman_index">
<code class="sig-name descname">boman_index</code><span class="sig-paren">(</span><em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.boman_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.boman_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the boman index of every sequence in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.
According to Boman, 2003, the boman index is a measure for protein-protein interactions and is calculated by
summing over all amino acid free energy of transfer [kcal/mol] between water and cyclohexane,[2] followed by
dividing by    sequence length.
([1] H. G. Boman, D. Wade, I. a Boman, B. Wåhlin, R. B. Merrifield, <em>FEBS Lett</em>. <strong>1989</strong>, <em>259</em>, 103–106.
[2] A. Radzick, R. Wolfenden, <em>Biochemistry</em> <strong>1988</strong>, <em>27</em>, 1664–1670.)</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">modlamp.core.aa_energies()</span></code></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of descriptor values in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">(</span><span class="s1">&#39;GLFDIVKKVVGALGSL&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">boman_index</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[-1.011875]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.calculate_MW">
<code class="sig-name descname">calculate_MW</code><span class="sig-paren">(</span><em class="sig-param">amide=False</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.calculate_MW"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.calculate_MW" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the molecular weight [g/mol] of every sequence in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>amide</strong> – {boolean} whether the sequences are C-terminally amidated (subtracts 0.95 from the MW).</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of descriptor values in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">(</span><span class="s1">&#39;IAESFKGHIPL&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">calculate_MW</span><span class="p">(</span><span class="n">amide</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 1210.43]])</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">modlamp.core.aa_weights()</span></code></p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version v2.1.5: </span>amide option added</p>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.calculate_all">
<code class="sig-name descname">calculate_all</code><span class="sig-paren">(</span><em class="sig-param">ph=7.4</em>, <em class="sig-param">amide=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.calculate_all"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.calculate_all" title="Permalink to this definition">¶</a></dt>
<dd><p>Method combining all global pydpi_py3 and appending them into the feature matrix in the attribute
<code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ph</strong> – {float} pH at which to calculate peptide charge</p></li>
<li><p><strong>amide</strong> – {boolean} whether the sequences have an amidated C-terminus.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of descriptor values in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">(</span><span class="s1">&#39;AFGHFKLKKLFIFGHERT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">calculate_all</span><span class="p">(</span><span class="n">amide</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">featurenames</span>
<span class="go">[&#39;Length&#39;, &#39;MW&#39;, &#39;ChargeDensity&#39;, &#39;pI&#39;, &#39;InstabilityInd&#39;, &#39;Aromaticity&#39;, &#39;AliphaticInd&#39;, &#39;BomanInd&#39;, &#39;HydRatio&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 18.,  2.17559000e+03,   1.87167619e-03,   1.16757812e+01, ...  1.10555556e+00,   4.44444444e-01]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">save_descriptor</span><span class="p">(</span><span class="s1">&#39;/path/to/outputfile.csv&#39;</span><span class="p">)</span>  <span class="c1"># save the descriptor data (with feature names header)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.calculate_charge">
<code class="sig-name descname">calculate_charge</code><span class="sig-paren">(</span><em class="sig-param">ph=7.0</em>, <em class="sig-param">amide=False</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.calculate_charge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.calculate_charge" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to overall charge of every sequence in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.</p>
<p>The method used is first described by Bjellqvist. In the case of amidation, the value for the ‘Cterm’ pKa is 15
(and Cterm is added to the pos_pKs dictionary.
The pKa scale is extracted from: <a class="reference external" href="http://www.hbcpnetbase.com/">http://www.hbcpnetbase.com/</a> (CRC Handbook of Chemistry and Physics, 96th ed).</p>
<p><strong>pos_pKs</strong> = {‘Nterm’: 9.38, ‘K’: 10.67, ‘R’: 12.10, ‘H’: 6.04}</p>
<p><strong>neg_pKs</strong> = {‘Cterm’: 2.15, ‘D’: 3.71, ‘E’: 4.15, ‘C’: 8.14, ‘Y’: 10.10}</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ph</strong> – {float} ph at which to calculate peptide charge.</p></li>
<li><p><strong>amide</strong> – {boolean} whether the sequences have an amidated C-terminus.</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of descriptor values in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">(</span><span class="s1">&#39;KLAKFGKRSELVALSG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">calculate_charge</span><span class="p">(</span><span class="n">ph</span><span class="o">=</span><span class="mf">7.4</span><span class="p">,</span> <span class="n">amide</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 3.989]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.charge_density">
<code class="sig-name descname">charge_density</code><span class="sig-paren">(</span><em class="sig-param">ph=7.0</em>, <em class="sig-param">amide=False</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.charge_density"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.charge_density" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the charge density (charge / MW) of every sequences in the attributes <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ph</strong> – {float} pH at which to calculate peptide charge.</p></li>
<li><p><strong>amide</strong> – {boolean} whether the sequences have an amidated C-terminus.</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of descriptor values in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">(</span><span class="s1">&#39;GNSDLLIEQRTLLASDEF&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">charge_density</span><span class="p">(</span><span class="n">ph</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">amide</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[-0.00097119]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.formula">
<code class="sig-name descname">formula</code><span class="sig-paren">(</span><em class="sig-param">amide=False</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.formula"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.formula" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the molecular formula of every sequence in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>amide</strong> – {boolean} whether the sequences are C-terminally amidated.</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of molecular formulas {str} in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">([</span><span class="s1">&#39;KADSFLSADGHSADFSLDKKLKERL&#39;</span><span class="p">,</span> <span class="s1">&#39;ERTILSDFPQWWFASLDFLNC&#39;</span><span class="p">,</span> <span class="s1">&#39;ACDEFGHIKLMNPQRSTVWY&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">formula</span><span class="p">(</span><span class="n">amide</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">v</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">C122 H197 N35 O39</span>
<span class="go">C121 H168 N28 O33 S</span>
<span class="go">C106 H157 N29 O30 S2</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">modlamp.core.aa_formulas()</span></code></p>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version v2.7.6.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.hydrophobic_ratio">
<code class="sig-name descname">hydrophobic_ratio</code><span class="sig-paren">(</span><em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.hydrophobic_ratio"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.hydrophobic_ratio" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the hydrophobic ratio of every sequence in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>, which is the
relative frequency of the amino acids <strong>A,C,F,I,L,M &amp; V</strong>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of descriptor values in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">(</span><span class="s1">&#39;VALLYWRTVLLAIII&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">hydrophobic_ratio</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 0.73333333]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.instability_index">
<code class="sig-name descname">instability_index</code><span class="sig-paren">(</span><em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.instability_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.instability_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the instability of every sequence in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.
The instability index is a prediction of protein stability based on the amino acid composition.
([1] K. Guruprasad, B. V Reddy, M. W. Pandit, Protein Eng. 1990, 4, 155–161.)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of descriptor values in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">(</span><span class="s1">&#39;LLASMNDLLAKRST&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">instability_index</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 63.95714286]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.isoelectric_point">
<code class="sig-name descname">isoelectric_point</code><span class="sig-paren">(</span><em class="sig-param">amide=False</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.isoelectric_point"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.isoelectric_point" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the isoelectric point of every sequence in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.
The pK scale is extracted from: <a class="reference external" href="http://www.hbcpnetbase.com/">http://www.hbcpnetbase.com/</a> (CRC Handbook of Chemistry and Physics, 96th ed).</p>
<blockquote>
<div><p><strong>pos_pKs</strong> = {‘Nterm’: 9.38, ‘K’: 10.67, ‘R’: 12.10, ‘H’: 6.04}</p>
<p><strong>neg_pKs</strong> = {‘Cterm’: 2.15, ‘D’: 3.71, ‘E’: 4.15, ‘C’: 8.14, ‘Y’: 10.10}</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>amide</strong> – {boolean} whether the sequences have an amidated C-terminus.</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of descriptor values in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">(</span><span class="s1">&#39;KLFDIKFGHIPQRST&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">isoelectric_point</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 10.6796875]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.length">
<code class="sig-name descname">length</code><span class="sig-paren">(</span><em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#GlobalDescriptor.length"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor.length" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to calculate the length (total AA count) of every sequence in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>array of sequence lengths in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">GlobalDescriptor</span><span class="p">([</span><span class="s1">&#39;AFDGHLKI&#39;</span><span class="p">,</span><span class="s1">&#39;KKLQRSDLLRTK&#39;</span><span class="p">,</span><span class="s1">&#39;KKLASCNNIPPR&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">length</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 8.], [12.], [12.]])</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor">
<em class="property">class </em><code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.descriptors_modlamp.</code><code class="sig-name descname">PeptideDescriptor</code><span class="sig-paren">(</span><em class="sig-param">seqs</em>, <em class="sig-param">scalename='Eisenberg'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#PeptideDescriptor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor" title="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor"><code class="xref py py-class docutils literal notranslate"><span class="pre">propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor</span></code></a></p>
<p>Base class for peptide pydpi_py3. The following <strong>amino acid descriptor scales</strong> are available for descriptor
calculation:</p>
<ul class="simple">
<li><p><strong>AASI</strong>           (An amino acid selectivity index scale for helical antimicrobial peptides, <em>[1] D. Juretić, D. Vukicević, N. Ilić, N. Antcheva, A. Tossi, J. Chem. Inf. Model. 2009, 49, 2873–2882.</em>)</p></li>
<li><p><strong>ABHPRK</strong>         (modlabs inhouse physicochemical feature scale (Acidic, Basic, Hydrophobic, Polar, aRomatic, Kink-inducer)</p></li>
<li><p><strong>argos</strong>          (Argos hydrophobicity amino acid scale, <em>[2] Argos, P., Rao, J. K. M. &amp; Hargrave, P. A., Eur. J. Biochem. 2005, 128, 565–575.</em>)</p></li>
<li><p><strong>bulkiness</strong>      (Amino acid side chain bulkiness scale, <em>[3] J. M. Zimmerman, N. Eliezer, R. Simha, J. Theor. Biol. 1968, 21, 170–201.</em>)</p></li>
<li><p><strong>charge_phys</strong>    (Amino acid charge at pH 7.0 - Hystidine charge +0.1.)</p></li>
<li><p><strong>charge_acid</strong>    (Amino acid charge at acidic pH - Hystidine charge +1.0.)</p></li>
<li><p><strong>cougar</strong>         (modlabs inhouse selection of global peptide pydpi_py3)</p></li>
<li><p><strong>eisenberg</strong>      (the Eisenberg hydrophobicity consensus amino acid scale, <em>[4] D. Eisenberg, R. M. Weiss, T. C. Terwilliger, W. Wilcox, Faraday Symp. Chem. Soc. 1982, 17, 109.</em>)</p></li>
<li><p><strong>Ez</strong>             (potential that assesses energies of insertion of amino acid side chains into lipid bilayers, <em>[5] A. Senes, D. C. Chadi, P. B. Law, R. F. S. Walters, V. Nanda, W. F. DeGrado, J. Mol. Biol. 2007, 366, 436–448.</em>)</p></li>
<li><p><strong>flexibility</strong>    (amino acid side chain flexibilitiy scale, <em>[6] R. Bhaskaran, P. K. Ponnuswamy, Int. J. Pept. Protein Res. 1988, 32, 241–255.</em>)</p></li>
<li><p><strong>grantham</strong>       (amino acid side chain composition, polarity and molecular volume, <em>[8] Grantham, R. Science. 185, 862–864 (1974).</em>)</p></li>
<li><p><strong>gravy</strong>          (GRAVY hydrophobicity amino acid scale, <em>[9] J. Kyte, R. F. Doolittle, J. Mol. Biol. 1982, 157, 105–132.</em>)</p></li>
<li><p><strong>hopp-woods</strong>     (Hopp-Woods amino acid hydrophobicity scale,*[10] T. P. Hopp, K. R. Woods, Proc. Natl. Acad. Sci. 1981, 78, 3824–3828.*)</p></li>
<li><p><strong>ISAECI</strong>         (Isotropic Surface Area (ISA) and Electronic Charge Index (ECI) of amino acid side chains, <em>[11] E. R. Collantes, W. J. Dunn, J. Med. Chem. 1995, 38, 2705–2713.</em>)</p></li>
<li><p><strong>janin</strong>          (Janin hydrophobicity amino acid scale, <em>[12] J. L. Cornette, K. B. Cease, H. Margalit, J. L. Spouge, J. A. Berzofsky, C. DeLisi, J. Mol. Biol. 1987, 195, 659–685.</em>)</p></li>
<li><p><strong>kytedoolittle</strong>  (Kyte &amp; Doolittle hydrophobicity amino acid scale, <em>[13] J. Kyte, R. F. Doolittle, J. Mol. Biol. 1982, 157, 105–132.</em>)</p></li>
<li><p><strong>levitt_alpha</strong>   (Levitt amino acid alpha-helix propensity scale, extracted from <a class="reference external" href="http://web.expasy.org/protscale">http://web.expasy.org/protscale</a>. <em>[14] M. Levitt, Biochemistry 1978, 17, 4277-4285.</em>)</p></li>
<li><p><strong>MSS</strong>            (A graph-theoretical index that reflects topological shape and size of amino acid side chains, <em>[15] C. Raychaudhury, A. Banerjee, P. Bag, S. Roy, J. Chem. Inf. Comput. Sci. 1999, 39, 248–254.</em>)</p></li>
<li><p><strong>MSW</strong>            (Amino acid scale based on a pca of the molecular surface based WHIM descriptor (MS-WHIM), extended to natural amino acids, <em>[16] A. Zaliani, E. Gancia, J. Chem. Inf. Comput. Sci 1999, 39, 525–533.</em>)</p></li>
<li><p><strong>pepArc</strong>         (modlabs pharmacophoric feature scale, dimensions are: hydrophobicity, polarity, positive charge, negative charge, proline.)</p></li>
<li><p><strong>pepcats</strong>        (modlabs pharmacophoric feature based PEPCATS scale, <em>[17] C. P. Koch, A. M. Perna, M. Pillong, N. K. Todoroff, P. Wrede, G. Folkers, J. A. Hiss, G. Schneider, PLoS Comput. Biol. 2013, 9, e1003088.</em>)</p></li>
<li><p><strong>polarity</strong>       (Amino acid polarity scale, <em>[18] J. M. Zimmerman, N. Eliezer, R. Simha, J. Theor. Biol. 1968, 21, 170–201.</em>)</p></li>
<li><p><strong>PPCALI</strong>         (modlabs inhouse scale derived from a pca of 143 amino acid property scales, <em>[19] C. P. Koch, A. M. Perna, M. Pillong, N. K. Todoroff, P. Wrede, G. Folkers, J. A. Hiss, G. Schneider, PLoS Comput. Biol. 2013, 9, e1003088.</em>)</p></li>
<li><p><strong>refractivity</strong>   (Relative amino acid refractivity values, <em>[20] T. L. McMeekin, M. Wilensky, M. L. Groves, Biochem. Biophys. Res. Commun. 1962, 7, 151–156.</em>)</p></li>
<li><p><strong>t_scale</strong>        (A pca derived scale based on amino acid side chain properties calculated with 6 different probes of the GRID program, <em>[21] M. Cocchi, E. Johansson, Quant. Struct. Act. Relationships 1993, 12, 1–8.</em>)</p></li>
<li><p><strong>TM_tend</strong>        (Amino acid transmembrane propensity scale, extracted from <a class="reference external" href="http://web.expasy.org/protscale">http://web.expasy.org/protscale</a>, <em>[22] Zhao, G., London E. Protein Sci. 2006, 15, 1987-2001.</em>)</p></li>
<li><p><strong>z3</strong>             (The original three dimensional Z-scale, <em>[23] S. Hellberg, M. Sjöström, B. Skagerberg, S. Wold, J. Med. Chem. 1987, 30, 1126–1135.</em>)</p></li>
<li><p><strong>z5</strong>             (The extended five dimensional Z-scale, <em>[24] M. Sandberg, L. Eriksson, J. Jonsson, M. Sjöström, S. Wold, J. Med. Chem. 1998, 41, 2481–2491.</em>)</p></li>
</ul>
<p>Further, amino acid scale independent methods can be calculated with help of the <a class="reference internal" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor" title="propythia.adjuv_functions.features_functions.descriptors_modlamp.GlobalDescriptor"><code class="xref py py-class docutils literal notranslate"><span class="pre">GlobalDescriptor</span></code></a> class.</p>
<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_arc">
<code class="sig-name descname">calculate_arc</code><span class="sig-paren">(</span><em class="sig-param">modality='max'</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#PeptideDescriptor.calculate_arc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_arc" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for calculating property arcs as seen in the helical wheel plot. Use for binary amino acid scales only.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>modality</strong> – modality of the arc to calculate, to choose between “max” and “mean”.</p></li>
<li><p><strong>append</strong> – if true, append to current descriptor stored in the descriptor attribute.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>calculated descriptor as numpy.array in the descriptor attribute.</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">arc</span> <span class="o">=</span> <span class="n">PeptideDescriptor</span><span class="p">(</span><span class="s2">&quot;KLLKLLKKLLKLLK&quot;</span><span class="p">,</span> <span class="n">scalename</span><span class="o">=</span><span class="s2">&quot;peparc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">arc</span><span class="o">.</span><span class="n">calculate_arc</span><span class="p">(</span><span class="n">modality</span><span class="o">=</span><span class="s2">&quot;max&quot;</span><span class="p">,</span> <span class="n">append</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">arc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[200, 160, 160,   0,   0]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_autocorr">
<code class="sig-name descname">calculate_autocorr</code><span class="sig-paren">(</span><em class="sig-param">window</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#PeptideDescriptor.calculate_autocorr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_autocorr" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for auto-correlating the amino acid values for a given descriptor scale</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> – {int} correlation window for descriptor calculation in a sliding window approach</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>calculated descriptor numpy.array in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span> <span class="o">=</span> <span class="n">PeptideDescriptor</span><span class="p">(</span><span class="s1">&#39;GLFDIVKKVVGALGSL&#39;</span><span class="p">,</span><span class="s1">&#39;PPCALI&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">calculate_autocorr</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[  1.28442339e+00,   1.29025116e+00,   1.03240901e+00, .... ]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">descriptor</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(1, 133)</span>
</pre></div>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version v.2.3.0.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_crosscorr">
<code class="sig-name descname">calculate_crosscorr</code><span class="sig-paren">(</span><em class="sig-param">window</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#PeptideDescriptor.calculate_crosscorr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_crosscorr" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for cross-correlating the amino acid values for a given descriptor scale</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> – {int} correlation window for descriptor calculation in a sliding window approach</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>calculated descriptor numpy.array in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span> <span class="o">=</span> <span class="n">PeptideDescriptor</span><span class="p">(</span><span class="s1">&#39;GLFDIVKKVVGALGSL&#39;</span><span class="p">,</span><span class="s1">&#39;pepcats&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">calculate_crosscorr</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 0.6875    ,  0.46666667,  0.42857143,  0.61538462,  0.58333333, ... ]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">descriptor</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(1, 147)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_global">
<code class="sig-name descname">calculate_global</code><span class="sig-paren">(</span><em class="sig-param">window=1000</em>, <em class="sig-param">modality='max'</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#PeptideDescriptor.calculate_global"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_global" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for calculating a global / window averaging descriptor value of a given AA scale</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> – {int} amino acid window in which to calculate the moment. If the sequence is shorter than the
window, the length of the sequence is taken.</p></li>
<li><p><strong>modality</strong> – {‘max’ or ‘mean’} Calculate respectively maximum or mean hydrophobic moment.</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Calculated descriptor as a numpy.array in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code> and all possible global
values in <code class="xref py py-attr docutils literal notranslate"><span class="pre">all_globs</span></code> (needed for the <a class="reference internal" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_profile" title="propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_profile"><code class="xref py py-func docutils literal notranslate"><span class="pre">calculate_profile()</span></code></a> method)</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span> <span class="o">=</span> <span class="n">PeptideDescriptor</span><span class="p">(</span><span class="s1">&#39;GLFDIVKKVVGALGSL&#39;</span><span class="p">,</span><span class="s1">&#39;eisenberg&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">calculate_global</span><span class="p">(</span><span class="n">window</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">modality</span><span class="o">=</span><span class="s1">&#39;max&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 0.44875]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_moment">
<code class="sig-name descname">calculate_moment</code><span class="sig-paren">(</span><em class="sig-param">window=1000</em>, <em class="sig-param">angle=100</em>, <em class="sig-param">modality='max'</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#PeptideDescriptor.calculate_moment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_moment" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for calculating the maximum or mean moment of the amino acid values for a given descriptor scale and
window.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> – {int} amino acid window in which to calculate the moment. If the sequence is shorter than the
window, the length of the sequence is taken. So if the default window of 1000 is chosen, for all sequences
shorter than 1000, the <strong>global</strong> hydrophobic moment will be calculated. Otherwise, the maximal
hydrophiobic moment for the chosen window size found in the sequence will be returned.</p></li>
<li><p><strong>angle</strong> – {int} angle in which to calculate the moment. <strong>100</strong> for alpha helices, <strong>180</strong> for beta sheets.</p></li>
<li><p><strong>modality</strong> – {‘all’, ‘max’ or ‘mean’} Calculate respectively maximum or mean hydrophobic moment. If all,
moments for all windows are returned.</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Calculated descriptor as a numpy.array in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code> and all possible global
values in <code class="xref py py-attr docutils literal notranslate"><span class="pre">all_moms</span></code> (needed for the <a class="reference internal" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_profile" title="propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_profile"><code class="xref py py-func docutils literal notranslate"><span class="pre">calculate_profile()</span></code></a> method)</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span> <span class="o">=</span> <span class="n">PeptideDescriptor</span><span class="p">(</span><span class="s1">&#39;GLFDIVKKVVGALGSL&#39;</span><span class="p">,</span> <span class="s1">&#39;eisenberg&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">calculate_moment</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 0.48790226]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_profile">
<code class="sig-name descname">calculate_profile</code><span class="sig-paren">(</span><em class="sig-param">prof_type='uH'</em>, <em class="sig-param">window=7</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#PeptideDescriptor.calculate_profile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.calculate_profile" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for calculating hydrophobicity or hydrophobic moment profiles for given sequences and fitting for
slope and intercept. The hydrophobicity scale used is “eisenberg”</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prof_type</strong> – prof_type of profile, available: ‘H’ for hydrophobicity or ‘uH’ for hydrophobic moment</p></li>
<li><p><strong>window</strong> – {int} size of sliding window used (odd-numbered).</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Fitted slope and intercept of calculated profile for every given sequence in the attribute
<code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span> <span class="o">=</span> <span class="n">PeptideDescriptor</span><span class="p">(</span><span class="s1">&#39;KLLKLLKKVVGALG&#39;</span><span class="p">,</span><span class="s1">&#39;kytedoolittle&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">calculate_profile</span><span class="p">(</span><span class="n">prof_type</span><span class="o">=</span><span class="s1">&#39;H&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 0.03731293,  0.19246599]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.load_scale">
<code class="sig-name descname">load_scale</code><span class="sig-paren">(</span><em class="sig-param">scalename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/descriptors_modlamp.html#PeptideDescriptor.load_scale"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.descriptors_modlamp.PeptideDescriptor.load_scale" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to load amino acid values from a given scale</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>scalename</strong> – {str} name of the amino acid scale to be loaded.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>loaded amino acid scale values in a dictionary in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">scale</span></code>.</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">modlamp.core.load_scale()</span></code></p>
</div>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.modlamp_core">
<span id="modlamp-core"></span><h2>modlamp_core<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.modlamp_core" title="Permalink to this headline">¶</a></h2>
<p>Core helper functions and classes for other modules. The two main classes are:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 75%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Class</p></th>
<th class="head"><p>Characteristics</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseSequence</span></code></p></td>
<td><p>Base class inheriting to all sequence classes in the module <code class="xref py py-mod docutils literal notranslate"><span class="pre">modlamp.sequences</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseDescriptor</span></code></p></td>
<td><p>Base class inheriting to the two descriptor classes in <code class="xref py py-mod docutils literal notranslate"><span class="pre">modlamp.pydpi_py3</span></code></p></td>
</tr>
</tbody>
</table>
<dl class="class">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor">
<em class="property">class </em><code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.modlamp_core.</code><code class="sig-name descname">BaseDescriptor</code><span class="sig-paren">(</span><em class="sig-param">seqs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class inheriting to both peptide descriptor classes <code class="xref py py-class docutils literal notranslate"><span class="pre">modlamp.pydpi_py3.GlobalDescriptor</span></code> and
<code class="xref py py-class docutils literal notranslate"><span class="pre">modlamp.pydpi_py3.PeptideDescriptor</span></code>.</p>
<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.count_aa">
<code class="sig-name descname">count_aa</code><span class="sig-paren">(</span><em class="sig-param">scale='relative'</em>, <em class="sig-param">average=False</em>, <em class="sig-param">append=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.count_aa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.count_aa" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for producing the amino acid distribution for the given sequences as a descriptor</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scale</strong> – {‘absolute’ or ‘relative’} defines whether counts or frequencies are given for each AA</p></li>
<li><p><strong>average</strong> – {boolean} whether the averaged amino acid counts for all sequences should be returned</p></li>
<li><p><strong>append</strong> – {boolean} whether the produced descriptor values should be appended to the existing ones in the
attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the amino acid distributions for every sequence individually in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span> <span class="o">=</span> <span class="n">PeptideDescriptor</span><span class="p">(</span><span class="s1">&#39;ACDEFGHIKLMNPQRSTVWY&#39;</span><span class="p">)</span> <span class="c1"># aa_count() does not depend on the descriptor scale</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">count_aa</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 0.05,  0.05,  0.05,  0.05,  0.05,  0.05,  0.05,  0.05,  0.05, ... ]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AMP</span><span class="o">.</span><span class="n">descriptor</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(1, 20)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-func docutils literal notranslate"><span class="pre">modlamp.core.count_aa()</span></code></p>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.feature_scaling">
<code class="sig-name descname">feature_scaling</code><span class="sig-paren">(</span><em class="sig-param">stype='standard'</em>, <em class="sig-param">fit=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.feature_scaling"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.feature_scaling" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for feature scaling of the calculated descriptor matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>stype</strong> – {‘standard’ or ‘minmax’} type of scaling to be used</p></li>
<li><p><strong>fit</strong> – {boolean} defines whether the used scaler is first fitting on the data (True) or
whether the already fitted scaler in <code class="xref py py-attr docutils literal notranslate"><span class="pre">scaler</span></code> should be used to transform (False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>scaled descriptor values in <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">D</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[0.155],[0.34],[0.16235294],[-0.08842105],[0.116]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">D</span><span class="o">.</span><span class="n">feature_scaling</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="s1">&#39;minmax&#39;</span><span class="p">,</span><span class="n">fit</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">array([[0.56818182],[1.],[0.5853447],[0.],[0.47714988]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.feature_shuffle">
<code class="sig-name descname">feature_shuffle</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.feature_shuffle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.feature_shuffle" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for shuffling feature columns randomly.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>descriptor matrix with shuffled feature columns in <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code></p>
</dd>
<dt class="field-even">Example</dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">D</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[0.80685625,167.05234375,39.56818125,-0.26338667,155.16888667,33.48778]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">D</span><span class="o">.</span><span class="n">feature_shuffle</span><span class="p">()</span>
<span class="go">array([[155.16888667,-0.26338667,167.05234375,0.80685625,39.56818125,33.48778]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.filter_aa">
<code class="sig-name descname">filter_aa</code><span class="sig-paren">(</span><em class="sig-param">amino_acids</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.filter_aa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.filter_aa" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to filter out corresponding names and descriptor values of sequences with given amino acids in the
argument list <em>aminoacids</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>amino_acids</strong> – list of amino acids to be filtered</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>filtered list of sequences, descriptor values, target values and names in the corresponding attributes.</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">BaseSequence</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AAALLLIIIKKK&#39;</span><span class="p">,</span> <span class="s1">&#39;CCEERRT&#39;</span><span class="p">,</span> <span class="s1">&#39;LLVVIIFFFQQ&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">filter_aa</span><span class="p">([</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;AAALLLIIIKKK&#39;, &#39;LLVVIIFFFQQ&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.filter_duplicates">
<code class="sig-name descname">filter_duplicates</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.filter_duplicates"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.filter_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to filter duplicates in the sequences from the class attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>filtered sequences list in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> and corresponding names.</p>
</dd>
<dt class="field-even">Example</dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">BaseDescriptor</span><span class="p">([</span><span class="s1">&#39;KLLKLLKKLLKLLK&#39;</span><span class="p">,</span> <span class="s1">&#39;KLLKLLKKLLKLLK&#39;</span><span class="p">,</span> <span class="s1">&#39;KLAKLAKKLAKLAK&#39;</span><span class="p">,</span> <span class="s1">&#39;KLAKLAKKLAKLAK&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">filter_duplicates</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;KLLKLLKKLLKLLK&#39;, &#39;KLAKLAKKLAKLAK&#39;]</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version v2.2.5.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.filter_sequences">
<code class="sig-name descname">filter_sequences</code><span class="sig-paren">(</span><em class="sig-param">sequences</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.filter_sequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.filter_sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to filter out entries for given sequences in <em>sequences</em> out of a descriptor instance. All
corresponding attribute values of these sequences (e.g. in <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code>, <code class="xref py py-attr docutils literal notranslate"><span class="pre">name</span></code>) are deleted
as well. The method returns an updated descriptor instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequences</strong> – {list} sequences to be filtered out of the whole instance, including corresponding data</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>updated instance without filtered sequences</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KLLKLLKKLLKLLK&#39;</span><span class="p">,</span> <span class="s1">&#39;ACDEFGHIK&#39;</span><span class="p">,</span> <span class="s1">&#39;GLFDIVKKVV&#39;</span><span class="p">,</span> <span class="s1">&#39;GLFDIVKKVVGALG&#39;</span><span class="p">,</span> <span class="s1">&#39;GLFDIVKKVVGALGSL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">PeptideDescriptor</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="s1">&#39;pepcats&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">calculate_crosscorr</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">filter_sequences</span><span class="p">(</span><span class="s1">&#39;KLLKLLKKLLKLLK&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;ACDEFGHIK&#39;, &#39;GLFDIVKKVV&#39;, &#39;GLFDIVKKVVGALG&#39;, &#39;GLFDIVKKVVGALGSL&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.filter_values">
<code class="sig-name descname">filter_values</code><span class="sig-paren">(</span><em class="sig-param">values</em>, <em class="sig-param">operator='=='</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.filter_values"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.filter_values" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to filter the descriptor matrix in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code> for a given list of values (same
size as the number of features in the descriptor matrix!) The operator option tells the method whether to
filter for values equal, lower, higher ect. to the given values in the <em>values</em> array.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>values</strong> – {list} values to filter the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">descriptor</span></code> for</p></li>
<li><p><strong>operator</strong> – {str} filter criterion, available the operators <code class="docutils literal notranslate"><span class="pre">==</span></code>, <code class="docutils literal notranslate"><span class="pre">&lt;</span></code>, <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>, <code class="docutils literal notranslate"><span class="pre">&lt;=``and</span> <span class="pre">``&gt;=</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>descriptor matrix and updated sequences containing only entries with descriptor values given in
<em>values</em> in the corresponding attributes.</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>  <span class="c1"># desc = BaseDescriptor instance</span>
<span class="go">array([[ 0.7666517 ],</span>
<span class="go">       [ 0.38373498]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">filter_values</span><span class="p">([</span><span class="mf">0.5</span><span class="p">],</span> <span class="s1">&#39;&lt;&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span>
<span class="go">array([[ 0.38373498]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.keep_natural_aa">
<code class="sig-name descname">keep_natural_aa</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.keep_natural_aa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.keep_natural_aa" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to filter out sequences that do not contain natural amino acids. If the sequence contains a character
that is not in [‘A’,’C’,’D,’E’,’F’,’G’,’H’,’I’,’K’,’L’,’M’,’N’,’P’,’Q’,’R’,’S’,’T’,’V’,’W’,’Y’].</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>filtered sequence list in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>. The other attributes are also filtered
accordingly (if present).</p>
</dd>
<dt class="field-even">Example</dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">BaseSequence</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;BBBsdflUasUJfBJ&#39;</span><span class="p">,</span> <span class="s1">&#39;GLFDIVKKVVGALGSL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">keep_natural_aa</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;GLFDIVKKVVGALGSL&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.load_descriptordata">
<code class="sig-name descname">load_descriptordata</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">delimiter='</em>, <em class="sig-param">'</em>, <em class="sig-param">targets=False</em>, <em class="sig-param">skip_header=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.load_descriptordata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.load_descriptordata" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to load any data file with sequences and descriptor values and save it to a new insatnce of the
class <code class="xref py py-class docutils literal notranslate"><span class="pre">modlamp.pydpi_py3.PeptideDescriptor</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Headers are not considered. To skip initial lines in the file, use the <em>skip_header</em> option.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> – {str} filename of the data file to be loaded</p></li>
<li><p><strong>delimiter</strong> – {str} column delimiter</p></li>
<li><p><strong>targets</strong> – {boolean} whether last column in the file contains a target class vector</p></li>
<li><p><strong>skip_header</strong> – {int} number of initial lines to skip in the file</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>loaded sequences, descriptor values and targets in the corresponding attributes.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.minmax_selection">
<code class="sig-name descname">minmax_selection</code><span class="sig-paren">(</span><em class="sig-param">iterations</em>, <em class="sig-param">distmetric='euclidean'</em>, <em class="sig-param">seed=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.minmax_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.minmax_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to select a specified number of sequences according to the minmax algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>iterations</strong> – {int} Number of sequences to retrieve.</p></li>
<li><p><strong>distmetric</strong> – Distance metric to calculate the distances between the sequences in descriptor space.
Choose from ‘euclidean’ or ‘minkowsky’.</p></li>
<li><p><strong>seed</strong> – {int} Set a random seed for numpy to pick the first sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>updated instance</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><strong>SciPy</strong> <a class="reference external" href="http://docs.scipy.org/doc/scipy/reference/spatial.distance.html">http://docs.scipy.org/doc/scipy/reference/spatial.distance.html</a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.random_selection">
<code class="sig-name descname">random_selection</code><span class="sig-paren">(</span><em class="sig-param">num</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.random_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.random_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to randomly select a specified number of sequences (with names and pydpi_py3 if present) out of a given
descriptor instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num</strong> – {int} number of entries to be randomly selected</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>updated instance</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">h</span> <span class="o">=</span> <span class="n">Helices</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">28</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">h</span><span class="o">.</span><span class="n">generate_helices</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span> <span class="o">=</span> <span class="n">PeptideDescriptor</span><span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">sequences</span><span class="p">,</span> <span class="s1">&#39;eisenberg&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">calculate_moment</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">desc</span><span class="o">.</span><span class="n">sequences</span><span class="p">)</span>
<span class="go">100</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span><span class="p">)</span>
<span class="go">100</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span><span class="o">.</span><span class="n">random_selection</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span><span class="p">)</span>
<span class="go">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">desc</span><span class="o">.</span><span class="n">descriptor</span><span class="p">)</span>
<span class="go">10</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version v2.2.3.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.read_fasta">
<code class="sig-name descname">read_fasta</code><span class="sig-paren">(</span><em class="sig-param">filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.read_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.read_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for loading sequences from a <code class="docutils literal notranslate"><span class="pre">.FASTA</span></code> formatted file into the attributes <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> and
<code class="xref py py-attr docutils literal notranslate"><span class="pre">names</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filename</strong> – {str} <code class="docutils literal notranslate"><span class="pre">.FASTA</span></code> file with sequences and headers to read</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>{list} sequences in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> with corresponding sequence names in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">names</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.save_descriptor">
<code class="sig-name descname">save_descriptor</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">delimiter='</em>, <em class="sig-param">'</em>, <em class="sig-param">targets=None</em>, <em class="sig-param">header=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.save_descriptor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.save_descriptor" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to save the descriptor values to a .csv/.txt file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> – filename of the output file</p></li>
<li><p><strong>delimiter</strong> – column delimiter</p></li>
<li><p><strong>targets</strong> – target class vector to be added to descriptor (same length as <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>)</p></li>
<li><p><strong>header</strong> – {str} header to be written at the beginning of the file (if <code class="docutils literal notranslate"><span class="pre">None</span></code>: feature names are taken)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>output file with peptide names and descriptor values</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.save_fasta">
<code class="sig-name descname">save_fasta</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">names=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.save_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.save_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for saving sequences from <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> to a <code class="docutils literal notranslate"><span class="pre">.FASTA</span></code> formatted file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> – {str} filename of the output <code class="docutils literal notranslate"><span class="pre">.FASTA</span></code> file</p></li>
<li><p><strong>names</strong> – {bool} whether sequence names from self.names should be saved as sequence identifiers</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a FASTA formatted file containing the generated sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.sequence_order_shuffle">
<code class="sig-name descname">sequence_order_shuffle</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseDescriptor.sequence_order_shuffle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseDescriptor.sequence_order_shuffle" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for shuffling sequence order in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>sequences in <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> with shuffled order in the list.</p>
</dd>
<dt class="field-even">Example</dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">D</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;LILRALKGAARALKVA&#39;,&#39;VKIAKIALKIIKGLG&#39;,&#39;VGVRLIKGIGRVARGAI&#39;,&#39;LRGLRGVIRGGKAIVRVGK&#39;,&#39;GGKLVRLIARIGKGV&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">D</span><span class="o">.</span><span class="n">sequence_order_shuffle</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">D</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;VGVRLIKGIGRVARGAI&#39;,&#39;LILRALKGAARALKVA&#39;,&#39;LRGLRGVIRGGKAIVRVGK&#39;,&#39;GGKLVRLIARIGKGV&#39;,&#39;VKIAKIALKIIKGLG&#39;]</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence">
<em class="property">class </em><code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.modlamp_core.</code><code class="sig-name descname">BaseSequence</code><span class="sig-paren">(</span><em class="sig-param">seqnum</em>, <em class="sig-param">lenmin=7</em>, <em class="sig-param">lenmax=28</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseSequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for sequence classes in the module <code class="xref py py-mod docutils literal notranslate"><span class="pre">modlamp.sequences</span></code>.
It contains amino acid probabilities for different sequence generation classes.</p>
<p>The following amino acid probabilities are used: (extracted from the
<a class="reference external" href="http://aps.unmc.edu/AP/statistic/statistic.php">APD3</a>, March 17, 2016)</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 9%" />
<col style="width: 13%" />
<col style="width: 19%" />
<col style="width: 28%" />
<col style="width: 31%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>AA</p></th>
<th class="head"><p>rand</p></th>
<th class="head"><p>AMP</p></th>
<th class="head"><p>AMPnoCM</p></th>
<th class="head"><p>randnoCM</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>A</p></td>
<td><p>0.05</p></td>
<td><p>0.0766</p></td>
<td><p>0.0812275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-odd"><td><p>C</p></td>
<td><p>0.05</p></td>
<td><p>0.071</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
</tr>
<tr class="row-even"><td><p>D</p></td>
<td><p>0.05</p></td>
<td><p>0.026</p></td>
<td><p>0.0306275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-odd"><td><p>E</p></td>
<td><p>0.05</p></td>
<td><p>0.0264</p></td>
<td><p>0.0310275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-even"><td><p>F</p></td>
<td><p>0.05</p></td>
<td><p>0.0405</p></td>
<td><p>0.0451275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-odd"><td><p>G</p></td>
<td><p>0.05</p></td>
<td><p>0.1172</p></td>
<td><p>0.1218275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-even"><td><p>H</p></td>
<td><p>0.05</p></td>
<td><p>0.021</p></td>
<td><p>0.0256275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-odd"><td><p>I</p></td>
<td><p>0.05</p></td>
<td><p>0.061</p></td>
<td><p>0.0656275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-even"><td><p>K</p></td>
<td><p>0.05</p></td>
<td><p>0.0958</p></td>
<td><p>0.1004275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-odd"><td><p>L</p></td>
<td><p>0.05</p></td>
<td><p>0.0838</p></td>
<td><p>0.0884275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-even"><td><p>M</p></td>
<td><p>0.05</p></td>
<td><p>0.0123</p></td>
<td><p>0.0</p></td>
<td><p>0.0</p></td>
</tr>
<tr class="row-odd"><td><p>N</p></td>
<td><p>0.05</p></td>
<td><p>0.0386</p></td>
<td><p>0.0432275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-even"><td><p>P</p></td>
<td><p>0.05</p></td>
<td><p>0.0463</p></td>
<td><p>0.0509275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-odd"><td><p>Q</p></td>
<td><p>0.05</p></td>
<td><p>0.0251</p></td>
<td><p>0.0297275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-even"><td><p>R</p></td>
<td><p>0.05</p></td>
<td><p>0.0545</p></td>
<td><p>0.0591275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-odd"><td><p>S</p></td>
<td><p>0.05</p></td>
<td><p>0.0613</p></td>
<td><p>0.0659275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-even"><td><p>T</p></td>
<td><p>0.05</p></td>
<td><p>0.0455</p></td>
<td><p>0.0501275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-odd"><td><p>V</p></td>
<td><p>0.05</p></td>
<td><p>0.0572</p></td>
<td><p>0.0618275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-even"><td><p>W</p></td>
<td><p>0.05</p></td>
<td><p>0.0155</p></td>
<td><p>0.0201275</p></td>
<td><p>0.05555555</p></td>
</tr>
<tr class="row-odd"><td><p>Y</p></td>
<td><p>0.05</p></td>
<td><p>0.0244</p></td>
<td><p>0.0290275</p></td>
<td><p>0.05555555</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.clean">
<code class="sig-name descname">clean</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseSequence.clean"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.clean" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to clean / clear / empty the attributes <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> and <code class="xref py py-attr docutils literal notranslate"><span class="pre">names</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>freshly initialized, empty class attributes.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.filter_aa">
<code class="sig-name descname">filter_aa</code><span class="sig-paren">(</span><em class="sig-param">amino_acids</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseSequence.filter_aa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.filter_aa" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to filter out corresponding names and descriptor values of sequences with given amino acids in the
argument list <em>aminoacids</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>amino_acids</strong> – {list} amino acids to be filtered</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>filtered list of sequences names in the corresponding attributes.</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">BaseSequence</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AAALLLIIIKKK&#39;</span><span class="p">,</span> <span class="s1">&#39;CCEERRT&#39;</span><span class="p">,</span> <span class="s1">&#39;LLVVIIFFFQQ&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">filter_aa</span><span class="p">([</span><span class="s1">&#39;C&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;AAALLLIIIKKK&#39;, &#39;LLVVIIFFFQQ&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.filter_duplicates">
<code class="sig-name descname">filter_duplicates</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseSequence.filter_duplicates"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.filter_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to filter duplicates in the sequences from the class attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>filtered sequences list in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> and corresponding names.</p>
</dd>
<dt class="field-even">Example</dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">BaseSequence</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KLLKLLKKLLKLLK&#39;</span><span class="p">,</span> <span class="s1">&#39;KLLKLLKKLLKLLK&#39;</span><span class="p">,</span> <span class="s1">&#39;KLAKLAKKLAKLAK&#39;</span><span class="p">,</span> <span class="s1">&#39;KLAKLAKKLAKLAK&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">filter_duplicates</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;KLLKLLKKLLKLLK&#39;, &#39;KLAKLAKKLAKLAK&#39;]</span>
</pre></div>
</div>
<div class="versionadded">
<p><span class="versionmodified added">New in version v2.2.5.</span></p>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.keep_natural_aa">
<code class="sig-name descname">keep_natural_aa</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseSequence.keep_natural_aa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.keep_natural_aa" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to filter out sequences that do not contain natural amino acids. If the sequence contains a character
that is not in <code class="docutils literal notranslate"><span class="pre">['A','C','D,'E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y']</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>filtered sequence list in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>. The other attributes are also filtered
accordingly (if present).</p>
</dd>
<dt class="field-even">Example</dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">BaseSequence</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;BBBsdflUasUJfBJ&#39;</span><span class="p">,</span> <span class="s1">&#39;GLFDIVKKVVGALGSL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">keep_natural_aa</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;GLFDIVKKVVGALGSL&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.mutate_AA">
<code class="sig-name descname">mutate_AA</code><span class="sig-paren">(</span><em class="sig-param">nr</em>, <em class="sig-param">prob</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseSequence.mutate_AA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.mutate_AA" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to mutate with <strong>prob</strong> probability a <strong>nr</strong> of positions per sequence randomly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nr</strong> – number of mutations to perform per sequence</p></li>
<li><p><strong>prob</strong> – probability of mutating a sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>mutated sequences in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code>.</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">BaseSequence</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IAKAGRAIIK&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">mutate_AA</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mf">1.</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;NAKAGRAWIK&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.save_fasta">
<code class="sig-name descname">save_fasta</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">names=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#BaseSequence.save_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.BaseSequence.save_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to save generated sequences in a <code class="docutils literal notranslate"><span class="pre">.FASTA</span></code> formatted file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> – output filename in which the sequences from <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> are safed in fasta format.</p></li>
<li><p><strong>names</strong> – {bool} whether sequence names from <code class="xref py py-attr docutils literal notranslate"><span class="pre">names</span></code> should be saved as sequence identifiers</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a FASTA formatted file containing the generated sequences</p>
</dd>
<dt class="field-odd">Example</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">BaseSequence</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KLLSLSLALDLLS&#39;</span><span class="p">,</span> <span class="s1">&#39;KLPERTVVNSSDF&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Sequence1&#39;</span><span class="p">,</span> <span class="s1">&#39;Sequence2&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">save_fasta</span><span class="p">(</span><span class="s1">&#39;/location/of/fasta/file.fasta&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.aa_energies">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.modlamp_core.</code><code class="sig-name descname">aa_energies</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#aa_energies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.aa_energies" title="Permalink to this definition">¶</a></dt>
<dd><p>Function holding free energies of transfer between cyclohexane and water for all natural amino acids.
H. G. Boman, D. Wade, I. a Boman, B. Wåhlin, R. B. Merrifield, <em>FEBS Lett</em>. <strong>1989</strong>, <em>259</em>, 103–106.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dictionary with amino acid letters and corresponding energies.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.aa_formulas">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.modlamp_core.</code><code class="sig-name descname">aa_formulas</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#aa_formulas"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.aa_formulas" title="Permalink to this definition">¶</a></dt>
<dd><p>Function returning the molecular formulas of all amino acids. All amino acids are considered in the neutral form
(uncharged).</p>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.aa_weights">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.modlamp_core.</code><code class="sig-name descname">aa_weights</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#aa_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.aa_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Function holding molecular weight data on all natural amino acids.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dictionary with amino acid letters and corresponding weights</p>
</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified added">New in version v2.4.1.</span></p>
</div>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.count_aas">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.modlamp_core.</code><code class="sig-name descname">count_aas</code><span class="sig-paren">(</span><em class="sig-param">seq</em>, <em class="sig-param">scale='relative'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#count_aas"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.count_aas" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to count the amino acids occuring in a given sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – {str} amino acid sequence</p></li>
<li><p><strong>scale</strong> – {‘absolute’ or ‘relative’} defines whether counts or frequencies are given for each AA</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>{dict} dictionary with amino acids as keys and their counts in the sequence as values.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.load_scale">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.modlamp_core.</code><code class="sig-name descname">load_scale</code><span class="sig-paren">(</span><em class="sig-param">scalename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#load_scale"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.load_scale" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to load scale values for a given amino acid scale</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>scalename</strong> – amino acid scale name, for available scales see the
<code class="xref py py-class docutils literal notranslate"><span class="pre">modlamp.pydpi_py3.PeptideDescriptor()</span></code> documentation.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>amino acid scale values in dictionary format.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.ngrams_apd">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.modlamp_core.</code><code class="sig-name descname">ngrams_apd</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#ngrams_apd"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.ngrams_apd" title="Permalink to this definition">¶</a></dt>
<dd><p>Function returning the most frequent 2-, 3- and 4-grams from all sequences in the <a class="reference external" href="http://aps.unmc.edu/AP/">APD3</a>, version August 2016 with 2727 sequences.
For all 2, 3 and 4grams, all possible ngrams were generated from all sequences and the top 50 most frequent
assembled into a list. Finally, leading and tailing spaces were striped and duplicates as well as ngrams containing
spaces were removed.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>numpy.array containing most frequent ngrams</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.read_fasta">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.modlamp_core.</code><code class="sig-name descname">read_fasta</code><span class="sig-paren">(</span><em class="sig-param">inputfile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#read_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.read_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for loading sequences from a FASTA formatted file into <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> &amp; <code class="xref py py-attr docutils literal notranslate"><span class="pre">names</span></code>.
This method is used by the base class <code class="xref py py-class docutils literal notranslate"><span class="pre">modlamp.pydpi_py3.PeptideDescriptor</span></code> if the input is a FASTA file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>inputfile</strong> – .fasta file with sequences and headers to read</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of sequences in the attribute <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> with corresponding sequence names in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">names</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.modlamp_core.save_fasta">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.modlamp_core.</code><code class="sig-name descname">save_fasta</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">sequences</em>, <em class="sig-param">names=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/modlamp_core.html#save_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.modlamp_core.save_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Method for saving sequences in the instance <code class="xref py py-attr docutils literal notranslate"><span class="pre">sequences</span></code> to a file in FASTA format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> – {str} output filename (ending .fasta)</p></li>
<li><p><strong>sequences</strong> – {list} sequences to be saved to file</p></li>
<li><p><strong>names</strong> – {list} whether sequence names from self.names should be saved as sequence identifiers</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a FASTA formatted file containing the generated sequences</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.pcp">
<span id="pcp"></span><h2>pcp<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.pcp" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.pseudo_aac">
<span id="pseudo-aac"></span><h2>pseudo_aac<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.pseudo_aac" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>Instead of using the conventional 20-D amino acid composition to represent the sample</p>
<p>of a protein, Prof. Kuo-Chen Chou proposed the pseudo amino acid (PseAA) composition</p>
<p>in order for inluding the sequence-order information. Based on the concept of Chou’s</p>
<p>pseudo amino acid composition, the server PseAA was designed in a flexible way, allowing</p>
<p>users to generate various kinds of pseudo amino acid composition for a given protein</p>
<p>sequence by selecting different parameters and their combinations. This module aims at</p>
<p>computing two types of PseAA pydpi_py3: Type I and Type II.</p>
<p>You can freely use and distribute it. If you have any problem, you could contact</p>
<p>with us timely.</p>
<p>References:</p>
<p>[1]: Kuo-Chen Chou. Prediction of Protein Cellular Attributes Using Pseudo-Amino Acid</p>
<p>Composition. PROTEINS: Structure, Function, and Genetics, 2001, 43: 246-255.</p>
<p>[2]: <a class="reference external" href="http://www.csbio.sjtu.edu.cn/bioinf/PseAAC/">http://www.csbio.sjtu.edu.cn/bioinf/PseAAC/</a></p>
<p>[3]: <a class="reference external" href="http://www.csbio.sjtu.edu.cn/bioinf/PseAAC/type2.htm">http://www.csbio.sjtu.edu.cn/bioinf/PseAAC/type2.htm</a></p>
<p>[4]: Kuo-Chen Chou. Using amphiphilic pseudo amino acid composition to predict enzyme</p>
<p>subfamily classes. Bioinformatics, 2005,21,10-19.</p>
<p>Authors: Dongsheng Cao and Yizeng Liang.</p>
<p>Date: 2012.9.2</p>
<p>Email: <a class="reference external" href="mailto:oriental-cds&#37;&#52;&#48;163&#46;com">oriental-cds<span>&#64;</span>163<span>&#46;</span>com</a></p>
<p>The hydrophobicity values are from JACS, 1962, 84: 4240-4246. (C. Tanford).</p>
<p>The hydrophilicity values are from PNAS, 1981, 78:3824-3828 (T.P.Hopp &amp; K.R.Woods).</p>
<p>The side-chain mass for each of the 20 amino acids.</p>
<p>CRC Handbook of Chemistry and Physics, 66th ed., CRC Press, Boca Raton, Florida (1985).</p>
<p>R.M.C. Dawson, D.C. Elliott, W.H. Elliott, K.M. Jones, Data for Biochemical Research 3rd ed.,</p>
<p>Clarendon Press Oxford (1986).</p>
<p>Altered and converted to python 3.6 by Ana Marta Sequeira 05/2019</p>
<hr class="docutils" />
<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.get_a_pseudo_aac">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">get_a_pseudo_aac</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">lamda=30</em>, <em class="sig-param">weight=0.5</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#get_a_pseudo_aac"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.get_a_pseudo_aac" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing all of type II pseudo-amino acid compostion based on the given
properties. Note that the number of PAAC strongly depends on the lamda value. if lamda = 20,
we can obtain 20+20=40 PAAC. The size of these values depends on the choice of lamda and weight simultaneously.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>lamda</strong> – lamda factor reflects the rank of correlation and is a non-Negative integer, such as 15.
Note that (1)lamda should NOT be larger than the length of input protein sequence;
(2) lamda must be non-Negative integer, such as 0, 1, 2, …; (3) when lamda =0, the
output of PseAA server is the 20-D amino acid composition.</p></li>
<li><p><strong>weight</strong> – weight factor is designed for the users to put weight on the additional PseAA components
with respect to the conventional AA components. The user can select any value within the
region from 0.05 to 0.7 for the weight factor.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing calculated 20+lamda PAAC</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.get_a_pseudo_aac1">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">get_a_pseudo_aac1</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">lamda=30</em>, <em class="sig-param">weight=0.5</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#get_a_pseudo_aac1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.get_a_pseudo_aac1" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the first 20 of type II pseudo-amino acid compostion based on     [_Hydrophobicity,_hydrophilicity].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – </p></li>
<li><p><strong>lamda</strong> – </p></li>
<li><p><strong>weight</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.get_a_pseudo_aac2">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">get_a_pseudo_aac2</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">lamda=30</em>, <em class="sig-param">weight=0.5</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#get_a_pseudo_aac2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.get_a_pseudo_aac2" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the last lamda of type II pseudo-amino acid compostion based on       [_Hydrophobicity,_hydrophilicity].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – </p></li>
<li><p><strong>lamda</strong> – </p></li>
<li><p><strong>weight</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.get_aa_composition">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">get_aa_composition</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#get_aa_composition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.get_aa_composition" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the composition of Amino acids for a given protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing the composition of 20 amino acids.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.get_correlation_function">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">get_correlation_function</code><span class="sig-paren">(</span><em class="sig-param">Ri='S'</em>, <em class="sig-param">Rj='D'</em>, <em class="sig-param">AAP=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#get_correlation_function"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.get_correlation_function" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the correlation between two given amino acids using the given properties.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>Ri</strong> – amino acids</p></li>
<li><p><strong>Rj</strong> – amino acids</p></li>
<li><p><strong>AAP</strong> – list form containing the properties, each of which is a dict form</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>correlation value between two amino acids</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.get_pseudo_aac">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">get_pseudo_aac</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">lamda=30</em>, <em class="sig-param">weight=0.05</em>, <em class="sig-param">AAP=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#get_pseudo_aac"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.get_pseudo_aac" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing all of type I pseudo-amino acid compostion based on the given
properties. Note that the number of PAAC strongly depends on the lamda value. if lamda = 20,
we can obtain 20+20=40 PAAC. The size of these values depends on the choice of lamda and weight simultaneously.
You must specify some properties into AAP.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>lamda</strong> – lamda factor reflects the rank of correlation and is a non-Negative integer, such as 15.
Note that (1)lamda should NOT be larger than the length of input protein sequence;
(2) lamda must be non-Negative integer, such as 0, 1, 2, …; (3) when lamda =0, the
output of PseAA server is the 20-D amino acid composition.</p></li>
<li><p><strong>weight</strong> – weight factor is designed for the users to put weight on the additional PseAA components
with respect to the conventional AA components. The user can select any value within the
region from 0.05 to 0.7 for the weight factor.</p></li>
<li><p><strong>AAP</strong> – AAP is a list form containing the properties, each of which is a dict form</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing calculated 20+lamda PAAC</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.get_pseudo_aac1">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">get_pseudo_aac1</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">lamda=30</em>, <em class="sig-param">weight=0.05</em>, <em class="sig-param">AAP=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#get_pseudo_aac1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.get_pseudo_aac1" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the first 20 of type I pseudo-amino acid compostion based on the given properties.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – </p></li>
<li><p><strong>lamda</strong> – </p></li>
<li><p><strong>weight</strong> – </p></li>
<li><p><strong>AAP</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.get_pseudo_aac2">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">get_pseudo_aac2</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">lamda=30</em>, <em class="sig-param">weight=0.05</em>, <em class="sig-param">AAP=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#get_pseudo_aac2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.get_pseudo_aac2" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the last lamda of type I pseudo-amino acid compostion based on the given properties.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – </p></li>
<li><p><strong>lamda</strong> – </p></li>
<li><p><strong>weight</strong> – </p></li>
<li><p><strong>AAP</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.get_sequence_order_correlation_factor">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">get_sequence_order_correlation_factor</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">k=1</em>, <em class="sig-param">AAP=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#get_sequence_order_correlation_factor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.get_sequence_order_correlation_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the Sequence order correlation factor with gap equal to k based on the given properities.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>k</strong> – gap</p></li>
<li><p><strong>AAP</strong> – AAP is a list form containing the properties, each of which is a dict form</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>correlation factor value with the gap equal to k</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.get_sequence_order_correlation_factor_for_apaac">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">get_sequence_order_correlation_factor_for_apaac</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">k=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#get_sequence_order_correlation_factor_for_apaac"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.get_sequence_order_correlation_factor_for_apaac" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the Sequence order correlation factor with gap equal to k based on [_Hydrophobicity,_hydrophilicity]
for APAAC (type II PseAAC) .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>k</strong> – gap</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>correlation factor value with the gap equal to k</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.pseudo_aac.normalize_each_aap">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.pseudo_aac.</code><code class="sig-name descname">normalize_each_aap</code><span class="sig-paren">(</span><em class="sig-param">aap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/pseudo_aac.html#normalize_each_aap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.pseudo_aac.normalize_each_aap" title="Permalink to this definition">¶</a></dt>
<dd><p>All of the amino acid indices are centralized and standardized before the calculation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>aap</strong> – dict form containing the properties of 20 amino acids</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing the normalized properties of 20 amino acids.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions.quasi_sequence_order">
<span id="quasi-sequence-order"></span><h2>quasi_sequence_order<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions.quasi_sequence_order" title="Permalink to this headline">¶</a></h2>
<p>given protein sequence. We can obtain two types of pydpi_py3: Sequence-order-coupling</p>
<p>number and quasi-sequence-order pydpi_py3. Two distance matrixes between 20 amino acids</p>
<p>are employed. You can freely use and distribute it. If you have any problem, please contact</p>
<p>us immediately.</p>
<p>References:</p>
<p>[1]:Kuo-Chen Chou. Prediction of Protein Subcellar Locations by Incorporating</p>
<p>Quasi-Sequence-Order Effect. Biochemical and Biophysical Research Communications</p>
<p>2000, 278, 477-483.</p>
<p>[2]: Kuo-Chen Chou and Yu-Dong Cai. Prediction of Protein sucellular locations by</p>
<p>GO-FunD-PseAA predictor, Biochemical and Biophysical Research Communications,</p>
<p>2004, 320, 1236-1239.</p>
<p>[3]:Gisbert Schneider and Paul wrede. The Rational Design of Amino Acid</p>
<p>Sequences by Artifical Neural Networks and Simulated Molecular Evolution: Do</p>
<p>Novo Design of an Idealized Leader Cleavge Site. Biophys Journal, 1994, 66,</p>
<p>335-344.</p>
<p>Authors: Dongsheng Cao and Yizeng Liang.</p>
<p>Date: 2012.09.03</p>
<p>Email: <a class="reference external" href="mailto:oriental-cds&#37;&#52;&#48;163&#46;com">oriental-cds<span>&#64;</span>163<span>&#46;</span>com</a></p>
<p>Altered and converted to python 3.6 by Ana Marta Sequeira 05/2019</p>
<hr class="docutils" />
<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_aa_composition">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_aa_composition</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_aa_composition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_aa_composition" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the composition of Amino acids for a given protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing the composition of 20 amino acids.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_quasi_sequence_order</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">weight=0.1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_quasi_sequence_order"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing quasi-sequence-order for a given protein.
[1]:Kuo-Chen Chou. Prediction of Protein Subcellar Locations by
Incorporating Quasi-Sequence-Order Effect. Biochemical and Biophysical
Research Communications 2000, 278, 477-483.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>maxlag</strong> – maxlag is the maximum lag and the length of the protein should be larger than maxlag. default is 30.</p></li>
<li><p><strong>weight</strong> – weight is a weight factor.  please see reference 1 for its choice. default is 0.1.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing all quasi-sequence-order</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order1">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_quasi_sequence_order1</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">weight=0.1</em>, <em class="sig-param">distancematrix={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_quasi_sequence_order1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order1" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the first 20 quasi-sequence-order for a given protein sequence.
see method get_quasi_sequence_order for the choice of parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – </p></li>
<li><p><strong>maxlag</strong> – </p></li>
<li><p><strong>weight</strong> – </p></li>
<li><p><strong>distancematrix</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order1_grant">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_quasi_sequence_order1_grant</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">weight=0.1</em>, <em class="sig-param">distancematrix={'AA': 0.0</em>, <em class="sig-param">'AC': 0.112</em>, <em class="sig-param">'AD': 0.819</em>, <em class="sig-param">'AE': 0.827</em>, <em class="sig-param">'AF': 0.54</em>, <em class="sig-param">'AG': 0.208</em>, <em class="sig-param">'AH': 0.696</em>, <em class="sig-param">'AI': 0.407</em>, <em class="sig-param">'AK': 0.891</em>, <em class="sig-param">'AL': 0.406</em>, <em class="sig-param">'AM': 0.379</em>, <em class="sig-param">'AN': 0.318</em>, <em class="sig-param">'AP': 0.191</em>, <em class="sig-param">'AQ': 0.372</em>, <em class="sig-param">'AR': 1.0</em>, <em class="sig-param">'AS': 0.094</em>, <em class="sig-param">'AT': 0.22</em>, <em class="sig-param">'AV': 0.273</em>, <em class="sig-param">'AW': 0.739</em>, <em class="sig-param">'AY': 0.552</em>, <em class="sig-param">'CA': 0.114</em>, <em class="sig-param">'CC': 0.0</em>, <em class="sig-param">'CD': 0.847</em>, <em class="sig-param">'CE': 0.838</em>, <em class="sig-param">'CF': 0.437</em>, <em class="sig-param">'CG': 0.32</em>, <em class="sig-param">'CH': 0.66</em>, <em class="sig-param">'CI': 0.304</em>, <em class="sig-param">'CK': 0.887</em>, <em class="sig-param">'CL': 0.301</em>, <em class="sig-param">'CM': 0.277</em>, <em class="sig-param">'CN': 0.324</em>, <em class="sig-param">'CP': 0.157</em>, <em class="sig-param">'CQ': 0.341</em>, <em class="sig-param">'CR': 1.0</em>, <em class="sig-param">'CS': 0.176</em>, <em class="sig-param">'CT': 0.233</em>, <em class="sig-param">'CV': 0.167</em>, <em class="sig-param">'CW': 0.639</em>, <em class="sig-param">'CY': 0.457</em>, <em class="sig-param">'DA': 0.729</em>, <em class="sig-param">'DC': 0.742</em>, <em class="sig-param">'DD': 0.0</em>, <em class="sig-param">'DE': 0.124</em>, <em class="sig-param">'DF': 0.924</em>, <em class="sig-param">'DG': 0.697</em>, <em class="sig-param">'DH': 0.435</em>, <em class="sig-param">'DI': 0.847</em>, <em class="sig-param">'DK': 0.249</em>, <em class="sig-param">'DL': 0.841</em>, <em class="sig-param">'DM': 0.819</em>, <em class="sig-param">'DN': 0.56</em>, <em class="sig-param">'DP': 0.657</em>, <em class="sig-param">'DQ': 0.584</em>, <em class="sig-param">'DR': 0.295</em>, <em class="sig-param">'DS': 0.667</em>, <em class="sig-param">'DT': 0.649</em>, <em class="sig-param">'DV': 0.797</em>, <em class="sig-param">'DW': 1.0</em>, <em class="sig-param">'DY': 0.836</em>, <em class="sig-param">'EA': 0.79</em>, <em class="sig-param">'EC': 0.788</em>, <em class="sig-param">'ED': 0.133</em>, <em class="sig-param">'EE': 0.0</em>, <em class="sig-param">'EF': 0.932</em>, <em class="sig-param">'EG': 0.779</em>, <em class="sig-param">'EH': 0.406</em>, <em class="sig-param">'EI': 0.86</em>, <em class="sig-param">'EK': 0.143</em>, <em class="sig-param">'EL': 0.854</em>, <em class="sig-param">'EM': 0.83</em>, <em class="sig-param">'EN': 0.599</em>, <em class="sig-param">'EP': 0.688</em>, <em class="sig-param">'EQ': 0.598</em>, <em class="sig-param">'ER': 0.234</em>, <em class="sig-param">'ES': 0.726</em>, <em class="sig-param">'ET': 0.682</em>, <em class="sig-param">'EV': 0.824</em>, <em class="sig-param">'EW': 1.0</em>, <em class="sig-param">'EY': 0.837</em>, <em class="sig-param">'FA': 0.508</em>, <em class="sig-param">'FC': 0.405</em>, <em class="sig-param">'FD': 0.977</em>, <em class="sig-param">'FE': 0.918</em>, <em class="sig-param">'FF': 0.0</em>, <em class="sig-param">'FG': 0.69</em>, <em class="sig-param">'FH': 0.663</em>, <em class="sig-param">'FI': 0.128</em>, <em class="sig-param">'FK': 0.903</em>, <em class="sig-param">'FL': 0.131</em>, <em class="sig-param">'FM': 0.169</em>, <em class="sig-param">'FN': 0.541</em>, <em class="sig-param">'FP': 0.42</em>, <em class="sig-param">'FQ': 0.459</em>, <em class="sig-param">'FR': 1.0</em>, <em class="sig-param">'FS': 0.548</em>, <em class="sig-param">'FT': 0.499</em>, <em class="sig-param">'FV': 0.252</em>, <em class="sig-param">'FW': 0.207</em>, <em class="sig-param">'FY': 0.179</em>, <em class="sig-param">'GA': 0.206</em>, <em class="sig-param">'GC': 0.312</em>, <em class="sig-param">'GD': 0.776</em>, <em class="sig-param">'GE': 0.807</em>, <em class="sig-param">'GF': 0.727</em>, <em class="sig-param">'GG': 0.0</em>, <em class="sig-param">'GH': 0.769</em>, <em class="sig-param">'GI': 0.592</em>, <em class="sig-param">'GK': 0.894</em>, <em class="sig-param">'GL': 0.591</em>, <em class="sig-param">'GM': 0.557</em>, <em class="sig-param">'GN': 0.381</em>, <em class="sig-param">'GP': 0.323</em>, <em class="sig-param">'GQ': 0.467</em>, <em class="sig-param">'GR': 1.0</em>, <em class="sig-param">'GS': 0.158</em>, <em class="sig-param">'GT': 0.272</em>, <em class="sig-param">'GV': 0.464</em>, <em class="sig-param">'GW': 0.923</em>, <em class="sig-param">'GY': 0.728</em>, <em class="sig-param">'HA': 0.896</em>, <em class="sig-param">'HC': 0.836</em>, <em class="sig-param">'HD': 0.629</em>, <em class="sig-param">'HE': 0.547</em>, <em class="sig-param">'HF': 0.907</em>, <em class="sig-param">'HG': 1.0</em>, <em class="sig-param">'HH': 0.0</em>, <em class="sig-param">'HI': 0.848</em>, <em class="sig-param">'HK': 0.566</em>, <em class="sig-param">'HL': 0.842</em>, <em class="sig-param">'HM': 0.825</em>, <em class="sig-param">'HN': 0.754</em>, <em class="sig-param">'HP': 0.777</em>, <em class="sig-param">'HQ': 0.716</em>, <em class="sig-param">'HR': 0.697</em>, <em class="sig-param">'HS': 0.865</em>, <em class="sig-param">'HT': 0.834</em>, <em class="sig-param">'HV': 0.831</em>, <em class="sig-param">'HW': 0.981</em>, <em class="sig-param">'HY': 0.821</em>, <em class="sig-param">'IA': 0.403</em>, <em class="sig-param">'IC': 0.296</em>, <em class="sig-param">'ID': 0.942</em>, <em class="sig-param">'IE': 0.891</em>, <em class="sig-param">'IF': 0.134</em>, <em class="sig-param">'IG': 0.592</em>, <em class="sig-param">'IH': 0.652</em>, <em class="sig-param">'II': 0.0</em>, <em class="sig-param">'IK': 0.892</em>, <em class="sig-param">'IL': 0.013</em>, <em class="sig-param">'IM': 0.057</em>, <em class="sig-param">'IN': 0.457</em>, <em class="sig-param">'IP': 0.311</em>, <em class="sig-param">'IQ': 0.383</em>, <em class="sig-param">'IR': 1.0</em>, <em class="sig-param">'IS': 0.443</em>, <em class="sig-param">'IT': 0.396</em>, <em class="sig-param">'IV': 0.133</em>, <em class="sig-param">'IW': 0.339</em>, <em class="sig-param">'IY': 0.213</em>, <em class="sig-param">'KA': 0.889</em>, <em class="sig-param">'KC': 0.871</em>, <em class="sig-param">'KD': 0.279</em>, <em class="sig-param">'KE': 0.149</em>, <em class="sig-param">'KF': 0.957</em>, <em class="sig-param">'KG': 0.9</em>, <em class="sig-param">'KH': 0.438</em>, <em class="sig-param">'KI': 0.899</em>, <em class="sig-param">'KK': 0.0</em>, <em class="sig-param">'KL': 0.892</em>, <em class="sig-param">'KM': 0.871</em>, <em class="sig-param">'KN': 0.667</em>, <em class="sig-param">'KP': 0.757</em>, <em class="sig-param">'KQ': 0.639</em>, <em class="sig-param">'KR': 0.154</em>, <em class="sig-param">'KS': 0.825</em>, <em class="sig-param">'KT': 0.759</em>, <em class="sig-param">'KV': 0.882</em>, <em class="sig-param">'KW': 1.0</em>, <em class="sig-param">'KY': 0.848</em>, <em class="sig-param">'LA': 0.405</em>, <em class="sig-param">'LC': 0.296</em>, <em class="sig-param">'LD': 0.944</em>, <em class="sig-param">'LE': 0.892</em>, <em class="sig-param">'LF': 0.139</em>, <em class="sig-param">'LG': 0.596</em>, <em class="sig-param">'LH': 0.653</em>, <em class="sig-param">'LI': 0.013</em>, <em class="sig-param">'LK': 0.893</em>, <em class="sig-param">'LL': 0.0</em>, <em class="sig-param">'LM': 0.062</em>, <em class="sig-param">'LN': 0.452</em>, <em class="sig-param">'LP': 0.309</em>, <em class="sig-param">'LQ': 0.376</em>, <em class="sig-param">'LR': 1.0</em>, <em class="sig-param">'LS': 0.443</em>, <em class="sig-param">'LT': 0.397</em>, <em class="sig-param">'LV': 0.133</em>, <em class="sig-param">'LW': 0.341</em>, <em class="sig-param">'LY': 0.205</em>, <em class="sig-param">'MA': 0.383</em>, <em class="sig-param">'MC': 0.276</em>, <em class="sig-param">'MD': 0.932</em>, <em class="sig-param">'ME': 0.879</em>, <em class="sig-param">'MF': 0.182</em>, <em class="sig-param">'MG': 0.569</em>, <em class="sig-param">'MH': 0.648</em>, <em class="sig-param">'MI': 0.058</em>, <em class="sig-param">'MK': 0.884</em>, <em class="sig-param">'ML': 0.062</em>, <em class="sig-param">'MM': 0.0</em>, <em class="sig-param">'MN': 0.447</em>, <em class="sig-param">'MP': 0.285</em>, <em class="sig-param">'MQ': 0.372</em>, <em class="sig-param">'MR': 1.0</em>, <em class="sig-param">'MS': 0.417</em>, <em class="sig-param">'MT': 0.358</em>, <em class="sig-param">'MV': 0.12</em>, <em class="sig-param">'MW': 0.391</em>, <em class="sig-param">'MY': 0.255</em>, <em class="sig-param">'NA': 0.424</em>, <em class="sig-param">'NC': 0.425</em>, <em class="sig-param">'ND': 0.838</em>, <em class="sig-param">'NE': 0.835</em>, <em class="sig-param">'NF': 0.766</em>, <em class="sig-param">'NG': 0.512</em>, <em class="sig-param">'NH': 0.78</em>, <em class="sig-param">'NI': 0.615</em>, <em class="sig-param">'NK': 0.891</em>, <em class="sig-param">'NL': 0.603</em>, <em class="sig-param">'NM': 0.588</em>, <em class="sig-param">'NN': 0.0</em>, <em class="sig-param">'NP': 0.266</em>, <em class="sig-param">'NQ': 0.175</em>, <em class="sig-param">'NR': 1.0</em>, <em class="sig-param">'NS': 0.361</em>, <em class="sig-param">'NT': 0.368</em>, <em class="sig-param">'NV': 0.503</em>, <em class="sig-param">'NW': 0.945</em>, <em class="sig-param">'NY': 0.641</em>, <em class="sig-param">'PA': 0.22</em>, <em class="sig-param">'PC': 0.179</em>, <em class="sig-param">'PD': 0.852</em>, <em class="sig-param">'PE': 0.831</em>, <em class="sig-param">'PF': 0.515</em>, <em class="sig-param">'PG': 0.376</em>, <em class="sig-param">'PH': 0.696</em>, <em class="sig-param">'PI': 0.363</em>, <em class="sig-param">'PK': 0.875</em>, <em class="sig-param">'PL': 0.357</em>, <em class="sig-param">'PM': 0.326</em>, <em class="sig-param">'PN': 0.231</em>, <em class="sig-param">'PP': 0.0</em>, <em class="sig-param">'PQ': 0.228</em>, <em class="sig-param">'PR': 1.0</em>, <em class="sig-param">'PS': 0.196</em>, <em class="sig-param">'PT': 0.161</em>, <em class="sig-param">'PV': 0.244</em>, <em class="sig-param">'PW': 0.72</em>, <em class="sig-param">'PY': 0.481</em>, <em class="sig-param">'QA': 0.512</em>, <em class="sig-param">'QC': 0.462</em>, <em class="sig-param">'QD': 0.903</em>, <em class="sig-param">'QE': 0.861</em>, <em class="sig-param">'QF': 0.671</em>, <em class="sig-param">'QG': 0.648</em>, <em class="sig-param">'QH': 0.765</em>, <em class="sig-param">'QI': 0.532</em>, <em class="sig-param">'QK': 0.881</em>, <em class="sig-param">'QL': 0.518</em>, <em class="sig-param">'QM': 0.505</em>, <em class="sig-param">'QN': 0.181</em>, <em class="sig-param">'QP': 0.272</em>, <em class="sig-param">'QQ': 0.0</em>, <em class="sig-param">'QR': 1.0</em>, <em class="sig-param">'QS': 0.461</em>, <em class="sig-param">'QT': 0.389</em>, <em class="sig-param">'QV': 0.464</em>, <em class="sig-param">'QW': 0.831</em>, <em class="sig-param">'QY': 0.522</em>, <em class="sig-param">'RA': 0.919</em>, <em class="sig-param">'RC': 0.905</em>, <em class="sig-param">'RD': 0.305</em>, <em class="sig-param">'RE': 0.225</em>, <em class="sig-param">'RF': 0.977</em>, <em class="sig-param">'RG': 0.928</em>, <em class="sig-param">'RH': 0.498</em>, <em class="sig-param">'RI': 0.929</em>, <em class="sig-param">'RK': 0.141</em>, <em class="sig-param">'RL': 0.92</em>, <em class="sig-param">'RM': 0.908</em>, <em class="sig-param">'RN': 0.69</em>, <em class="sig-param">'RP': 0.796</em>, <em class="sig-param">'RQ': 0.668</em>, <em class="sig-param">'RR': 0.0</em>, <em class="sig-param">'RS': 0.86</em>, <em class="sig-param">'RT': 0.808</em>, <em class="sig-param">'RV': 0.914</em>, <em class="sig-param">'RW': 1.0</em>, <em class="sig-param">'RY': 0.859</em>, <em class="sig-param">'SA': 0.1</em>, <em class="sig-param">'SC': 0.185</em>, <em class="sig-param">'SD': 0.801</em>, <em class="sig-param">'SE': 0.812</em>, <em class="sig-param">'SF': 0.622</em>, <em class="sig-param">'SG': 0.17</em>, <em class="sig-param">'SH': 0.718</em>, <em class="sig-param">'SI': 0.478</em>, <em class="sig-param">'SK': 0.883</em>, <em class="sig-param">'SL': 0.474</em>, <em class="sig-param">'SM': 0.44</em>, <em class="sig-param">'SN': 0.289</em>, <em class="sig-param">'SP': 0.181</em>, <em class="sig-param">'SQ': 0.358</em>, <em class="sig-param">'SR': 1.0</em>, <em class="sig-param">'SS': 0.0</em>, <em class="sig-param">'ST': 0.174</em>, <em class="sig-param">'SV': 0.342</em>, <em class="sig-param">'SW': 0.827</em>, <em class="sig-param">'SY': 0.615</em>, <em class="sig-param">'TA': 0.251</em>, <em class="sig-param">'TC': 0.261</em>, <em class="sig-param">'TD': 0.83</em>, <em class="sig-param">'TE': 0.812</em>, <em class="sig-param">'TF': 0.604</em>, <em class="sig-param">'TG': 0.312</em>, <em class="sig-param">'TH': 0.737</em>, <em class="sig-param">'TI': 0.455</em>, <em class="sig-param">'TK': 0.866</em>, <em class="sig-param">'TL': 0.453</em>, <em class="sig-param">'TM': 0.403</em>, <em class="sig-param">'TN': 0.315</em>, <em class="sig-param">'TP': 0.159</em>, <em class="sig-param">'TQ': 0.322</em>, <em class="sig-param">'TR': 1.0</em>, <em class="sig-param">'TS': 0.185</em>, <em class="sig-param">'TT': 0.0</em>, <em class="sig-param">'TV': 0.345</em>, <em class="sig-param">'TW': 0.816</em>, <em class="sig-param">'TY': 0.596</em>, <em class="sig-param">'VA': 0.275</em>, <em class="sig-param">'VC': 0.165</em>, <em class="sig-param">'VD': 0.9</em>, <em class="sig-param">'VE': 0.867</em>, <em class="sig-param">'VF': 0.269</em>, <em class="sig-param">'VG': 0.471</em>, <em class="sig-param">'VH': 0.649</em>, <em class="sig-param">'VI': 0.135</em>, <em class="sig-param">'VK': 0.889</em>, <em class="sig-param">'VL': 0.134</em>, <em class="sig-param">'VM': 0.12</em>, <em class="sig-param">'VN': 0.38</em>, <em class="sig-param">'VP': 0.212</em>, <em class="sig-param">'VQ': 0.339</em>, <em class="sig-param">'VR': 1.0</em>, <em class="sig-param">'VS': 0.322</em>, <em class="sig-param">'VT': 0.305</em>, <em class="sig-param">'VV': 0.0</em>, <em class="sig-param">'VW': 0.472</em>, <em class="sig-param">'VY': 0.31</em>, <em class="sig-param">'WA': 0.658</em>, <em class="sig-param">'WC': 0.56</em>, <em class="sig-param">'WD': 1.0</em>, <em class="sig-param">'WE': 0.931</em>, <em class="sig-param">'WF': 0.196</em>, <em class="sig-param">'WG': 0.829</em>, <em class="sig-param">'WH': 0.678</em>, <em class="sig-param">'WI': 0.305</em>, <em class="sig-param">'WK': 0.892</em>, <em class="sig-param">'WL': 0.304</em>, <em class="sig-param">'WM': 0.344</em>, <em class="sig-param">'WN': 0.631</em>, <em class="sig-param">'WP': 0.555</em>, <em class="sig-param">'WQ': 0.538</em>, <em class="sig-param">'WR': 0.968</em>, <em class="sig-param">'WS': 0.689</em>, <em class="sig-param">'WT': 0.638</em>, <em class="sig-param">'WV': 0.418</em>, <em class="sig-param">'WW': 0.0</em>, <em class="sig-param">'WY': 0.204</em>, <em class="sig-param">'YA': 0.587</em>, <em class="sig-param">'YC': 0.478</em>, <em class="sig-param">'YD': 1.0</em>, <em class="sig-param">'YE': 0.932</em>, <em class="sig-param">'YF': 0.202</em>, <em class="sig-param">'YG': 0.782</em>, <em class="sig-param">'YH': 0.678</em>, <em class="sig-param">'YI': 0.23</em>, <em class="sig-param">'YK': 0.904</em>, <em class="sig-param">'YL': 0.219</em>, <em class="sig-param">'YM': 0.268</em>, <em class="sig-param">'YN': 0.512</em>, <em class="sig-param">'YP': 0.444</em>, <em class="sig-param">'YQ': 0.404</em>, <em class="sig-param">'YR': 0.995</em>, <em class="sig-param">'YS': 0.612</em>, <em class="sig-param">'YT': 0.557</em>, <em class="sig-param">'YV': 0.328</em>, <em class="sig-param">'YW': 0.244</em>, <em class="sig-param">'YY': 0.0}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_quasi_sequence_order1_grant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order1_grant" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the first 20 quasi-sequence-order for a given protein sequence.
see method get_quasi_sequence_order for the choice of parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – </p></li>
<li><p><strong>maxlag</strong> – </p></li>
<li><p><strong>weight</strong> – </p></li>
<li><p><strong>distancematrix</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order1_sw">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_quasi_sequence_order1_sw</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">weight=0.1</em>, <em class="sig-param">distancematrix={'AA': 0.0</em>, <em class="sig-param">'AC': 0.112</em>, <em class="sig-param">'AD': 0.819</em>, <em class="sig-param">'AE': 0.827</em>, <em class="sig-param">'AF': 0.54</em>, <em class="sig-param">'AG': 0.208</em>, <em class="sig-param">'AH': 0.696</em>, <em class="sig-param">'AI': 0.407</em>, <em class="sig-param">'AK': 0.891</em>, <em class="sig-param">'AL': 0.406</em>, <em class="sig-param">'AM': 0.379</em>, <em class="sig-param">'AN': 0.318</em>, <em class="sig-param">'AP': 0.191</em>, <em class="sig-param">'AQ': 0.372</em>, <em class="sig-param">'AR': 1.0</em>, <em class="sig-param">'AS': 0.094</em>, <em class="sig-param">'AT': 0.22</em>, <em class="sig-param">'AV': 0.273</em>, <em class="sig-param">'AW': 0.739</em>, <em class="sig-param">'AY': 0.552</em>, <em class="sig-param">'CA': 0.114</em>, <em class="sig-param">'CC': 0.0</em>, <em class="sig-param">'CD': 0.847</em>, <em class="sig-param">'CE': 0.838</em>, <em class="sig-param">'CF': 0.437</em>, <em class="sig-param">'CG': 0.32</em>, <em class="sig-param">'CH': 0.66</em>, <em class="sig-param">'CI': 0.304</em>, <em class="sig-param">'CK': 0.887</em>, <em class="sig-param">'CL': 0.301</em>, <em class="sig-param">'CM': 0.277</em>, <em class="sig-param">'CN': 0.324</em>, <em class="sig-param">'CP': 0.157</em>, <em class="sig-param">'CQ': 0.341</em>, <em class="sig-param">'CR': 1.0</em>, <em class="sig-param">'CS': 0.176</em>, <em class="sig-param">'CT': 0.233</em>, <em class="sig-param">'CV': 0.167</em>, <em class="sig-param">'CW': 0.639</em>, <em class="sig-param">'CY': 0.457</em>, <em class="sig-param">'DA': 0.729</em>, <em class="sig-param">'DC': 0.742</em>, <em class="sig-param">'DD': 0.0</em>, <em class="sig-param">'DE': 0.124</em>, <em class="sig-param">'DF': 0.924</em>, <em class="sig-param">'DG': 0.697</em>, <em class="sig-param">'DH': 0.435</em>, <em class="sig-param">'DI': 0.847</em>, <em class="sig-param">'DK': 0.249</em>, <em class="sig-param">'DL': 0.841</em>, <em class="sig-param">'DM': 0.819</em>, <em class="sig-param">'DN': 0.56</em>, <em class="sig-param">'DP': 0.657</em>, <em class="sig-param">'DQ': 0.584</em>, <em class="sig-param">'DR': 0.295</em>, <em class="sig-param">'DS': 0.667</em>, <em class="sig-param">'DT': 0.649</em>, <em class="sig-param">'DV': 0.797</em>, <em class="sig-param">'DW': 1.0</em>, <em class="sig-param">'DY': 0.836</em>, <em class="sig-param">'EA': 0.79</em>, <em class="sig-param">'EC': 0.788</em>, <em class="sig-param">'ED': 0.133</em>, <em class="sig-param">'EE': 0.0</em>, <em class="sig-param">'EF': 0.932</em>, <em class="sig-param">'EG': 0.779</em>, <em class="sig-param">'EH': 0.406</em>, <em class="sig-param">'EI': 0.86</em>, <em class="sig-param">'EK': 0.143</em>, <em class="sig-param">'EL': 0.854</em>, <em class="sig-param">'EM': 0.83</em>, <em class="sig-param">'EN': 0.599</em>, <em class="sig-param">'EP': 0.688</em>, <em class="sig-param">'EQ': 0.598</em>, <em class="sig-param">'ER': 0.234</em>, <em class="sig-param">'ES': 0.726</em>, <em class="sig-param">'ET': 0.682</em>, <em class="sig-param">'EV': 0.824</em>, <em class="sig-param">'EW': 1.0</em>, <em class="sig-param">'EY': 0.837</em>, <em class="sig-param">'FA': 0.508</em>, <em class="sig-param">'FC': 0.405</em>, <em class="sig-param">'FD': 0.977</em>, <em class="sig-param">'FE': 0.918</em>, <em class="sig-param">'FF': 0.0</em>, <em class="sig-param">'FG': 0.69</em>, <em class="sig-param">'FH': 0.663</em>, <em class="sig-param">'FI': 0.128</em>, <em class="sig-param">'FK': 0.903</em>, <em class="sig-param">'FL': 0.131</em>, <em class="sig-param">'FM': 0.169</em>, <em class="sig-param">'FN': 0.541</em>, <em class="sig-param">'FP': 0.42</em>, <em class="sig-param">'FQ': 0.459</em>, <em class="sig-param">'FR': 1.0</em>, <em class="sig-param">'FS': 0.548</em>, <em class="sig-param">'FT': 0.499</em>, <em class="sig-param">'FV': 0.252</em>, <em class="sig-param">'FW': 0.207</em>, <em class="sig-param">'FY': 0.179</em>, <em class="sig-param">'GA': 0.206</em>, <em class="sig-param">'GC': 0.312</em>, <em class="sig-param">'GD': 0.776</em>, <em class="sig-param">'GE': 0.807</em>, <em class="sig-param">'GF': 0.727</em>, <em class="sig-param">'GG': 0.0</em>, <em class="sig-param">'GH': 0.769</em>, <em class="sig-param">'GI': 0.592</em>, <em class="sig-param">'GK': 0.894</em>, <em class="sig-param">'GL': 0.591</em>, <em class="sig-param">'GM': 0.557</em>, <em class="sig-param">'GN': 0.381</em>, <em class="sig-param">'GP': 0.323</em>, <em class="sig-param">'GQ': 0.467</em>, <em class="sig-param">'GR': 1.0</em>, <em class="sig-param">'GS': 0.158</em>, <em class="sig-param">'GT': 0.272</em>, <em class="sig-param">'GV': 0.464</em>, <em class="sig-param">'GW': 0.923</em>, <em class="sig-param">'GY': 0.728</em>, <em class="sig-param">'HA': 0.896</em>, <em class="sig-param">'HC': 0.836</em>, <em class="sig-param">'HD': 0.629</em>, <em class="sig-param">'HE': 0.547</em>, <em class="sig-param">'HF': 0.907</em>, <em class="sig-param">'HG': 1.0</em>, <em class="sig-param">'HH': 0.0</em>, <em class="sig-param">'HI': 0.848</em>, <em class="sig-param">'HK': 0.566</em>, <em class="sig-param">'HL': 0.842</em>, <em class="sig-param">'HM': 0.825</em>, <em class="sig-param">'HN': 0.754</em>, <em class="sig-param">'HP': 0.777</em>, <em class="sig-param">'HQ': 0.716</em>, <em class="sig-param">'HR': 0.697</em>, <em class="sig-param">'HS': 0.865</em>, <em class="sig-param">'HT': 0.834</em>, <em class="sig-param">'HV': 0.831</em>, <em class="sig-param">'HW': 0.981</em>, <em class="sig-param">'HY': 0.821</em>, <em class="sig-param">'IA': 0.403</em>, <em class="sig-param">'IC': 0.296</em>, <em class="sig-param">'ID': 0.942</em>, <em class="sig-param">'IE': 0.891</em>, <em class="sig-param">'IF': 0.134</em>, <em class="sig-param">'IG': 0.592</em>, <em class="sig-param">'IH': 0.652</em>, <em class="sig-param">'II': 0.0</em>, <em class="sig-param">'IK': 0.892</em>, <em class="sig-param">'IL': 0.013</em>, <em class="sig-param">'IM': 0.057</em>, <em class="sig-param">'IN': 0.457</em>, <em class="sig-param">'IP': 0.311</em>, <em class="sig-param">'IQ': 0.383</em>, <em class="sig-param">'IR': 1.0</em>, <em class="sig-param">'IS': 0.443</em>, <em class="sig-param">'IT': 0.396</em>, <em class="sig-param">'IV': 0.133</em>, <em class="sig-param">'IW': 0.339</em>, <em class="sig-param">'IY': 0.213</em>, <em class="sig-param">'KA': 0.889</em>, <em class="sig-param">'KC': 0.871</em>, <em class="sig-param">'KD': 0.279</em>, <em class="sig-param">'KE': 0.149</em>, <em class="sig-param">'KF': 0.957</em>, <em class="sig-param">'KG': 0.9</em>, <em class="sig-param">'KH': 0.438</em>, <em class="sig-param">'KI': 0.899</em>, <em class="sig-param">'KK': 0.0</em>, <em class="sig-param">'KL': 0.892</em>, <em class="sig-param">'KM': 0.871</em>, <em class="sig-param">'KN': 0.667</em>, <em class="sig-param">'KP': 0.757</em>, <em class="sig-param">'KQ': 0.639</em>, <em class="sig-param">'KR': 0.154</em>, <em class="sig-param">'KS': 0.825</em>, <em class="sig-param">'KT': 0.759</em>, <em class="sig-param">'KV': 0.882</em>, <em class="sig-param">'KW': 1.0</em>, <em class="sig-param">'KY': 0.848</em>, <em class="sig-param">'LA': 0.405</em>, <em class="sig-param">'LC': 0.296</em>, <em class="sig-param">'LD': 0.944</em>, <em class="sig-param">'LE': 0.892</em>, <em class="sig-param">'LF': 0.139</em>, <em class="sig-param">'LG': 0.596</em>, <em class="sig-param">'LH': 0.653</em>, <em class="sig-param">'LI': 0.013</em>, <em class="sig-param">'LK': 0.893</em>, <em class="sig-param">'LL': 0.0</em>, <em class="sig-param">'LM': 0.062</em>, <em class="sig-param">'LN': 0.452</em>, <em class="sig-param">'LP': 0.309</em>, <em class="sig-param">'LQ': 0.376</em>, <em class="sig-param">'LR': 1.0</em>, <em class="sig-param">'LS': 0.443</em>, <em class="sig-param">'LT': 0.397</em>, <em class="sig-param">'LV': 0.133</em>, <em class="sig-param">'LW': 0.341</em>, <em class="sig-param">'LY': 0.205</em>, <em class="sig-param">'MA': 0.383</em>, <em class="sig-param">'MC': 0.276</em>, <em class="sig-param">'MD': 0.932</em>, <em class="sig-param">'ME': 0.879</em>, <em class="sig-param">'MF': 0.182</em>, <em class="sig-param">'MG': 0.569</em>, <em class="sig-param">'MH': 0.648</em>, <em class="sig-param">'MI': 0.058</em>, <em class="sig-param">'MK': 0.884</em>, <em class="sig-param">'ML': 0.062</em>, <em class="sig-param">'MM': 0.0</em>, <em class="sig-param">'MN': 0.447</em>, <em class="sig-param">'MP': 0.285</em>, <em class="sig-param">'MQ': 0.372</em>, <em class="sig-param">'MR': 1.0</em>, <em class="sig-param">'MS': 0.417</em>, <em class="sig-param">'MT': 0.358</em>, <em class="sig-param">'MV': 0.12</em>, <em class="sig-param">'MW': 0.391</em>, <em class="sig-param">'MY': 0.255</em>, <em class="sig-param">'NA': 0.424</em>, <em class="sig-param">'NC': 0.425</em>, <em class="sig-param">'ND': 0.838</em>, <em class="sig-param">'NE': 0.835</em>, <em class="sig-param">'NF': 0.766</em>, <em class="sig-param">'NG': 0.512</em>, <em class="sig-param">'NH': 0.78</em>, <em class="sig-param">'NI': 0.615</em>, <em class="sig-param">'NK': 0.891</em>, <em class="sig-param">'NL': 0.603</em>, <em class="sig-param">'NM': 0.588</em>, <em class="sig-param">'NN': 0.0</em>, <em class="sig-param">'NP': 0.266</em>, <em class="sig-param">'NQ': 0.175</em>, <em class="sig-param">'NR': 1.0</em>, <em class="sig-param">'NS': 0.361</em>, <em class="sig-param">'NT': 0.368</em>, <em class="sig-param">'NV': 0.503</em>, <em class="sig-param">'NW': 0.945</em>, <em class="sig-param">'NY': 0.641</em>, <em class="sig-param">'PA': 0.22</em>, <em class="sig-param">'PC': 0.179</em>, <em class="sig-param">'PD': 0.852</em>, <em class="sig-param">'PE': 0.831</em>, <em class="sig-param">'PF': 0.515</em>, <em class="sig-param">'PG': 0.376</em>, <em class="sig-param">'PH': 0.696</em>, <em class="sig-param">'PI': 0.363</em>, <em class="sig-param">'PK': 0.875</em>, <em class="sig-param">'PL': 0.357</em>, <em class="sig-param">'PM': 0.326</em>, <em class="sig-param">'PN': 0.231</em>, <em class="sig-param">'PP': 0.0</em>, <em class="sig-param">'PQ': 0.228</em>, <em class="sig-param">'PR': 1.0</em>, <em class="sig-param">'PS': 0.196</em>, <em class="sig-param">'PT': 0.161</em>, <em class="sig-param">'PV': 0.244</em>, <em class="sig-param">'PW': 0.72</em>, <em class="sig-param">'PY': 0.481</em>, <em class="sig-param">'QA': 0.512</em>, <em class="sig-param">'QC': 0.462</em>, <em class="sig-param">'QD': 0.903</em>, <em class="sig-param">'QE': 0.861</em>, <em class="sig-param">'QF': 0.671</em>, <em class="sig-param">'QG': 0.648</em>, <em class="sig-param">'QH': 0.765</em>, <em class="sig-param">'QI': 0.532</em>, <em class="sig-param">'QK': 0.881</em>, <em class="sig-param">'QL': 0.518</em>, <em class="sig-param">'QM': 0.505</em>, <em class="sig-param">'QN': 0.181</em>, <em class="sig-param">'QP': 0.272</em>, <em class="sig-param">'QQ': 0.0</em>, <em class="sig-param">'QR': 1.0</em>, <em class="sig-param">'QS': 0.461</em>, <em class="sig-param">'QT': 0.389</em>, <em class="sig-param">'QV': 0.464</em>, <em class="sig-param">'QW': 0.831</em>, <em class="sig-param">'QY': 0.522</em>, <em class="sig-param">'RA': 0.919</em>, <em class="sig-param">'RC': 0.905</em>, <em class="sig-param">'RD': 0.305</em>, <em class="sig-param">'RE': 0.225</em>, <em class="sig-param">'RF': 0.977</em>, <em class="sig-param">'RG': 0.928</em>, <em class="sig-param">'RH': 0.498</em>, <em class="sig-param">'RI': 0.929</em>, <em class="sig-param">'RK': 0.141</em>, <em class="sig-param">'RL': 0.92</em>, <em class="sig-param">'RM': 0.908</em>, <em class="sig-param">'RN': 0.69</em>, <em class="sig-param">'RP': 0.796</em>, <em class="sig-param">'RQ': 0.668</em>, <em class="sig-param">'RR': 0.0</em>, <em class="sig-param">'RS': 0.86</em>, <em class="sig-param">'RT': 0.808</em>, <em class="sig-param">'RV': 0.914</em>, <em class="sig-param">'RW': 1.0</em>, <em class="sig-param">'RY': 0.859</em>, <em class="sig-param">'SA': 0.1</em>, <em class="sig-param">'SC': 0.185</em>, <em class="sig-param">'SD': 0.801</em>, <em class="sig-param">'SE': 0.812</em>, <em class="sig-param">'SF': 0.622</em>, <em class="sig-param">'SG': 0.17</em>, <em class="sig-param">'SH': 0.718</em>, <em class="sig-param">'SI': 0.478</em>, <em class="sig-param">'SK': 0.883</em>, <em class="sig-param">'SL': 0.474</em>, <em class="sig-param">'SM': 0.44</em>, <em class="sig-param">'SN': 0.289</em>, <em class="sig-param">'SP': 0.181</em>, <em class="sig-param">'SQ': 0.358</em>, <em class="sig-param">'SR': 1.0</em>, <em class="sig-param">'SS': 0.0</em>, <em class="sig-param">'ST': 0.174</em>, <em class="sig-param">'SV': 0.342</em>, <em class="sig-param">'SW': 0.827</em>, <em class="sig-param">'SY': 0.615</em>, <em class="sig-param">'TA': 0.251</em>, <em class="sig-param">'TC': 0.261</em>, <em class="sig-param">'TD': 0.83</em>, <em class="sig-param">'TE': 0.812</em>, <em class="sig-param">'TF': 0.604</em>, <em class="sig-param">'TG': 0.312</em>, <em class="sig-param">'TH': 0.737</em>, <em class="sig-param">'TI': 0.455</em>, <em class="sig-param">'TK': 0.866</em>, <em class="sig-param">'TL': 0.453</em>, <em class="sig-param">'TM': 0.403</em>, <em class="sig-param">'TN': 0.315</em>, <em class="sig-param">'TP': 0.159</em>, <em class="sig-param">'TQ': 0.322</em>, <em class="sig-param">'TR': 1.0</em>, <em class="sig-param">'TS': 0.185</em>, <em class="sig-param">'TT': 0.0</em>, <em class="sig-param">'TV': 0.345</em>, <em class="sig-param">'TW': 0.816</em>, <em class="sig-param">'TY': 0.596</em>, <em class="sig-param">'VA': 0.275</em>, <em class="sig-param">'VC': 0.165</em>, <em class="sig-param">'VD': 0.9</em>, <em class="sig-param">'VE': 0.867</em>, <em class="sig-param">'VF': 0.269</em>, <em class="sig-param">'VG': 0.471</em>, <em class="sig-param">'VH': 0.649</em>, <em class="sig-param">'VI': 0.135</em>, <em class="sig-param">'VK': 0.889</em>, <em class="sig-param">'VL': 0.134</em>, <em class="sig-param">'VM': 0.12</em>, <em class="sig-param">'VN': 0.38</em>, <em class="sig-param">'VP': 0.212</em>, <em class="sig-param">'VQ': 0.339</em>, <em class="sig-param">'VR': 1.0</em>, <em class="sig-param">'VS': 0.322</em>, <em class="sig-param">'VT': 0.305</em>, <em class="sig-param">'VV': 0.0</em>, <em class="sig-param">'VW': 0.472</em>, <em class="sig-param">'VY': 0.31</em>, <em class="sig-param">'WA': 0.658</em>, <em class="sig-param">'WC': 0.56</em>, <em class="sig-param">'WD': 1.0</em>, <em class="sig-param">'WE': 0.931</em>, <em class="sig-param">'WF': 0.196</em>, <em class="sig-param">'WG': 0.829</em>, <em class="sig-param">'WH': 0.678</em>, <em class="sig-param">'WI': 0.305</em>, <em class="sig-param">'WK': 0.892</em>, <em class="sig-param">'WL': 0.304</em>, <em class="sig-param">'WM': 0.344</em>, <em class="sig-param">'WN': 0.631</em>, <em class="sig-param">'WP': 0.555</em>, <em class="sig-param">'WQ': 0.538</em>, <em class="sig-param">'WR': 0.968</em>, <em class="sig-param">'WS': 0.689</em>, <em class="sig-param">'WT': 0.638</em>, <em class="sig-param">'WV': 0.418</em>, <em class="sig-param">'WW': 0.0</em>, <em class="sig-param">'WY': 0.204</em>, <em class="sig-param">'YA': 0.587</em>, <em class="sig-param">'YC': 0.478</em>, <em class="sig-param">'YD': 1.0</em>, <em class="sig-param">'YE': 0.932</em>, <em class="sig-param">'YF': 0.202</em>, <em class="sig-param">'YG': 0.782</em>, <em class="sig-param">'YH': 0.678</em>, <em class="sig-param">'YI': 0.23</em>, <em class="sig-param">'YK': 0.904</em>, <em class="sig-param">'YL': 0.219</em>, <em class="sig-param">'YM': 0.268</em>, <em class="sig-param">'YN': 0.512</em>, <em class="sig-param">'YP': 0.444</em>, <em class="sig-param">'YQ': 0.404</em>, <em class="sig-param">'YR': 0.995</em>, <em class="sig-param">'YS': 0.612</em>, <em class="sig-param">'YT': 0.557</em>, <em class="sig-param">'YV': 0.328</em>, <em class="sig-param">'YW': 0.244</em>, <em class="sig-param">'YY': 0.0}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_quasi_sequence_order1_sw"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order1_sw" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the first 20 quasi-sequence-order for a given protein sequence.
see method get_quasi_sequence_order for the choice of parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – </p></li>
<li><p><strong>maxlag</strong> – </p></li>
<li><p><strong>weight</strong> – </p></li>
<li><p><strong>distancematrix</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order2">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_quasi_sequence_order2</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">weight=0.1</em>, <em class="sig-param">distancematrix={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_quasi_sequence_order2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order2" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the last maxlag quasi-sequence-order for a given protein sequence.
see method get_quasi_sequence_order for the choice of parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – </p></li>
<li><p><strong>maxlag</strong> – </p></li>
<li><p><strong>weight</strong> – </p></li>
<li><p><strong>distancematrix</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order2_grant">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_quasi_sequence_order2_grant</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">weight=0.1</em>, <em class="sig-param">distancematrix={'AA': 0.0</em>, <em class="sig-param">'AC': 0.112</em>, <em class="sig-param">'AD': 0.819</em>, <em class="sig-param">'AE': 0.827</em>, <em class="sig-param">'AF': 0.54</em>, <em class="sig-param">'AG': 0.208</em>, <em class="sig-param">'AH': 0.696</em>, <em class="sig-param">'AI': 0.407</em>, <em class="sig-param">'AK': 0.891</em>, <em class="sig-param">'AL': 0.406</em>, <em class="sig-param">'AM': 0.379</em>, <em class="sig-param">'AN': 0.318</em>, <em class="sig-param">'AP': 0.191</em>, <em class="sig-param">'AQ': 0.372</em>, <em class="sig-param">'AR': 1.0</em>, <em class="sig-param">'AS': 0.094</em>, <em class="sig-param">'AT': 0.22</em>, <em class="sig-param">'AV': 0.273</em>, <em class="sig-param">'AW': 0.739</em>, <em class="sig-param">'AY': 0.552</em>, <em class="sig-param">'CA': 0.114</em>, <em class="sig-param">'CC': 0.0</em>, <em class="sig-param">'CD': 0.847</em>, <em class="sig-param">'CE': 0.838</em>, <em class="sig-param">'CF': 0.437</em>, <em class="sig-param">'CG': 0.32</em>, <em class="sig-param">'CH': 0.66</em>, <em class="sig-param">'CI': 0.304</em>, <em class="sig-param">'CK': 0.887</em>, <em class="sig-param">'CL': 0.301</em>, <em class="sig-param">'CM': 0.277</em>, <em class="sig-param">'CN': 0.324</em>, <em class="sig-param">'CP': 0.157</em>, <em class="sig-param">'CQ': 0.341</em>, <em class="sig-param">'CR': 1.0</em>, <em class="sig-param">'CS': 0.176</em>, <em class="sig-param">'CT': 0.233</em>, <em class="sig-param">'CV': 0.167</em>, <em class="sig-param">'CW': 0.639</em>, <em class="sig-param">'CY': 0.457</em>, <em class="sig-param">'DA': 0.729</em>, <em class="sig-param">'DC': 0.742</em>, <em class="sig-param">'DD': 0.0</em>, <em class="sig-param">'DE': 0.124</em>, <em class="sig-param">'DF': 0.924</em>, <em class="sig-param">'DG': 0.697</em>, <em class="sig-param">'DH': 0.435</em>, <em class="sig-param">'DI': 0.847</em>, <em class="sig-param">'DK': 0.249</em>, <em class="sig-param">'DL': 0.841</em>, <em class="sig-param">'DM': 0.819</em>, <em class="sig-param">'DN': 0.56</em>, <em class="sig-param">'DP': 0.657</em>, <em class="sig-param">'DQ': 0.584</em>, <em class="sig-param">'DR': 0.295</em>, <em class="sig-param">'DS': 0.667</em>, <em class="sig-param">'DT': 0.649</em>, <em class="sig-param">'DV': 0.797</em>, <em class="sig-param">'DW': 1.0</em>, <em class="sig-param">'DY': 0.836</em>, <em class="sig-param">'EA': 0.79</em>, <em class="sig-param">'EC': 0.788</em>, <em class="sig-param">'ED': 0.133</em>, <em class="sig-param">'EE': 0.0</em>, <em class="sig-param">'EF': 0.932</em>, <em class="sig-param">'EG': 0.779</em>, <em class="sig-param">'EH': 0.406</em>, <em class="sig-param">'EI': 0.86</em>, <em class="sig-param">'EK': 0.143</em>, <em class="sig-param">'EL': 0.854</em>, <em class="sig-param">'EM': 0.83</em>, <em class="sig-param">'EN': 0.599</em>, <em class="sig-param">'EP': 0.688</em>, <em class="sig-param">'EQ': 0.598</em>, <em class="sig-param">'ER': 0.234</em>, <em class="sig-param">'ES': 0.726</em>, <em class="sig-param">'ET': 0.682</em>, <em class="sig-param">'EV': 0.824</em>, <em class="sig-param">'EW': 1.0</em>, <em class="sig-param">'EY': 0.837</em>, <em class="sig-param">'FA': 0.508</em>, <em class="sig-param">'FC': 0.405</em>, <em class="sig-param">'FD': 0.977</em>, <em class="sig-param">'FE': 0.918</em>, <em class="sig-param">'FF': 0.0</em>, <em class="sig-param">'FG': 0.69</em>, <em class="sig-param">'FH': 0.663</em>, <em class="sig-param">'FI': 0.128</em>, <em class="sig-param">'FK': 0.903</em>, <em class="sig-param">'FL': 0.131</em>, <em class="sig-param">'FM': 0.169</em>, <em class="sig-param">'FN': 0.541</em>, <em class="sig-param">'FP': 0.42</em>, <em class="sig-param">'FQ': 0.459</em>, <em class="sig-param">'FR': 1.0</em>, <em class="sig-param">'FS': 0.548</em>, <em class="sig-param">'FT': 0.499</em>, <em class="sig-param">'FV': 0.252</em>, <em class="sig-param">'FW': 0.207</em>, <em class="sig-param">'FY': 0.179</em>, <em class="sig-param">'GA': 0.206</em>, <em class="sig-param">'GC': 0.312</em>, <em class="sig-param">'GD': 0.776</em>, <em class="sig-param">'GE': 0.807</em>, <em class="sig-param">'GF': 0.727</em>, <em class="sig-param">'GG': 0.0</em>, <em class="sig-param">'GH': 0.769</em>, <em class="sig-param">'GI': 0.592</em>, <em class="sig-param">'GK': 0.894</em>, <em class="sig-param">'GL': 0.591</em>, <em class="sig-param">'GM': 0.557</em>, <em class="sig-param">'GN': 0.381</em>, <em class="sig-param">'GP': 0.323</em>, <em class="sig-param">'GQ': 0.467</em>, <em class="sig-param">'GR': 1.0</em>, <em class="sig-param">'GS': 0.158</em>, <em class="sig-param">'GT': 0.272</em>, <em class="sig-param">'GV': 0.464</em>, <em class="sig-param">'GW': 0.923</em>, <em class="sig-param">'GY': 0.728</em>, <em class="sig-param">'HA': 0.896</em>, <em class="sig-param">'HC': 0.836</em>, <em class="sig-param">'HD': 0.629</em>, <em class="sig-param">'HE': 0.547</em>, <em class="sig-param">'HF': 0.907</em>, <em class="sig-param">'HG': 1.0</em>, <em class="sig-param">'HH': 0.0</em>, <em class="sig-param">'HI': 0.848</em>, <em class="sig-param">'HK': 0.566</em>, <em class="sig-param">'HL': 0.842</em>, <em class="sig-param">'HM': 0.825</em>, <em class="sig-param">'HN': 0.754</em>, <em class="sig-param">'HP': 0.777</em>, <em class="sig-param">'HQ': 0.716</em>, <em class="sig-param">'HR': 0.697</em>, <em class="sig-param">'HS': 0.865</em>, <em class="sig-param">'HT': 0.834</em>, <em class="sig-param">'HV': 0.831</em>, <em class="sig-param">'HW': 0.981</em>, <em class="sig-param">'HY': 0.821</em>, <em class="sig-param">'IA': 0.403</em>, <em class="sig-param">'IC': 0.296</em>, <em class="sig-param">'ID': 0.942</em>, <em class="sig-param">'IE': 0.891</em>, <em class="sig-param">'IF': 0.134</em>, <em class="sig-param">'IG': 0.592</em>, <em class="sig-param">'IH': 0.652</em>, <em class="sig-param">'II': 0.0</em>, <em class="sig-param">'IK': 0.892</em>, <em class="sig-param">'IL': 0.013</em>, <em class="sig-param">'IM': 0.057</em>, <em class="sig-param">'IN': 0.457</em>, <em class="sig-param">'IP': 0.311</em>, <em class="sig-param">'IQ': 0.383</em>, <em class="sig-param">'IR': 1.0</em>, <em class="sig-param">'IS': 0.443</em>, <em class="sig-param">'IT': 0.396</em>, <em class="sig-param">'IV': 0.133</em>, <em class="sig-param">'IW': 0.339</em>, <em class="sig-param">'IY': 0.213</em>, <em class="sig-param">'KA': 0.889</em>, <em class="sig-param">'KC': 0.871</em>, <em class="sig-param">'KD': 0.279</em>, <em class="sig-param">'KE': 0.149</em>, <em class="sig-param">'KF': 0.957</em>, <em class="sig-param">'KG': 0.9</em>, <em class="sig-param">'KH': 0.438</em>, <em class="sig-param">'KI': 0.899</em>, <em class="sig-param">'KK': 0.0</em>, <em class="sig-param">'KL': 0.892</em>, <em class="sig-param">'KM': 0.871</em>, <em class="sig-param">'KN': 0.667</em>, <em class="sig-param">'KP': 0.757</em>, <em class="sig-param">'KQ': 0.639</em>, <em class="sig-param">'KR': 0.154</em>, <em class="sig-param">'KS': 0.825</em>, <em class="sig-param">'KT': 0.759</em>, <em class="sig-param">'KV': 0.882</em>, <em class="sig-param">'KW': 1.0</em>, <em class="sig-param">'KY': 0.848</em>, <em class="sig-param">'LA': 0.405</em>, <em class="sig-param">'LC': 0.296</em>, <em class="sig-param">'LD': 0.944</em>, <em class="sig-param">'LE': 0.892</em>, <em class="sig-param">'LF': 0.139</em>, <em class="sig-param">'LG': 0.596</em>, <em class="sig-param">'LH': 0.653</em>, <em class="sig-param">'LI': 0.013</em>, <em class="sig-param">'LK': 0.893</em>, <em class="sig-param">'LL': 0.0</em>, <em class="sig-param">'LM': 0.062</em>, <em class="sig-param">'LN': 0.452</em>, <em class="sig-param">'LP': 0.309</em>, <em class="sig-param">'LQ': 0.376</em>, <em class="sig-param">'LR': 1.0</em>, <em class="sig-param">'LS': 0.443</em>, <em class="sig-param">'LT': 0.397</em>, <em class="sig-param">'LV': 0.133</em>, <em class="sig-param">'LW': 0.341</em>, <em class="sig-param">'LY': 0.205</em>, <em class="sig-param">'MA': 0.383</em>, <em class="sig-param">'MC': 0.276</em>, <em class="sig-param">'MD': 0.932</em>, <em class="sig-param">'ME': 0.879</em>, <em class="sig-param">'MF': 0.182</em>, <em class="sig-param">'MG': 0.569</em>, <em class="sig-param">'MH': 0.648</em>, <em class="sig-param">'MI': 0.058</em>, <em class="sig-param">'MK': 0.884</em>, <em class="sig-param">'ML': 0.062</em>, <em class="sig-param">'MM': 0.0</em>, <em class="sig-param">'MN': 0.447</em>, <em class="sig-param">'MP': 0.285</em>, <em class="sig-param">'MQ': 0.372</em>, <em class="sig-param">'MR': 1.0</em>, <em class="sig-param">'MS': 0.417</em>, <em class="sig-param">'MT': 0.358</em>, <em class="sig-param">'MV': 0.12</em>, <em class="sig-param">'MW': 0.391</em>, <em class="sig-param">'MY': 0.255</em>, <em class="sig-param">'NA': 0.424</em>, <em class="sig-param">'NC': 0.425</em>, <em class="sig-param">'ND': 0.838</em>, <em class="sig-param">'NE': 0.835</em>, <em class="sig-param">'NF': 0.766</em>, <em class="sig-param">'NG': 0.512</em>, <em class="sig-param">'NH': 0.78</em>, <em class="sig-param">'NI': 0.615</em>, <em class="sig-param">'NK': 0.891</em>, <em class="sig-param">'NL': 0.603</em>, <em class="sig-param">'NM': 0.588</em>, <em class="sig-param">'NN': 0.0</em>, <em class="sig-param">'NP': 0.266</em>, <em class="sig-param">'NQ': 0.175</em>, <em class="sig-param">'NR': 1.0</em>, <em class="sig-param">'NS': 0.361</em>, <em class="sig-param">'NT': 0.368</em>, <em class="sig-param">'NV': 0.503</em>, <em class="sig-param">'NW': 0.945</em>, <em class="sig-param">'NY': 0.641</em>, <em class="sig-param">'PA': 0.22</em>, <em class="sig-param">'PC': 0.179</em>, <em class="sig-param">'PD': 0.852</em>, <em class="sig-param">'PE': 0.831</em>, <em class="sig-param">'PF': 0.515</em>, <em class="sig-param">'PG': 0.376</em>, <em class="sig-param">'PH': 0.696</em>, <em class="sig-param">'PI': 0.363</em>, <em class="sig-param">'PK': 0.875</em>, <em class="sig-param">'PL': 0.357</em>, <em class="sig-param">'PM': 0.326</em>, <em class="sig-param">'PN': 0.231</em>, <em class="sig-param">'PP': 0.0</em>, <em class="sig-param">'PQ': 0.228</em>, <em class="sig-param">'PR': 1.0</em>, <em class="sig-param">'PS': 0.196</em>, <em class="sig-param">'PT': 0.161</em>, <em class="sig-param">'PV': 0.244</em>, <em class="sig-param">'PW': 0.72</em>, <em class="sig-param">'PY': 0.481</em>, <em class="sig-param">'QA': 0.512</em>, <em class="sig-param">'QC': 0.462</em>, <em class="sig-param">'QD': 0.903</em>, <em class="sig-param">'QE': 0.861</em>, <em class="sig-param">'QF': 0.671</em>, <em class="sig-param">'QG': 0.648</em>, <em class="sig-param">'QH': 0.765</em>, <em class="sig-param">'QI': 0.532</em>, <em class="sig-param">'QK': 0.881</em>, <em class="sig-param">'QL': 0.518</em>, <em class="sig-param">'QM': 0.505</em>, <em class="sig-param">'QN': 0.181</em>, <em class="sig-param">'QP': 0.272</em>, <em class="sig-param">'QQ': 0.0</em>, <em class="sig-param">'QR': 1.0</em>, <em class="sig-param">'QS': 0.461</em>, <em class="sig-param">'QT': 0.389</em>, <em class="sig-param">'QV': 0.464</em>, <em class="sig-param">'QW': 0.831</em>, <em class="sig-param">'QY': 0.522</em>, <em class="sig-param">'RA': 0.919</em>, <em class="sig-param">'RC': 0.905</em>, <em class="sig-param">'RD': 0.305</em>, <em class="sig-param">'RE': 0.225</em>, <em class="sig-param">'RF': 0.977</em>, <em class="sig-param">'RG': 0.928</em>, <em class="sig-param">'RH': 0.498</em>, <em class="sig-param">'RI': 0.929</em>, <em class="sig-param">'RK': 0.141</em>, <em class="sig-param">'RL': 0.92</em>, <em class="sig-param">'RM': 0.908</em>, <em class="sig-param">'RN': 0.69</em>, <em class="sig-param">'RP': 0.796</em>, <em class="sig-param">'RQ': 0.668</em>, <em class="sig-param">'RR': 0.0</em>, <em class="sig-param">'RS': 0.86</em>, <em class="sig-param">'RT': 0.808</em>, <em class="sig-param">'RV': 0.914</em>, <em class="sig-param">'RW': 1.0</em>, <em class="sig-param">'RY': 0.859</em>, <em class="sig-param">'SA': 0.1</em>, <em class="sig-param">'SC': 0.185</em>, <em class="sig-param">'SD': 0.801</em>, <em class="sig-param">'SE': 0.812</em>, <em class="sig-param">'SF': 0.622</em>, <em class="sig-param">'SG': 0.17</em>, <em class="sig-param">'SH': 0.718</em>, <em class="sig-param">'SI': 0.478</em>, <em class="sig-param">'SK': 0.883</em>, <em class="sig-param">'SL': 0.474</em>, <em class="sig-param">'SM': 0.44</em>, <em class="sig-param">'SN': 0.289</em>, <em class="sig-param">'SP': 0.181</em>, <em class="sig-param">'SQ': 0.358</em>, <em class="sig-param">'SR': 1.0</em>, <em class="sig-param">'SS': 0.0</em>, <em class="sig-param">'ST': 0.174</em>, <em class="sig-param">'SV': 0.342</em>, <em class="sig-param">'SW': 0.827</em>, <em class="sig-param">'SY': 0.615</em>, <em class="sig-param">'TA': 0.251</em>, <em class="sig-param">'TC': 0.261</em>, <em class="sig-param">'TD': 0.83</em>, <em class="sig-param">'TE': 0.812</em>, <em class="sig-param">'TF': 0.604</em>, <em class="sig-param">'TG': 0.312</em>, <em class="sig-param">'TH': 0.737</em>, <em class="sig-param">'TI': 0.455</em>, <em class="sig-param">'TK': 0.866</em>, <em class="sig-param">'TL': 0.453</em>, <em class="sig-param">'TM': 0.403</em>, <em class="sig-param">'TN': 0.315</em>, <em class="sig-param">'TP': 0.159</em>, <em class="sig-param">'TQ': 0.322</em>, <em class="sig-param">'TR': 1.0</em>, <em class="sig-param">'TS': 0.185</em>, <em class="sig-param">'TT': 0.0</em>, <em class="sig-param">'TV': 0.345</em>, <em class="sig-param">'TW': 0.816</em>, <em class="sig-param">'TY': 0.596</em>, <em class="sig-param">'VA': 0.275</em>, <em class="sig-param">'VC': 0.165</em>, <em class="sig-param">'VD': 0.9</em>, <em class="sig-param">'VE': 0.867</em>, <em class="sig-param">'VF': 0.269</em>, <em class="sig-param">'VG': 0.471</em>, <em class="sig-param">'VH': 0.649</em>, <em class="sig-param">'VI': 0.135</em>, <em class="sig-param">'VK': 0.889</em>, <em class="sig-param">'VL': 0.134</em>, <em class="sig-param">'VM': 0.12</em>, <em class="sig-param">'VN': 0.38</em>, <em class="sig-param">'VP': 0.212</em>, <em class="sig-param">'VQ': 0.339</em>, <em class="sig-param">'VR': 1.0</em>, <em class="sig-param">'VS': 0.322</em>, <em class="sig-param">'VT': 0.305</em>, <em class="sig-param">'VV': 0.0</em>, <em class="sig-param">'VW': 0.472</em>, <em class="sig-param">'VY': 0.31</em>, <em class="sig-param">'WA': 0.658</em>, <em class="sig-param">'WC': 0.56</em>, <em class="sig-param">'WD': 1.0</em>, <em class="sig-param">'WE': 0.931</em>, <em class="sig-param">'WF': 0.196</em>, <em class="sig-param">'WG': 0.829</em>, <em class="sig-param">'WH': 0.678</em>, <em class="sig-param">'WI': 0.305</em>, <em class="sig-param">'WK': 0.892</em>, <em class="sig-param">'WL': 0.304</em>, <em class="sig-param">'WM': 0.344</em>, <em class="sig-param">'WN': 0.631</em>, <em class="sig-param">'WP': 0.555</em>, <em class="sig-param">'WQ': 0.538</em>, <em class="sig-param">'WR': 0.968</em>, <em class="sig-param">'WS': 0.689</em>, <em class="sig-param">'WT': 0.638</em>, <em class="sig-param">'WV': 0.418</em>, <em class="sig-param">'WW': 0.0</em>, <em class="sig-param">'WY': 0.204</em>, <em class="sig-param">'YA': 0.587</em>, <em class="sig-param">'YC': 0.478</em>, <em class="sig-param">'YD': 1.0</em>, <em class="sig-param">'YE': 0.932</em>, <em class="sig-param">'YF': 0.202</em>, <em class="sig-param">'YG': 0.782</em>, <em class="sig-param">'YH': 0.678</em>, <em class="sig-param">'YI': 0.23</em>, <em class="sig-param">'YK': 0.904</em>, <em class="sig-param">'YL': 0.219</em>, <em class="sig-param">'YM': 0.268</em>, <em class="sig-param">'YN': 0.512</em>, <em class="sig-param">'YP': 0.444</em>, <em class="sig-param">'YQ': 0.404</em>, <em class="sig-param">'YR': 0.995</em>, <em class="sig-param">'YS': 0.612</em>, <em class="sig-param">'YT': 0.557</em>, <em class="sig-param">'YV': 0.328</em>, <em class="sig-param">'YW': 0.244</em>, <em class="sig-param">'YY': 0.0}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_quasi_sequence_order2_grant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order2_grant" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the last maxlag quasi-sequence-order for a given protein sequence.
see method get_quasi_sequence_order for the choice of parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – </p></li>
<li><p><strong>maxlag</strong> – </p></li>
<li><p><strong>weight</strong> – </p></li>
<li><p><strong>distancematrix</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order2_sw">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_quasi_sequence_order2_sw</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">weight=0.1</em>, <em class="sig-param">distancematrix={'AA': 0.0</em>, <em class="sig-param">'AC': 0.112</em>, <em class="sig-param">'AD': 0.819</em>, <em class="sig-param">'AE': 0.827</em>, <em class="sig-param">'AF': 0.54</em>, <em class="sig-param">'AG': 0.208</em>, <em class="sig-param">'AH': 0.696</em>, <em class="sig-param">'AI': 0.407</em>, <em class="sig-param">'AK': 0.891</em>, <em class="sig-param">'AL': 0.406</em>, <em class="sig-param">'AM': 0.379</em>, <em class="sig-param">'AN': 0.318</em>, <em class="sig-param">'AP': 0.191</em>, <em class="sig-param">'AQ': 0.372</em>, <em class="sig-param">'AR': 1.0</em>, <em class="sig-param">'AS': 0.094</em>, <em class="sig-param">'AT': 0.22</em>, <em class="sig-param">'AV': 0.273</em>, <em class="sig-param">'AW': 0.739</em>, <em class="sig-param">'AY': 0.552</em>, <em class="sig-param">'CA': 0.114</em>, <em class="sig-param">'CC': 0.0</em>, <em class="sig-param">'CD': 0.847</em>, <em class="sig-param">'CE': 0.838</em>, <em class="sig-param">'CF': 0.437</em>, <em class="sig-param">'CG': 0.32</em>, <em class="sig-param">'CH': 0.66</em>, <em class="sig-param">'CI': 0.304</em>, <em class="sig-param">'CK': 0.887</em>, <em class="sig-param">'CL': 0.301</em>, <em class="sig-param">'CM': 0.277</em>, <em class="sig-param">'CN': 0.324</em>, <em class="sig-param">'CP': 0.157</em>, <em class="sig-param">'CQ': 0.341</em>, <em class="sig-param">'CR': 1.0</em>, <em class="sig-param">'CS': 0.176</em>, <em class="sig-param">'CT': 0.233</em>, <em class="sig-param">'CV': 0.167</em>, <em class="sig-param">'CW': 0.639</em>, <em class="sig-param">'CY': 0.457</em>, <em class="sig-param">'DA': 0.729</em>, <em class="sig-param">'DC': 0.742</em>, <em class="sig-param">'DD': 0.0</em>, <em class="sig-param">'DE': 0.124</em>, <em class="sig-param">'DF': 0.924</em>, <em class="sig-param">'DG': 0.697</em>, <em class="sig-param">'DH': 0.435</em>, <em class="sig-param">'DI': 0.847</em>, <em class="sig-param">'DK': 0.249</em>, <em class="sig-param">'DL': 0.841</em>, <em class="sig-param">'DM': 0.819</em>, <em class="sig-param">'DN': 0.56</em>, <em class="sig-param">'DP': 0.657</em>, <em class="sig-param">'DQ': 0.584</em>, <em class="sig-param">'DR': 0.295</em>, <em class="sig-param">'DS': 0.667</em>, <em class="sig-param">'DT': 0.649</em>, <em class="sig-param">'DV': 0.797</em>, <em class="sig-param">'DW': 1.0</em>, <em class="sig-param">'DY': 0.836</em>, <em class="sig-param">'EA': 0.79</em>, <em class="sig-param">'EC': 0.788</em>, <em class="sig-param">'ED': 0.133</em>, <em class="sig-param">'EE': 0.0</em>, <em class="sig-param">'EF': 0.932</em>, <em class="sig-param">'EG': 0.779</em>, <em class="sig-param">'EH': 0.406</em>, <em class="sig-param">'EI': 0.86</em>, <em class="sig-param">'EK': 0.143</em>, <em class="sig-param">'EL': 0.854</em>, <em class="sig-param">'EM': 0.83</em>, <em class="sig-param">'EN': 0.599</em>, <em class="sig-param">'EP': 0.688</em>, <em class="sig-param">'EQ': 0.598</em>, <em class="sig-param">'ER': 0.234</em>, <em class="sig-param">'ES': 0.726</em>, <em class="sig-param">'ET': 0.682</em>, <em class="sig-param">'EV': 0.824</em>, <em class="sig-param">'EW': 1.0</em>, <em class="sig-param">'EY': 0.837</em>, <em class="sig-param">'FA': 0.508</em>, <em class="sig-param">'FC': 0.405</em>, <em class="sig-param">'FD': 0.977</em>, <em class="sig-param">'FE': 0.918</em>, <em class="sig-param">'FF': 0.0</em>, <em class="sig-param">'FG': 0.69</em>, <em class="sig-param">'FH': 0.663</em>, <em class="sig-param">'FI': 0.128</em>, <em class="sig-param">'FK': 0.903</em>, <em class="sig-param">'FL': 0.131</em>, <em class="sig-param">'FM': 0.169</em>, <em class="sig-param">'FN': 0.541</em>, <em class="sig-param">'FP': 0.42</em>, <em class="sig-param">'FQ': 0.459</em>, <em class="sig-param">'FR': 1.0</em>, <em class="sig-param">'FS': 0.548</em>, <em class="sig-param">'FT': 0.499</em>, <em class="sig-param">'FV': 0.252</em>, <em class="sig-param">'FW': 0.207</em>, <em class="sig-param">'FY': 0.179</em>, <em class="sig-param">'GA': 0.206</em>, <em class="sig-param">'GC': 0.312</em>, <em class="sig-param">'GD': 0.776</em>, <em class="sig-param">'GE': 0.807</em>, <em class="sig-param">'GF': 0.727</em>, <em class="sig-param">'GG': 0.0</em>, <em class="sig-param">'GH': 0.769</em>, <em class="sig-param">'GI': 0.592</em>, <em class="sig-param">'GK': 0.894</em>, <em class="sig-param">'GL': 0.591</em>, <em class="sig-param">'GM': 0.557</em>, <em class="sig-param">'GN': 0.381</em>, <em class="sig-param">'GP': 0.323</em>, <em class="sig-param">'GQ': 0.467</em>, <em class="sig-param">'GR': 1.0</em>, <em class="sig-param">'GS': 0.158</em>, <em class="sig-param">'GT': 0.272</em>, <em class="sig-param">'GV': 0.464</em>, <em class="sig-param">'GW': 0.923</em>, <em class="sig-param">'GY': 0.728</em>, <em class="sig-param">'HA': 0.896</em>, <em class="sig-param">'HC': 0.836</em>, <em class="sig-param">'HD': 0.629</em>, <em class="sig-param">'HE': 0.547</em>, <em class="sig-param">'HF': 0.907</em>, <em class="sig-param">'HG': 1.0</em>, <em class="sig-param">'HH': 0.0</em>, <em class="sig-param">'HI': 0.848</em>, <em class="sig-param">'HK': 0.566</em>, <em class="sig-param">'HL': 0.842</em>, <em class="sig-param">'HM': 0.825</em>, <em class="sig-param">'HN': 0.754</em>, <em class="sig-param">'HP': 0.777</em>, <em class="sig-param">'HQ': 0.716</em>, <em class="sig-param">'HR': 0.697</em>, <em class="sig-param">'HS': 0.865</em>, <em class="sig-param">'HT': 0.834</em>, <em class="sig-param">'HV': 0.831</em>, <em class="sig-param">'HW': 0.981</em>, <em class="sig-param">'HY': 0.821</em>, <em class="sig-param">'IA': 0.403</em>, <em class="sig-param">'IC': 0.296</em>, <em class="sig-param">'ID': 0.942</em>, <em class="sig-param">'IE': 0.891</em>, <em class="sig-param">'IF': 0.134</em>, <em class="sig-param">'IG': 0.592</em>, <em class="sig-param">'IH': 0.652</em>, <em class="sig-param">'II': 0.0</em>, <em class="sig-param">'IK': 0.892</em>, <em class="sig-param">'IL': 0.013</em>, <em class="sig-param">'IM': 0.057</em>, <em class="sig-param">'IN': 0.457</em>, <em class="sig-param">'IP': 0.311</em>, <em class="sig-param">'IQ': 0.383</em>, <em class="sig-param">'IR': 1.0</em>, <em class="sig-param">'IS': 0.443</em>, <em class="sig-param">'IT': 0.396</em>, <em class="sig-param">'IV': 0.133</em>, <em class="sig-param">'IW': 0.339</em>, <em class="sig-param">'IY': 0.213</em>, <em class="sig-param">'KA': 0.889</em>, <em class="sig-param">'KC': 0.871</em>, <em class="sig-param">'KD': 0.279</em>, <em class="sig-param">'KE': 0.149</em>, <em class="sig-param">'KF': 0.957</em>, <em class="sig-param">'KG': 0.9</em>, <em class="sig-param">'KH': 0.438</em>, <em class="sig-param">'KI': 0.899</em>, <em class="sig-param">'KK': 0.0</em>, <em class="sig-param">'KL': 0.892</em>, <em class="sig-param">'KM': 0.871</em>, <em class="sig-param">'KN': 0.667</em>, <em class="sig-param">'KP': 0.757</em>, <em class="sig-param">'KQ': 0.639</em>, <em class="sig-param">'KR': 0.154</em>, <em class="sig-param">'KS': 0.825</em>, <em class="sig-param">'KT': 0.759</em>, <em class="sig-param">'KV': 0.882</em>, <em class="sig-param">'KW': 1.0</em>, <em class="sig-param">'KY': 0.848</em>, <em class="sig-param">'LA': 0.405</em>, <em class="sig-param">'LC': 0.296</em>, <em class="sig-param">'LD': 0.944</em>, <em class="sig-param">'LE': 0.892</em>, <em class="sig-param">'LF': 0.139</em>, <em class="sig-param">'LG': 0.596</em>, <em class="sig-param">'LH': 0.653</em>, <em class="sig-param">'LI': 0.013</em>, <em class="sig-param">'LK': 0.893</em>, <em class="sig-param">'LL': 0.0</em>, <em class="sig-param">'LM': 0.062</em>, <em class="sig-param">'LN': 0.452</em>, <em class="sig-param">'LP': 0.309</em>, <em class="sig-param">'LQ': 0.376</em>, <em class="sig-param">'LR': 1.0</em>, <em class="sig-param">'LS': 0.443</em>, <em class="sig-param">'LT': 0.397</em>, <em class="sig-param">'LV': 0.133</em>, <em class="sig-param">'LW': 0.341</em>, <em class="sig-param">'LY': 0.205</em>, <em class="sig-param">'MA': 0.383</em>, <em class="sig-param">'MC': 0.276</em>, <em class="sig-param">'MD': 0.932</em>, <em class="sig-param">'ME': 0.879</em>, <em class="sig-param">'MF': 0.182</em>, <em class="sig-param">'MG': 0.569</em>, <em class="sig-param">'MH': 0.648</em>, <em class="sig-param">'MI': 0.058</em>, <em class="sig-param">'MK': 0.884</em>, <em class="sig-param">'ML': 0.062</em>, <em class="sig-param">'MM': 0.0</em>, <em class="sig-param">'MN': 0.447</em>, <em class="sig-param">'MP': 0.285</em>, <em class="sig-param">'MQ': 0.372</em>, <em class="sig-param">'MR': 1.0</em>, <em class="sig-param">'MS': 0.417</em>, <em class="sig-param">'MT': 0.358</em>, <em class="sig-param">'MV': 0.12</em>, <em class="sig-param">'MW': 0.391</em>, <em class="sig-param">'MY': 0.255</em>, <em class="sig-param">'NA': 0.424</em>, <em class="sig-param">'NC': 0.425</em>, <em class="sig-param">'ND': 0.838</em>, <em class="sig-param">'NE': 0.835</em>, <em class="sig-param">'NF': 0.766</em>, <em class="sig-param">'NG': 0.512</em>, <em class="sig-param">'NH': 0.78</em>, <em class="sig-param">'NI': 0.615</em>, <em class="sig-param">'NK': 0.891</em>, <em class="sig-param">'NL': 0.603</em>, <em class="sig-param">'NM': 0.588</em>, <em class="sig-param">'NN': 0.0</em>, <em class="sig-param">'NP': 0.266</em>, <em class="sig-param">'NQ': 0.175</em>, <em class="sig-param">'NR': 1.0</em>, <em class="sig-param">'NS': 0.361</em>, <em class="sig-param">'NT': 0.368</em>, <em class="sig-param">'NV': 0.503</em>, <em class="sig-param">'NW': 0.945</em>, <em class="sig-param">'NY': 0.641</em>, <em class="sig-param">'PA': 0.22</em>, <em class="sig-param">'PC': 0.179</em>, <em class="sig-param">'PD': 0.852</em>, <em class="sig-param">'PE': 0.831</em>, <em class="sig-param">'PF': 0.515</em>, <em class="sig-param">'PG': 0.376</em>, <em class="sig-param">'PH': 0.696</em>, <em class="sig-param">'PI': 0.363</em>, <em class="sig-param">'PK': 0.875</em>, <em class="sig-param">'PL': 0.357</em>, <em class="sig-param">'PM': 0.326</em>, <em class="sig-param">'PN': 0.231</em>, <em class="sig-param">'PP': 0.0</em>, <em class="sig-param">'PQ': 0.228</em>, <em class="sig-param">'PR': 1.0</em>, <em class="sig-param">'PS': 0.196</em>, <em class="sig-param">'PT': 0.161</em>, <em class="sig-param">'PV': 0.244</em>, <em class="sig-param">'PW': 0.72</em>, <em class="sig-param">'PY': 0.481</em>, <em class="sig-param">'QA': 0.512</em>, <em class="sig-param">'QC': 0.462</em>, <em class="sig-param">'QD': 0.903</em>, <em class="sig-param">'QE': 0.861</em>, <em class="sig-param">'QF': 0.671</em>, <em class="sig-param">'QG': 0.648</em>, <em class="sig-param">'QH': 0.765</em>, <em class="sig-param">'QI': 0.532</em>, <em class="sig-param">'QK': 0.881</em>, <em class="sig-param">'QL': 0.518</em>, <em class="sig-param">'QM': 0.505</em>, <em class="sig-param">'QN': 0.181</em>, <em class="sig-param">'QP': 0.272</em>, <em class="sig-param">'QQ': 0.0</em>, <em class="sig-param">'QR': 1.0</em>, <em class="sig-param">'QS': 0.461</em>, <em class="sig-param">'QT': 0.389</em>, <em class="sig-param">'QV': 0.464</em>, <em class="sig-param">'QW': 0.831</em>, <em class="sig-param">'QY': 0.522</em>, <em class="sig-param">'RA': 0.919</em>, <em class="sig-param">'RC': 0.905</em>, <em class="sig-param">'RD': 0.305</em>, <em class="sig-param">'RE': 0.225</em>, <em class="sig-param">'RF': 0.977</em>, <em class="sig-param">'RG': 0.928</em>, <em class="sig-param">'RH': 0.498</em>, <em class="sig-param">'RI': 0.929</em>, <em class="sig-param">'RK': 0.141</em>, <em class="sig-param">'RL': 0.92</em>, <em class="sig-param">'RM': 0.908</em>, <em class="sig-param">'RN': 0.69</em>, <em class="sig-param">'RP': 0.796</em>, <em class="sig-param">'RQ': 0.668</em>, <em class="sig-param">'RR': 0.0</em>, <em class="sig-param">'RS': 0.86</em>, <em class="sig-param">'RT': 0.808</em>, <em class="sig-param">'RV': 0.914</em>, <em class="sig-param">'RW': 1.0</em>, <em class="sig-param">'RY': 0.859</em>, <em class="sig-param">'SA': 0.1</em>, <em class="sig-param">'SC': 0.185</em>, <em class="sig-param">'SD': 0.801</em>, <em class="sig-param">'SE': 0.812</em>, <em class="sig-param">'SF': 0.622</em>, <em class="sig-param">'SG': 0.17</em>, <em class="sig-param">'SH': 0.718</em>, <em class="sig-param">'SI': 0.478</em>, <em class="sig-param">'SK': 0.883</em>, <em class="sig-param">'SL': 0.474</em>, <em class="sig-param">'SM': 0.44</em>, <em class="sig-param">'SN': 0.289</em>, <em class="sig-param">'SP': 0.181</em>, <em class="sig-param">'SQ': 0.358</em>, <em class="sig-param">'SR': 1.0</em>, <em class="sig-param">'SS': 0.0</em>, <em class="sig-param">'ST': 0.174</em>, <em class="sig-param">'SV': 0.342</em>, <em class="sig-param">'SW': 0.827</em>, <em class="sig-param">'SY': 0.615</em>, <em class="sig-param">'TA': 0.251</em>, <em class="sig-param">'TC': 0.261</em>, <em class="sig-param">'TD': 0.83</em>, <em class="sig-param">'TE': 0.812</em>, <em class="sig-param">'TF': 0.604</em>, <em class="sig-param">'TG': 0.312</em>, <em class="sig-param">'TH': 0.737</em>, <em class="sig-param">'TI': 0.455</em>, <em class="sig-param">'TK': 0.866</em>, <em class="sig-param">'TL': 0.453</em>, <em class="sig-param">'TM': 0.403</em>, <em class="sig-param">'TN': 0.315</em>, <em class="sig-param">'TP': 0.159</em>, <em class="sig-param">'TQ': 0.322</em>, <em class="sig-param">'TR': 1.0</em>, <em class="sig-param">'TS': 0.185</em>, <em class="sig-param">'TT': 0.0</em>, <em class="sig-param">'TV': 0.345</em>, <em class="sig-param">'TW': 0.816</em>, <em class="sig-param">'TY': 0.596</em>, <em class="sig-param">'VA': 0.275</em>, <em class="sig-param">'VC': 0.165</em>, <em class="sig-param">'VD': 0.9</em>, <em class="sig-param">'VE': 0.867</em>, <em class="sig-param">'VF': 0.269</em>, <em class="sig-param">'VG': 0.471</em>, <em class="sig-param">'VH': 0.649</em>, <em class="sig-param">'VI': 0.135</em>, <em class="sig-param">'VK': 0.889</em>, <em class="sig-param">'VL': 0.134</em>, <em class="sig-param">'VM': 0.12</em>, <em class="sig-param">'VN': 0.38</em>, <em class="sig-param">'VP': 0.212</em>, <em class="sig-param">'VQ': 0.339</em>, <em class="sig-param">'VR': 1.0</em>, <em class="sig-param">'VS': 0.322</em>, <em class="sig-param">'VT': 0.305</em>, <em class="sig-param">'VV': 0.0</em>, <em class="sig-param">'VW': 0.472</em>, <em class="sig-param">'VY': 0.31</em>, <em class="sig-param">'WA': 0.658</em>, <em class="sig-param">'WC': 0.56</em>, <em class="sig-param">'WD': 1.0</em>, <em class="sig-param">'WE': 0.931</em>, <em class="sig-param">'WF': 0.196</em>, <em class="sig-param">'WG': 0.829</em>, <em class="sig-param">'WH': 0.678</em>, <em class="sig-param">'WI': 0.305</em>, <em class="sig-param">'WK': 0.892</em>, <em class="sig-param">'WL': 0.304</em>, <em class="sig-param">'WM': 0.344</em>, <em class="sig-param">'WN': 0.631</em>, <em class="sig-param">'WP': 0.555</em>, <em class="sig-param">'WQ': 0.538</em>, <em class="sig-param">'WR': 0.968</em>, <em class="sig-param">'WS': 0.689</em>, <em class="sig-param">'WT': 0.638</em>, <em class="sig-param">'WV': 0.418</em>, <em class="sig-param">'WW': 0.0</em>, <em class="sig-param">'WY': 0.204</em>, <em class="sig-param">'YA': 0.587</em>, <em class="sig-param">'YC': 0.478</em>, <em class="sig-param">'YD': 1.0</em>, <em class="sig-param">'YE': 0.932</em>, <em class="sig-param">'YF': 0.202</em>, <em class="sig-param">'YG': 0.782</em>, <em class="sig-param">'YH': 0.678</em>, <em class="sig-param">'YI': 0.23</em>, <em class="sig-param">'YK': 0.904</em>, <em class="sig-param">'YL': 0.219</em>, <em class="sig-param">'YM': 0.268</em>, <em class="sig-param">'YN': 0.512</em>, <em class="sig-param">'YP': 0.444</em>, <em class="sig-param">'YQ': 0.404</em>, <em class="sig-param">'YR': 0.995</em>, <em class="sig-param">'YS': 0.612</em>, <em class="sig-param">'YT': 0.557</em>, <em class="sig-param">'YV': 0.328</em>, <em class="sig-param">'YW': 0.244</em>, <em class="sig-param">'YY': 0.0}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_quasi_sequence_order2_sw"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_order2_sw" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the last maxlag quasi-sequence-order for a given protein sequence .
see method get_quasi_sequence_order for the choice of parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – </p></li>
<li><p><strong>maxlag</strong> – </p></li>
<li><p><strong>weight</strong> – </p></li>
<li><p><strong>distancematrix</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_orderp">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_quasi_sequence_orderp</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">weight=0.1</em>, <em class="sig-param">distancematrix={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_quasi_sequence_orderp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_quasi_sequence_orderp" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing quasi-sequence-order pydpi_py3 for a given protein.
[1]:Kuo-Chen Chou. Prediction of Protein Subcellar Locations by
Incorporating Quasi-Sequence-Order Effect. Biochemical and Biophysical
Research Communications 2000, 278, 477-483.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>maxlag</strong> – maxlag is the maximum lag and the length of the protein should be larger than maxlag. default is 30.</p></li>
<li><p><strong>weight</strong> – weight is a weight factor.  please see reference 1 for its choice. default is 0.1.</p></li>
<li><p><strong>distancematrix</strong> – distancematrix is a dict form containing 400 distance values</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing all quasi-sequence-order</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_sequence_order_coupling_number">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_sequence_order_coupling_number</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">d=1</em>, <em class="sig-param">distancematrix={'AA': 0.0</em>, <em class="sig-param">'AC': 0.112</em>, <em class="sig-param">'AD': 0.819</em>, <em class="sig-param">'AE': 0.827</em>, <em class="sig-param">'AF': 0.54</em>, <em class="sig-param">'AG': 0.208</em>, <em class="sig-param">'AH': 0.696</em>, <em class="sig-param">'AI': 0.407</em>, <em class="sig-param">'AK': 0.891</em>, <em class="sig-param">'AL': 0.406</em>, <em class="sig-param">'AM': 0.379</em>, <em class="sig-param">'AN': 0.318</em>, <em class="sig-param">'AP': 0.191</em>, <em class="sig-param">'AQ': 0.372</em>, <em class="sig-param">'AR': 1.0</em>, <em class="sig-param">'AS': 0.094</em>, <em class="sig-param">'AT': 0.22</em>, <em class="sig-param">'AV': 0.273</em>, <em class="sig-param">'AW': 0.739</em>, <em class="sig-param">'AY': 0.552</em>, <em class="sig-param">'CA': 0.114</em>, <em class="sig-param">'CC': 0.0</em>, <em class="sig-param">'CD': 0.847</em>, <em class="sig-param">'CE': 0.838</em>, <em class="sig-param">'CF': 0.437</em>, <em class="sig-param">'CG': 0.32</em>, <em class="sig-param">'CH': 0.66</em>, <em class="sig-param">'CI': 0.304</em>, <em class="sig-param">'CK': 0.887</em>, <em class="sig-param">'CL': 0.301</em>, <em class="sig-param">'CM': 0.277</em>, <em class="sig-param">'CN': 0.324</em>, <em class="sig-param">'CP': 0.157</em>, <em class="sig-param">'CQ': 0.341</em>, <em class="sig-param">'CR': 1.0</em>, <em class="sig-param">'CS': 0.176</em>, <em class="sig-param">'CT': 0.233</em>, <em class="sig-param">'CV': 0.167</em>, <em class="sig-param">'CW': 0.639</em>, <em class="sig-param">'CY': 0.457</em>, <em class="sig-param">'DA': 0.729</em>, <em class="sig-param">'DC': 0.742</em>, <em class="sig-param">'DD': 0.0</em>, <em class="sig-param">'DE': 0.124</em>, <em class="sig-param">'DF': 0.924</em>, <em class="sig-param">'DG': 0.697</em>, <em class="sig-param">'DH': 0.435</em>, <em class="sig-param">'DI': 0.847</em>, <em class="sig-param">'DK': 0.249</em>, <em class="sig-param">'DL': 0.841</em>, <em class="sig-param">'DM': 0.819</em>, <em class="sig-param">'DN': 0.56</em>, <em class="sig-param">'DP': 0.657</em>, <em class="sig-param">'DQ': 0.584</em>, <em class="sig-param">'DR': 0.295</em>, <em class="sig-param">'DS': 0.667</em>, <em class="sig-param">'DT': 0.649</em>, <em class="sig-param">'DV': 0.797</em>, <em class="sig-param">'DW': 1.0</em>, <em class="sig-param">'DY': 0.836</em>, <em class="sig-param">'EA': 0.79</em>, <em class="sig-param">'EC': 0.788</em>, <em class="sig-param">'ED': 0.133</em>, <em class="sig-param">'EE': 0.0</em>, <em class="sig-param">'EF': 0.932</em>, <em class="sig-param">'EG': 0.779</em>, <em class="sig-param">'EH': 0.406</em>, <em class="sig-param">'EI': 0.86</em>, <em class="sig-param">'EK': 0.143</em>, <em class="sig-param">'EL': 0.854</em>, <em class="sig-param">'EM': 0.83</em>, <em class="sig-param">'EN': 0.599</em>, <em class="sig-param">'EP': 0.688</em>, <em class="sig-param">'EQ': 0.598</em>, <em class="sig-param">'ER': 0.234</em>, <em class="sig-param">'ES': 0.726</em>, <em class="sig-param">'ET': 0.682</em>, <em class="sig-param">'EV': 0.824</em>, <em class="sig-param">'EW': 1.0</em>, <em class="sig-param">'EY': 0.837</em>, <em class="sig-param">'FA': 0.508</em>, <em class="sig-param">'FC': 0.405</em>, <em class="sig-param">'FD': 0.977</em>, <em class="sig-param">'FE': 0.918</em>, <em class="sig-param">'FF': 0.0</em>, <em class="sig-param">'FG': 0.69</em>, <em class="sig-param">'FH': 0.663</em>, <em class="sig-param">'FI': 0.128</em>, <em class="sig-param">'FK': 0.903</em>, <em class="sig-param">'FL': 0.131</em>, <em class="sig-param">'FM': 0.169</em>, <em class="sig-param">'FN': 0.541</em>, <em class="sig-param">'FP': 0.42</em>, <em class="sig-param">'FQ': 0.459</em>, <em class="sig-param">'FR': 1.0</em>, <em class="sig-param">'FS': 0.548</em>, <em class="sig-param">'FT': 0.499</em>, <em class="sig-param">'FV': 0.252</em>, <em class="sig-param">'FW': 0.207</em>, <em class="sig-param">'FY': 0.179</em>, <em class="sig-param">'GA': 0.206</em>, <em class="sig-param">'GC': 0.312</em>, <em class="sig-param">'GD': 0.776</em>, <em class="sig-param">'GE': 0.807</em>, <em class="sig-param">'GF': 0.727</em>, <em class="sig-param">'GG': 0.0</em>, <em class="sig-param">'GH': 0.769</em>, <em class="sig-param">'GI': 0.592</em>, <em class="sig-param">'GK': 0.894</em>, <em class="sig-param">'GL': 0.591</em>, <em class="sig-param">'GM': 0.557</em>, <em class="sig-param">'GN': 0.381</em>, <em class="sig-param">'GP': 0.323</em>, <em class="sig-param">'GQ': 0.467</em>, <em class="sig-param">'GR': 1.0</em>, <em class="sig-param">'GS': 0.158</em>, <em class="sig-param">'GT': 0.272</em>, <em class="sig-param">'GV': 0.464</em>, <em class="sig-param">'GW': 0.923</em>, <em class="sig-param">'GY': 0.728</em>, <em class="sig-param">'HA': 0.896</em>, <em class="sig-param">'HC': 0.836</em>, <em class="sig-param">'HD': 0.629</em>, <em class="sig-param">'HE': 0.547</em>, <em class="sig-param">'HF': 0.907</em>, <em class="sig-param">'HG': 1.0</em>, <em class="sig-param">'HH': 0.0</em>, <em class="sig-param">'HI': 0.848</em>, <em class="sig-param">'HK': 0.566</em>, <em class="sig-param">'HL': 0.842</em>, <em class="sig-param">'HM': 0.825</em>, <em class="sig-param">'HN': 0.754</em>, <em class="sig-param">'HP': 0.777</em>, <em class="sig-param">'HQ': 0.716</em>, <em class="sig-param">'HR': 0.697</em>, <em class="sig-param">'HS': 0.865</em>, <em class="sig-param">'HT': 0.834</em>, <em class="sig-param">'HV': 0.831</em>, <em class="sig-param">'HW': 0.981</em>, <em class="sig-param">'HY': 0.821</em>, <em class="sig-param">'IA': 0.403</em>, <em class="sig-param">'IC': 0.296</em>, <em class="sig-param">'ID': 0.942</em>, <em class="sig-param">'IE': 0.891</em>, <em class="sig-param">'IF': 0.134</em>, <em class="sig-param">'IG': 0.592</em>, <em class="sig-param">'IH': 0.652</em>, <em class="sig-param">'II': 0.0</em>, <em class="sig-param">'IK': 0.892</em>, <em class="sig-param">'IL': 0.013</em>, <em class="sig-param">'IM': 0.057</em>, <em class="sig-param">'IN': 0.457</em>, <em class="sig-param">'IP': 0.311</em>, <em class="sig-param">'IQ': 0.383</em>, <em class="sig-param">'IR': 1.0</em>, <em class="sig-param">'IS': 0.443</em>, <em class="sig-param">'IT': 0.396</em>, <em class="sig-param">'IV': 0.133</em>, <em class="sig-param">'IW': 0.339</em>, <em class="sig-param">'IY': 0.213</em>, <em class="sig-param">'KA': 0.889</em>, <em class="sig-param">'KC': 0.871</em>, <em class="sig-param">'KD': 0.279</em>, <em class="sig-param">'KE': 0.149</em>, <em class="sig-param">'KF': 0.957</em>, <em class="sig-param">'KG': 0.9</em>, <em class="sig-param">'KH': 0.438</em>, <em class="sig-param">'KI': 0.899</em>, <em class="sig-param">'KK': 0.0</em>, <em class="sig-param">'KL': 0.892</em>, <em class="sig-param">'KM': 0.871</em>, <em class="sig-param">'KN': 0.667</em>, <em class="sig-param">'KP': 0.757</em>, <em class="sig-param">'KQ': 0.639</em>, <em class="sig-param">'KR': 0.154</em>, <em class="sig-param">'KS': 0.825</em>, <em class="sig-param">'KT': 0.759</em>, <em class="sig-param">'KV': 0.882</em>, <em class="sig-param">'KW': 1.0</em>, <em class="sig-param">'KY': 0.848</em>, <em class="sig-param">'LA': 0.405</em>, <em class="sig-param">'LC': 0.296</em>, <em class="sig-param">'LD': 0.944</em>, <em class="sig-param">'LE': 0.892</em>, <em class="sig-param">'LF': 0.139</em>, <em class="sig-param">'LG': 0.596</em>, <em class="sig-param">'LH': 0.653</em>, <em class="sig-param">'LI': 0.013</em>, <em class="sig-param">'LK': 0.893</em>, <em class="sig-param">'LL': 0.0</em>, <em class="sig-param">'LM': 0.062</em>, <em class="sig-param">'LN': 0.452</em>, <em class="sig-param">'LP': 0.309</em>, <em class="sig-param">'LQ': 0.376</em>, <em class="sig-param">'LR': 1.0</em>, <em class="sig-param">'LS': 0.443</em>, <em class="sig-param">'LT': 0.397</em>, <em class="sig-param">'LV': 0.133</em>, <em class="sig-param">'LW': 0.341</em>, <em class="sig-param">'LY': 0.205</em>, <em class="sig-param">'MA': 0.383</em>, <em class="sig-param">'MC': 0.276</em>, <em class="sig-param">'MD': 0.932</em>, <em class="sig-param">'ME': 0.879</em>, <em class="sig-param">'MF': 0.182</em>, <em class="sig-param">'MG': 0.569</em>, <em class="sig-param">'MH': 0.648</em>, <em class="sig-param">'MI': 0.058</em>, <em class="sig-param">'MK': 0.884</em>, <em class="sig-param">'ML': 0.062</em>, <em class="sig-param">'MM': 0.0</em>, <em class="sig-param">'MN': 0.447</em>, <em class="sig-param">'MP': 0.285</em>, <em class="sig-param">'MQ': 0.372</em>, <em class="sig-param">'MR': 1.0</em>, <em class="sig-param">'MS': 0.417</em>, <em class="sig-param">'MT': 0.358</em>, <em class="sig-param">'MV': 0.12</em>, <em class="sig-param">'MW': 0.391</em>, <em class="sig-param">'MY': 0.255</em>, <em class="sig-param">'NA': 0.424</em>, <em class="sig-param">'NC': 0.425</em>, <em class="sig-param">'ND': 0.838</em>, <em class="sig-param">'NE': 0.835</em>, <em class="sig-param">'NF': 0.766</em>, <em class="sig-param">'NG': 0.512</em>, <em class="sig-param">'NH': 0.78</em>, <em class="sig-param">'NI': 0.615</em>, <em class="sig-param">'NK': 0.891</em>, <em class="sig-param">'NL': 0.603</em>, <em class="sig-param">'NM': 0.588</em>, <em class="sig-param">'NN': 0.0</em>, <em class="sig-param">'NP': 0.266</em>, <em class="sig-param">'NQ': 0.175</em>, <em class="sig-param">'NR': 1.0</em>, <em class="sig-param">'NS': 0.361</em>, <em class="sig-param">'NT': 0.368</em>, <em class="sig-param">'NV': 0.503</em>, <em class="sig-param">'NW': 0.945</em>, <em class="sig-param">'NY': 0.641</em>, <em class="sig-param">'PA': 0.22</em>, <em class="sig-param">'PC': 0.179</em>, <em class="sig-param">'PD': 0.852</em>, <em class="sig-param">'PE': 0.831</em>, <em class="sig-param">'PF': 0.515</em>, <em class="sig-param">'PG': 0.376</em>, <em class="sig-param">'PH': 0.696</em>, <em class="sig-param">'PI': 0.363</em>, <em class="sig-param">'PK': 0.875</em>, <em class="sig-param">'PL': 0.357</em>, <em class="sig-param">'PM': 0.326</em>, <em class="sig-param">'PN': 0.231</em>, <em class="sig-param">'PP': 0.0</em>, <em class="sig-param">'PQ': 0.228</em>, <em class="sig-param">'PR': 1.0</em>, <em class="sig-param">'PS': 0.196</em>, <em class="sig-param">'PT': 0.161</em>, <em class="sig-param">'PV': 0.244</em>, <em class="sig-param">'PW': 0.72</em>, <em class="sig-param">'PY': 0.481</em>, <em class="sig-param">'QA': 0.512</em>, <em class="sig-param">'QC': 0.462</em>, <em class="sig-param">'QD': 0.903</em>, <em class="sig-param">'QE': 0.861</em>, <em class="sig-param">'QF': 0.671</em>, <em class="sig-param">'QG': 0.648</em>, <em class="sig-param">'QH': 0.765</em>, <em class="sig-param">'QI': 0.532</em>, <em class="sig-param">'QK': 0.881</em>, <em class="sig-param">'QL': 0.518</em>, <em class="sig-param">'QM': 0.505</em>, <em class="sig-param">'QN': 0.181</em>, <em class="sig-param">'QP': 0.272</em>, <em class="sig-param">'QQ': 0.0</em>, <em class="sig-param">'QR': 1.0</em>, <em class="sig-param">'QS': 0.461</em>, <em class="sig-param">'QT': 0.389</em>, <em class="sig-param">'QV': 0.464</em>, <em class="sig-param">'QW': 0.831</em>, <em class="sig-param">'QY': 0.522</em>, <em class="sig-param">'RA': 0.919</em>, <em class="sig-param">'RC': 0.905</em>, <em class="sig-param">'RD': 0.305</em>, <em class="sig-param">'RE': 0.225</em>, <em class="sig-param">'RF': 0.977</em>, <em class="sig-param">'RG': 0.928</em>, <em class="sig-param">'RH': 0.498</em>, <em class="sig-param">'RI': 0.929</em>, <em class="sig-param">'RK': 0.141</em>, <em class="sig-param">'RL': 0.92</em>, <em class="sig-param">'RM': 0.908</em>, <em class="sig-param">'RN': 0.69</em>, <em class="sig-param">'RP': 0.796</em>, <em class="sig-param">'RQ': 0.668</em>, <em class="sig-param">'RR': 0.0</em>, <em class="sig-param">'RS': 0.86</em>, <em class="sig-param">'RT': 0.808</em>, <em class="sig-param">'RV': 0.914</em>, <em class="sig-param">'RW': 1.0</em>, <em class="sig-param">'RY': 0.859</em>, <em class="sig-param">'SA': 0.1</em>, <em class="sig-param">'SC': 0.185</em>, <em class="sig-param">'SD': 0.801</em>, <em class="sig-param">'SE': 0.812</em>, <em class="sig-param">'SF': 0.622</em>, <em class="sig-param">'SG': 0.17</em>, <em class="sig-param">'SH': 0.718</em>, <em class="sig-param">'SI': 0.478</em>, <em class="sig-param">'SK': 0.883</em>, <em class="sig-param">'SL': 0.474</em>, <em class="sig-param">'SM': 0.44</em>, <em class="sig-param">'SN': 0.289</em>, <em class="sig-param">'SP': 0.181</em>, <em class="sig-param">'SQ': 0.358</em>, <em class="sig-param">'SR': 1.0</em>, <em class="sig-param">'SS': 0.0</em>, <em class="sig-param">'ST': 0.174</em>, <em class="sig-param">'SV': 0.342</em>, <em class="sig-param">'SW': 0.827</em>, <em class="sig-param">'SY': 0.615</em>, <em class="sig-param">'TA': 0.251</em>, <em class="sig-param">'TC': 0.261</em>, <em class="sig-param">'TD': 0.83</em>, <em class="sig-param">'TE': 0.812</em>, <em class="sig-param">'TF': 0.604</em>, <em class="sig-param">'TG': 0.312</em>, <em class="sig-param">'TH': 0.737</em>, <em class="sig-param">'TI': 0.455</em>, <em class="sig-param">'TK': 0.866</em>, <em class="sig-param">'TL': 0.453</em>, <em class="sig-param">'TM': 0.403</em>, <em class="sig-param">'TN': 0.315</em>, <em class="sig-param">'TP': 0.159</em>, <em class="sig-param">'TQ': 0.322</em>, <em class="sig-param">'TR': 1.0</em>, <em class="sig-param">'TS': 0.185</em>, <em class="sig-param">'TT': 0.0</em>, <em class="sig-param">'TV': 0.345</em>, <em class="sig-param">'TW': 0.816</em>, <em class="sig-param">'TY': 0.596</em>, <em class="sig-param">'VA': 0.275</em>, <em class="sig-param">'VC': 0.165</em>, <em class="sig-param">'VD': 0.9</em>, <em class="sig-param">'VE': 0.867</em>, <em class="sig-param">'VF': 0.269</em>, <em class="sig-param">'VG': 0.471</em>, <em class="sig-param">'VH': 0.649</em>, <em class="sig-param">'VI': 0.135</em>, <em class="sig-param">'VK': 0.889</em>, <em class="sig-param">'VL': 0.134</em>, <em class="sig-param">'VM': 0.12</em>, <em class="sig-param">'VN': 0.38</em>, <em class="sig-param">'VP': 0.212</em>, <em class="sig-param">'VQ': 0.339</em>, <em class="sig-param">'VR': 1.0</em>, <em class="sig-param">'VS': 0.322</em>, <em class="sig-param">'VT': 0.305</em>, <em class="sig-param">'VV': 0.0</em>, <em class="sig-param">'VW': 0.472</em>, <em class="sig-param">'VY': 0.31</em>, <em class="sig-param">'WA': 0.658</em>, <em class="sig-param">'WC': 0.56</em>, <em class="sig-param">'WD': 1.0</em>, <em class="sig-param">'WE': 0.931</em>, <em class="sig-param">'WF': 0.196</em>, <em class="sig-param">'WG': 0.829</em>, <em class="sig-param">'WH': 0.678</em>, <em class="sig-param">'WI': 0.305</em>, <em class="sig-param">'WK': 0.892</em>, <em class="sig-param">'WL': 0.304</em>, <em class="sig-param">'WM': 0.344</em>, <em class="sig-param">'WN': 0.631</em>, <em class="sig-param">'WP': 0.555</em>, <em class="sig-param">'WQ': 0.538</em>, <em class="sig-param">'WR': 0.968</em>, <em class="sig-param">'WS': 0.689</em>, <em class="sig-param">'WT': 0.638</em>, <em class="sig-param">'WV': 0.418</em>, <em class="sig-param">'WW': 0.0</em>, <em class="sig-param">'WY': 0.204</em>, <em class="sig-param">'YA': 0.587</em>, <em class="sig-param">'YC': 0.478</em>, <em class="sig-param">'YD': 1.0</em>, <em class="sig-param">'YE': 0.932</em>, <em class="sig-param">'YF': 0.202</em>, <em class="sig-param">'YG': 0.782</em>, <em class="sig-param">'YH': 0.678</em>, <em class="sig-param">'YI': 0.23</em>, <em class="sig-param">'YK': 0.904</em>, <em class="sig-param">'YL': 0.219</em>, <em class="sig-param">'YM': 0.268</em>, <em class="sig-param">'YN': 0.512</em>, <em class="sig-param">'YP': 0.444</em>, <em class="sig-param">'YQ': 0.404</em>, <em class="sig-param">'YR': 0.995</em>, <em class="sig-param">'YS': 0.612</em>, <em class="sig-param">'YT': 0.557</em>, <em class="sig-param">'YV': 0.328</em>, <em class="sig-param">'YW': 0.244</em>, <em class="sig-param">'YY': 0.0}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_sequence_order_coupling_number"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_sequence_order_coupling_number" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the dth-rank sequence order coupling number for a protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>d</strong> – d is the gap between two amino acids</p></li>
<li><p><strong>distancematrix</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>result is numeric value</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_sequence_order_coupling_number_grant">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_sequence_order_coupling_number_grant</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">distancematrix={'AA': 0.0</em>, <em class="sig-param">'AC': 0.112</em>, <em class="sig-param">'AD': 0.819</em>, <em class="sig-param">'AE': 0.827</em>, <em class="sig-param">'AF': 0.54</em>, <em class="sig-param">'AG': 0.208</em>, <em class="sig-param">'AH': 0.696</em>, <em class="sig-param">'AI': 0.407</em>, <em class="sig-param">'AK': 0.891</em>, <em class="sig-param">'AL': 0.406</em>, <em class="sig-param">'AM': 0.379</em>, <em class="sig-param">'AN': 0.318</em>, <em class="sig-param">'AP': 0.191</em>, <em class="sig-param">'AQ': 0.372</em>, <em class="sig-param">'AR': 1.0</em>, <em class="sig-param">'AS': 0.094</em>, <em class="sig-param">'AT': 0.22</em>, <em class="sig-param">'AV': 0.273</em>, <em class="sig-param">'AW': 0.739</em>, <em class="sig-param">'AY': 0.552</em>, <em class="sig-param">'CA': 0.114</em>, <em class="sig-param">'CC': 0.0</em>, <em class="sig-param">'CD': 0.847</em>, <em class="sig-param">'CE': 0.838</em>, <em class="sig-param">'CF': 0.437</em>, <em class="sig-param">'CG': 0.32</em>, <em class="sig-param">'CH': 0.66</em>, <em class="sig-param">'CI': 0.304</em>, <em class="sig-param">'CK': 0.887</em>, <em class="sig-param">'CL': 0.301</em>, <em class="sig-param">'CM': 0.277</em>, <em class="sig-param">'CN': 0.324</em>, <em class="sig-param">'CP': 0.157</em>, <em class="sig-param">'CQ': 0.341</em>, <em class="sig-param">'CR': 1.0</em>, <em class="sig-param">'CS': 0.176</em>, <em class="sig-param">'CT': 0.233</em>, <em class="sig-param">'CV': 0.167</em>, <em class="sig-param">'CW': 0.639</em>, <em class="sig-param">'CY': 0.457</em>, <em class="sig-param">'DA': 0.729</em>, <em class="sig-param">'DC': 0.742</em>, <em class="sig-param">'DD': 0.0</em>, <em class="sig-param">'DE': 0.124</em>, <em class="sig-param">'DF': 0.924</em>, <em class="sig-param">'DG': 0.697</em>, <em class="sig-param">'DH': 0.435</em>, <em class="sig-param">'DI': 0.847</em>, <em class="sig-param">'DK': 0.249</em>, <em class="sig-param">'DL': 0.841</em>, <em class="sig-param">'DM': 0.819</em>, <em class="sig-param">'DN': 0.56</em>, <em class="sig-param">'DP': 0.657</em>, <em class="sig-param">'DQ': 0.584</em>, <em class="sig-param">'DR': 0.295</em>, <em class="sig-param">'DS': 0.667</em>, <em class="sig-param">'DT': 0.649</em>, <em class="sig-param">'DV': 0.797</em>, <em class="sig-param">'DW': 1.0</em>, <em class="sig-param">'DY': 0.836</em>, <em class="sig-param">'EA': 0.79</em>, <em class="sig-param">'EC': 0.788</em>, <em class="sig-param">'ED': 0.133</em>, <em class="sig-param">'EE': 0.0</em>, <em class="sig-param">'EF': 0.932</em>, <em class="sig-param">'EG': 0.779</em>, <em class="sig-param">'EH': 0.406</em>, <em class="sig-param">'EI': 0.86</em>, <em class="sig-param">'EK': 0.143</em>, <em class="sig-param">'EL': 0.854</em>, <em class="sig-param">'EM': 0.83</em>, <em class="sig-param">'EN': 0.599</em>, <em class="sig-param">'EP': 0.688</em>, <em class="sig-param">'EQ': 0.598</em>, <em class="sig-param">'ER': 0.234</em>, <em class="sig-param">'ES': 0.726</em>, <em class="sig-param">'ET': 0.682</em>, <em class="sig-param">'EV': 0.824</em>, <em class="sig-param">'EW': 1.0</em>, <em class="sig-param">'EY': 0.837</em>, <em class="sig-param">'FA': 0.508</em>, <em class="sig-param">'FC': 0.405</em>, <em class="sig-param">'FD': 0.977</em>, <em class="sig-param">'FE': 0.918</em>, <em class="sig-param">'FF': 0.0</em>, <em class="sig-param">'FG': 0.69</em>, <em class="sig-param">'FH': 0.663</em>, <em class="sig-param">'FI': 0.128</em>, <em class="sig-param">'FK': 0.903</em>, <em class="sig-param">'FL': 0.131</em>, <em class="sig-param">'FM': 0.169</em>, <em class="sig-param">'FN': 0.541</em>, <em class="sig-param">'FP': 0.42</em>, <em class="sig-param">'FQ': 0.459</em>, <em class="sig-param">'FR': 1.0</em>, <em class="sig-param">'FS': 0.548</em>, <em class="sig-param">'FT': 0.499</em>, <em class="sig-param">'FV': 0.252</em>, <em class="sig-param">'FW': 0.207</em>, <em class="sig-param">'FY': 0.179</em>, <em class="sig-param">'GA': 0.206</em>, <em class="sig-param">'GC': 0.312</em>, <em class="sig-param">'GD': 0.776</em>, <em class="sig-param">'GE': 0.807</em>, <em class="sig-param">'GF': 0.727</em>, <em class="sig-param">'GG': 0.0</em>, <em class="sig-param">'GH': 0.769</em>, <em class="sig-param">'GI': 0.592</em>, <em class="sig-param">'GK': 0.894</em>, <em class="sig-param">'GL': 0.591</em>, <em class="sig-param">'GM': 0.557</em>, <em class="sig-param">'GN': 0.381</em>, <em class="sig-param">'GP': 0.323</em>, <em class="sig-param">'GQ': 0.467</em>, <em class="sig-param">'GR': 1.0</em>, <em class="sig-param">'GS': 0.158</em>, <em class="sig-param">'GT': 0.272</em>, <em class="sig-param">'GV': 0.464</em>, <em class="sig-param">'GW': 0.923</em>, <em class="sig-param">'GY': 0.728</em>, <em class="sig-param">'HA': 0.896</em>, <em class="sig-param">'HC': 0.836</em>, <em class="sig-param">'HD': 0.629</em>, <em class="sig-param">'HE': 0.547</em>, <em class="sig-param">'HF': 0.907</em>, <em class="sig-param">'HG': 1.0</em>, <em class="sig-param">'HH': 0.0</em>, <em class="sig-param">'HI': 0.848</em>, <em class="sig-param">'HK': 0.566</em>, <em class="sig-param">'HL': 0.842</em>, <em class="sig-param">'HM': 0.825</em>, <em class="sig-param">'HN': 0.754</em>, <em class="sig-param">'HP': 0.777</em>, <em class="sig-param">'HQ': 0.716</em>, <em class="sig-param">'HR': 0.697</em>, <em class="sig-param">'HS': 0.865</em>, <em class="sig-param">'HT': 0.834</em>, <em class="sig-param">'HV': 0.831</em>, <em class="sig-param">'HW': 0.981</em>, <em class="sig-param">'HY': 0.821</em>, <em class="sig-param">'IA': 0.403</em>, <em class="sig-param">'IC': 0.296</em>, <em class="sig-param">'ID': 0.942</em>, <em class="sig-param">'IE': 0.891</em>, <em class="sig-param">'IF': 0.134</em>, <em class="sig-param">'IG': 0.592</em>, <em class="sig-param">'IH': 0.652</em>, <em class="sig-param">'II': 0.0</em>, <em class="sig-param">'IK': 0.892</em>, <em class="sig-param">'IL': 0.013</em>, <em class="sig-param">'IM': 0.057</em>, <em class="sig-param">'IN': 0.457</em>, <em class="sig-param">'IP': 0.311</em>, <em class="sig-param">'IQ': 0.383</em>, <em class="sig-param">'IR': 1.0</em>, <em class="sig-param">'IS': 0.443</em>, <em class="sig-param">'IT': 0.396</em>, <em class="sig-param">'IV': 0.133</em>, <em class="sig-param">'IW': 0.339</em>, <em class="sig-param">'IY': 0.213</em>, <em class="sig-param">'KA': 0.889</em>, <em class="sig-param">'KC': 0.871</em>, <em class="sig-param">'KD': 0.279</em>, <em class="sig-param">'KE': 0.149</em>, <em class="sig-param">'KF': 0.957</em>, <em class="sig-param">'KG': 0.9</em>, <em class="sig-param">'KH': 0.438</em>, <em class="sig-param">'KI': 0.899</em>, <em class="sig-param">'KK': 0.0</em>, <em class="sig-param">'KL': 0.892</em>, <em class="sig-param">'KM': 0.871</em>, <em class="sig-param">'KN': 0.667</em>, <em class="sig-param">'KP': 0.757</em>, <em class="sig-param">'KQ': 0.639</em>, <em class="sig-param">'KR': 0.154</em>, <em class="sig-param">'KS': 0.825</em>, <em class="sig-param">'KT': 0.759</em>, <em class="sig-param">'KV': 0.882</em>, <em class="sig-param">'KW': 1.0</em>, <em class="sig-param">'KY': 0.848</em>, <em class="sig-param">'LA': 0.405</em>, <em class="sig-param">'LC': 0.296</em>, <em class="sig-param">'LD': 0.944</em>, <em class="sig-param">'LE': 0.892</em>, <em class="sig-param">'LF': 0.139</em>, <em class="sig-param">'LG': 0.596</em>, <em class="sig-param">'LH': 0.653</em>, <em class="sig-param">'LI': 0.013</em>, <em class="sig-param">'LK': 0.893</em>, <em class="sig-param">'LL': 0.0</em>, <em class="sig-param">'LM': 0.062</em>, <em class="sig-param">'LN': 0.452</em>, <em class="sig-param">'LP': 0.309</em>, <em class="sig-param">'LQ': 0.376</em>, <em class="sig-param">'LR': 1.0</em>, <em class="sig-param">'LS': 0.443</em>, <em class="sig-param">'LT': 0.397</em>, <em class="sig-param">'LV': 0.133</em>, <em class="sig-param">'LW': 0.341</em>, <em class="sig-param">'LY': 0.205</em>, <em class="sig-param">'MA': 0.383</em>, <em class="sig-param">'MC': 0.276</em>, <em class="sig-param">'MD': 0.932</em>, <em class="sig-param">'ME': 0.879</em>, <em class="sig-param">'MF': 0.182</em>, <em class="sig-param">'MG': 0.569</em>, <em class="sig-param">'MH': 0.648</em>, <em class="sig-param">'MI': 0.058</em>, <em class="sig-param">'MK': 0.884</em>, <em class="sig-param">'ML': 0.062</em>, <em class="sig-param">'MM': 0.0</em>, <em class="sig-param">'MN': 0.447</em>, <em class="sig-param">'MP': 0.285</em>, <em class="sig-param">'MQ': 0.372</em>, <em class="sig-param">'MR': 1.0</em>, <em class="sig-param">'MS': 0.417</em>, <em class="sig-param">'MT': 0.358</em>, <em class="sig-param">'MV': 0.12</em>, <em class="sig-param">'MW': 0.391</em>, <em class="sig-param">'MY': 0.255</em>, <em class="sig-param">'NA': 0.424</em>, <em class="sig-param">'NC': 0.425</em>, <em class="sig-param">'ND': 0.838</em>, <em class="sig-param">'NE': 0.835</em>, <em class="sig-param">'NF': 0.766</em>, <em class="sig-param">'NG': 0.512</em>, <em class="sig-param">'NH': 0.78</em>, <em class="sig-param">'NI': 0.615</em>, <em class="sig-param">'NK': 0.891</em>, <em class="sig-param">'NL': 0.603</em>, <em class="sig-param">'NM': 0.588</em>, <em class="sig-param">'NN': 0.0</em>, <em class="sig-param">'NP': 0.266</em>, <em class="sig-param">'NQ': 0.175</em>, <em class="sig-param">'NR': 1.0</em>, <em class="sig-param">'NS': 0.361</em>, <em class="sig-param">'NT': 0.368</em>, <em class="sig-param">'NV': 0.503</em>, <em class="sig-param">'NW': 0.945</em>, <em class="sig-param">'NY': 0.641</em>, <em class="sig-param">'PA': 0.22</em>, <em class="sig-param">'PC': 0.179</em>, <em class="sig-param">'PD': 0.852</em>, <em class="sig-param">'PE': 0.831</em>, <em class="sig-param">'PF': 0.515</em>, <em class="sig-param">'PG': 0.376</em>, <em class="sig-param">'PH': 0.696</em>, <em class="sig-param">'PI': 0.363</em>, <em class="sig-param">'PK': 0.875</em>, <em class="sig-param">'PL': 0.357</em>, <em class="sig-param">'PM': 0.326</em>, <em class="sig-param">'PN': 0.231</em>, <em class="sig-param">'PP': 0.0</em>, <em class="sig-param">'PQ': 0.228</em>, <em class="sig-param">'PR': 1.0</em>, <em class="sig-param">'PS': 0.196</em>, <em class="sig-param">'PT': 0.161</em>, <em class="sig-param">'PV': 0.244</em>, <em class="sig-param">'PW': 0.72</em>, <em class="sig-param">'PY': 0.481</em>, <em class="sig-param">'QA': 0.512</em>, <em class="sig-param">'QC': 0.462</em>, <em class="sig-param">'QD': 0.903</em>, <em class="sig-param">'QE': 0.861</em>, <em class="sig-param">'QF': 0.671</em>, <em class="sig-param">'QG': 0.648</em>, <em class="sig-param">'QH': 0.765</em>, <em class="sig-param">'QI': 0.532</em>, <em class="sig-param">'QK': 0.881</em>, <em class="sig-param">'QL': 0.518</em>, <em class="sig-param">'QM': 0.505</em>, <em class="sig-param">'QN': 0.181</em>, <em class="sig-param">'QP': 0.272</em>, <em class="sig-param">'QQ': 0.0</em>, <em class="sig-param">'QR': 1.0</em>, <em class="sig-param">'QS': 0.461</em>, <em class="sig-param">'QT': 0.389</em>, <em class="sig-param">'QV': 0.464</em>, <em class="sig-param">'QW': 0.831</em>, <em class="sig-param">'QY': 0.522</em>, <em class="sig-param">'RA': 0.919</em>, <em class="sig-param">'RC': 0.905</em>, <em class="sig-param">'RD': 0.305</em>, <em class="sig-param">'RE': 0.225</em>, <em class="sig-param">'RF': 0.977</em>, <em class="sig-param">'RG': 0.928</em>, <em class="sig-param">'RH': 0.498</em>, <em class="sig-param">'RI': 0.929</em>, <em class="sig-param">'RK': 0.141</em>, <em class="sig-param">'RL': 0.92</em>, <em class="sig-param">'RM': 0.908</em>, <em class="sig-param">'RN': 0.69</em>, <em class="sig-param">'RP': 0.796</em>, <em class="sig-param">'RQ': 0.668</em>, <em class="sig-param">'RR': 0.0</em>, <em class="sig-param">'RS': 0.86</em>, <em class="sig-param">'RT': 0.808</em>, <em class="sig-param">'RV': 0.914</em>, <em class="sig-param">'RW': 1.0</em>, <em class="sig-param">'RY': 0.859</em>, <em class="sig-param">'SA': 0.1</em>, <em class="sig-param">'SC': 0.185</em>, <em class="sig-param">'SD': 0.801</em>, <em class="sig-param">'SE': 0.812</em>, <em class="sig-param">'SF': 0.622</em>, <em class="sig-param">'SG': 0.17</em>, <em class="sig-param">'SH': 0.718</em>, <em class="sig-param">'SI': 0.478</em>, <em class="sig-param">'SK': 0.883</em>, <em class="sig-param">'SL': 0.474</em>, <em class="sig-param">'SM': 0.44</em>, <em class="sig-param">'SN': 0.289</em>, <em class="sig-param">'SP': 0.181</em>, <em class="sig-param">'SQ': 0.358</em>, <em class="sig-param">'SR': 1.0</em>, <em class="sig-param">'SS': 0.0</em>, <em class="sig-param">'ST': 0.174</em>, <em class="sig-param">'SV': 0.342</em>, <em class="sig-param">'SW': 0.827</em>, <em class="sig-param">'SY': 0.615</em>, <em class="sig-param">'TA': 0.251</em>, <em class="sig-param">'TC': 0.261</em>, <em class="sig-param">'TD': 0.83</em>, <em class="sig-param">'TE': 0.812</em>, <em class="sig-param">'TF': 0.604</em>, <em class="sig-param">'TG': 0.312</em>, <em class="sig-param">'TH': 0.737</em>, <em class="sig-param">'TI': 0.455</em>, <em class="sig-param">'TK': 0.866</em>, <em class="sig-param">'TL': 0.453</em>, <em class="sig-param">'TM': 0.403</em>, <em class="sig-param">'TN': 0.315</em>, <em class="sig-param">'TP': 0.159</em>, <em class="sig-param">'TQ': 0.322</em>, <em class="sig-param">'TR': 1.0</em>, <em class="sig-param">'TS': 0.185</em>, <em class="sig-param">'TT': 0.0</em>, <em class="sig-param">'TV': 0.345</em>, <em class="sig-param">'TW': 0.816</em>, <em class="sig-param">'TY': 0.596</em>, <em class="sig-param">'VA': 0.275</em>, <em class="sig-param">'VC': 0.165</em>, <em class="sig-param">'VD': 0.9</em>, <em class="sig-param">'VE': 0.867</em>, <em class="sig-param">'VF': 0.269</em>, <em class="sig-param">'VG': 0.471</em>, <em class="sig-param">'VH': 0.649</em>, <em class="sig-param">'VI': 0.135</em>, <em class="sig-param">'VK': 0.889</em>, <em class="sig-param">'VL': 0.134</em>, <em class="sig-param">'VM': 0.12</em>, <em class="sig-param">'VN': 0.38</em>, <em class="sig-param">'VP': 0.212</em>, <em class="sig-param">'VQ': 0.339</em>, <em class="sig-param">'VR': 1.0</em>, <em class="sig-param">'VS': 0.322</em>, <em class="sig-param">'VT': 0.305</em>, <em class="sig-param">'VV': 0.0</em>, <em class="sig-param">'VW': 0.472</em>, <em class="sig-param">'VY': 0.31</em>, <em class="sig-param">'WA': 0.658</em>, <em class="sig-param">'WC': 0.56</em>, <em class="sig-param">'WD': 1.0</em>, <em class="sig-param">'WE': 0.931</em>, <em class="sig-param">'WF': 0.196</em>, <em class="sig-param">'WG': 0.829</em>, <em class="sig-param">'WH': 0.678</em>, <em class="sig-param">'WI': 0.305</em>, <em class="sig-param">'WK': 0.892</em>, <em class="sig-param">'WL': 0.304</em>, <em class="sig-param">'WM': 0.344</em>, <em class="sig-param">'WN': 0.631</em>, <em class="sig-param">'WP': 0.555</em>, <em class="sig-param">'WQ': 0.538</em>, <em class="sig-param">'WR': 0.968</em>, <em class="sig-param">'WS': 0.689</em>, <em class="sig-param">'WT': 0.638</em>, <em class="sig-param">'WV': 0.418</em>, <em class="sig-param">'WW': 0.0</em>, <em class="sig-param">'WY': 0.204</em>, <em class="sig-param">'YA': 0.587</em>, <em class="sig-param">'YC': 0.478</em>, <em class="sig-param">'YD': 1.0</em>, <em class="sig-param">'YE': 0.932</em>, <em class="sig-param">'YF': 0.202</em>, <em class="sig-param">'YG': 0.782</em>, <em class="sig-param">'YH': 0.678</em>, <em class="sig-param">'YI': 0.23</em>, <em class="sig-param">'YK': 0.904</em>, <em class="sig-param">'YL': 0.219</em>, <em class="sig-param">'YM': 0.268</em>, <em class="sig-param">'YN': 0.512</em>, <em class="sig-param">'YP': 0.444</em>, <em class="sig-param">'YQ': 0.404</em>, <em class="sig-param">'YR': 0.995</em>, <em class="sig-param">'YS': 0.612</em>, <em class="sig-param">'YT': 0.557</em>, <em class="sig-param">'YV': 0.328</em>, <em class="sig-param">'YW': 0.244</em>, <em class="sig-param">'YY': 0.0}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_sequence_order_coupling_number_grant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_sequence_order_coupling_number_grant" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the sequence order coupling numbers from 1 to maxlag for a given protein sequence based on the
Grantham chemical distance      matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>maxlag</strong> – maxlag is the maximum lag and the length of the protein should be larger than maxlag. default is 30.</p></li>
<li><p><strong>distancematrix</strong> – distancematrix is a dict form containing Grantham chemical distance      matrix. omitted!</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing all sequence order coupling number        based on the Grantham chemical distance matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_sequence_order_coupling_number_sw">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_sequence_order_coupling_number_sw</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">distancematrix={'AA': 0.0</em>, <em class="sig-param">'AC': 0.112</em>, <em class="sig-param">'AD': 0.819</em>, <em class="sig-param">'AE': 0.827</em>, <em class="sig-param">'AF': 0.54</em>, <em class="sig-param">'AG': 0.208</em>, <em class="sig-param">'AH': 0.696</em>, <em class="sig-param">'AI': 0.407</em>, <em class="sig-param">'AK': 0.891</em>, <em class="sig-param">'AL': 0.406</em>, <em class="sig-param">'AM': 0.379</em>, <em class="sig-param">'AN': 0.318</em>, <em class="sig-param">'AP': 0.191</em>, <em class="sig-param">'AQ': 0.372</em>, <em class="sig-param">'AR': 1.0</em>, <em class="sig-param">'AS': 0.094</em>, <em class="sig-param">'AT': 0.22</em>, <em class="sig-param">'AV': 0.273</em>, <em class="sig-param">'AW': 0.739</em>, <em class="sig-param">'AY': 0.552</em>, <em class="sig-param">'CA': 0.114</em>, <em class="sig-param">'CC': 0.0</em>, <em class="sig-param">'CD': 0.847</em>, <em class="sig-param">'CE': 0.838</em>, <em class="sig-param">'CF': 0.437</em>, <em class="sig-param">'CG': 0.32</em>, <em class="sig-param">'CH': 0.66</em>, <em class="sig-param">'CI': 0.304</em>, <em class="sig-param">'CK': 0.887</em>, <em class="sig-param">'CL': 0.301</em>, <em class="sig-param">'CM': 0.277</em>, <em class="sig-param">'CN': 0.324</em>, <em class="sig-param">'CP': 0.157</em>, <em class="sig-param">'CQ': 0.341</em>, <em class="sig-param">'CR': 1.0</em>, <em class="sig-param">'CS': 0.176</em>, <em class="sig-param">'CT': 0.233</em>, <em class="sig-param">'CV': 0.167</em>, <em class="sig-param">'CW': 0.639</em>, <em class="sig-param">'CY': 0.457</em>, <em class="sig-param">'DA': 0.729</em>, <em class="sig-param">'DC': 0.742</em>, <em class="sig-param">'DD': 0.0</em>, <em class="sig-param">'DE': 0.124</em>, <em class="sig-param">'DF': 0.924</em>, <em class="sig-param">'DG': 0.697</em>, <em class="sig-param">'DH': 0.435</em>, <em class="sig-param">'DI': 0.847</em>, <em class="sig-param">'DK': 0.249</em>, <em class="sig-param">'DL': 0.841</em>, <em class="sig-param">'DM': 0.819</em>, <em class="sig-param">'DN': 0.56</em>, <em class="sig-param">'DP': 0.657</em>, <em class="sig-param">'DQ': 0.584</em>, <em class="sig-param">'DR': 0.295</em>, <em class="sig-param">'DS': 0.667</em>, <em class="sig-param">'DT': 0.649</em>, <em class="sig-param">'DV': 0.797</em>, <em class="sig-param">'DW': 1.0</em>, <em class="sig-param">'DY': 0.836</em>, <em class="sig-param">'EA': 0.79</em>, <em class="sig-param">'EC': 0.788</em>, <em class="sig-param">'ED': 0.133</em>, <em class="sig-param">'EE': 0.0</em>, <em class="sig-param">'EF': 0.932</em>, <em class="sig-param">'EG': 0.779</em>, <em class="sig-param">'EH': 0.406</em>, <em class="sig-param">'EI': 0.86</em>, <em class="sig-param">'EK': 0.143</em>, <em class="sig-param">'EL': 0.854</em>, <em class="sig-param">'EM': 0.83</em>, <em class="sig-param">'EN': 0.599</em>, <em class="sig-param">'EP': 0.688</em>, <em class="sig-param">'EQ': 0.598</em>, <em class="sig-param">'ER': 0.234</em>, <em class="sig-param">'ES': 0.726</em>, <em class="sig-param">'ET': 0.682</em>, <em class="sig-param">'EV': 0.824</em>, <em class="sig-param">'EW': 1.0</em>, <em class="sig-param">'EY': 0.837</em>, <em class="sig-param">'FA': 0.508</em>, <em class="sig-param">'FC': 0.405</em>, <em class="sig-param">'FD': 0.977</em>, <em class="sig-param">'FE': 0.918</em>, <em class="sig-param">'FF': 0.0</em>, <em class="sig-param">'FG': 0.69</em>, <em class="sig-param">'FH': 0.663</em>, <em class="sig-param">'FI': 0.128</em>, <em class="sig-param">'FK': 0.903</em>, <em class="sig-param">'FL': 0.131</em>, <em class="sig-param">'FM': 0.169</em>, <em class="sig-param">'FN': 0.541</em>, <em class="sig-param">'FP': 0.42</em>, <em class="sig-param">'FQ': 0.459</em>, <em class="sig-param">'FR': 1.0</em>, <em class="sig-param">'FS': 0.548</em>, <em class="sig-param">'FT': 0.499</em>, <em class="sig-param">'FV': 0.252</em>, <em class="sig-param">'FW': 0.207</em>, <em class="sig-param">'FY': 0.179</em>, <em class="sig-param">'GA': 0.206</em>, <em class="sig-param">'GC': 0.312</em>, <em class="sig-param">'GD': 0.776</em>, <em class="sig-param">'GE': 0.807</em>, <em class="sig-param">'GF': 0.727</em>, <em class="sig-param">'GG': 0.0</em>, <em class="sig-param">'GH': 0.769</em>, <em class="sig-param">'GI': 0.592</em>, <em class="sig-param">'GK': 0.894</em>, <em class="sig-param">'GL': 0.591</em>, <em class="sig-param">'GM': 0.557</em>, <em class="sig-param">'GN': 0.381</em>, <em class="sig-param">'GP': 0.323</em>, <em class="sig-param">'GQ': 0.467</em>, <em class="sig-param">'GR': 1.0</em>, <em class="sig-param">'GS': 0.158</em>, <em class="sig-param">'GT': 0.272</em>, <em class="sig-param">'GV': 0.464</em>, <em class="sig-param">'GW': 0.923</em>, <em class="sig-param">'GY': 0.728</em>, <em class="sig-param">'HA': 0.896</em>, <em class="sig-param">'HC': 0.836</em>, <em class="sig-param">'HD': 0.629</em>, <em class="sig-param">'HE': 0.547</em>, <em class="sig-param">'HF': 0.907</em>, <em class="sig-param">'HG': 1.0</em>, <em class="sig-param">'HH': 0.0</em>, <em class="sig-param">'HI': 0.848</em>, <em class="sig-param">'HK': 0.566</em>, <em class="sig-param">'HL': 0.842</em>, <em class="sig-param">'HM': 0.825</em>, <em class="sig-param">'HN': 0.754</em>, <em class="sig-param">'HP': 0.777</em>, <em class="sig-param">'HQ': 0.716</em>, <em class="sig-param">'HR': 0.697</em>, <em class="sig-param">'HS': 0.865</em>, <em class="sig-param">'HT': 0.834</em>, <em class="sig-param">'HV': 0.831</em>, <em class="sig-param">'HW': 0.981</em>, <em class="sig-param">'HY': 0.821</em>, <em class="sig-param">'IA': 0.403</em>, <em class="sig-param">'IC': 0.296</em>, <em class="sig-param">'ID': 0.942</em>, <em class="sig-param">'IE': 0.891</em>, <em class="sig-param">'IF': 0.134</em>, <em class="sig-param">'IG': 0.592</em>, <em class="sig-param">'IH': 0.652</em>, <em class="sig-param">'II': 0.0</em>, <em class="sig-param">'IK': 0.892</em>, <em class="sig-param">'IL': 0.013</em>, <em class="sig-param">'IM': 0.057</em>, <em class="sig-param">'IN': 0.457</em>, <em class="sig-param">'IP': 0.311</em>, <em class="sig-param">'IQ': 0.383</em>, <em class="sig-param">'IR': 1.0</em>, <em class="sig-param">'IS': 0.443</em>, <em class="sig-param">'IT': 0.396</em>, <em class="sig-param">'IV': 0.133</em>, <em class="sig-param">'IW': 0.339</em>, <em class="sig-param">'IY': 0.213</em>, <em class="sig-param">'KA': 0.889</em>, <em class="sig-param">'KC': 0.871</em>, <em class="sig-param">'KD': 0.279</em>, <em class="sig-param">'KE': 0.149</em>, <em class="sig-param">'KF': 0.957</em>, <em class="sig-param">'KG': 0.9</em>, <em class="sig-param">'KH': 0.438</em>, <em class="sig-param">'KI': 0.899</em>, <em class="sig-param">'KK': 0.0</em>, <em class="sig-param">'KL': 0.892</em>, <em class="sig-param">'KM': 0.871</em>, <em class="sig-param">'KN': 0.667</em>, <em class="sig-param">'KP': 0.757</em>, <em class="sig-param">'KQ': 0.639</em>, <em class="sig-param">'KR': 0.154</em>, <em class="sig-param">'KS': 0.825</em>, <em class="sig-param">'KT': 0.759</em>, <em class="sig-param">'KV': 0.882</em>, <em class="sig-param">'KW': 1.0</em>, <em class="sig-param">'KY': 0.848</em>, <em class="sig-param">'LA': 0.405</em>, <em class="sig-param">'LC': 0.296</em>, <em class="sig-param">'LD': 0.944</em>, <em class="sig-param">'LE': 0.892</em>, <em class="sig-param">'LF': 0.139</em>, <em class="sig-param">'LG': 0.596</em>, <em class="sig-param">'LH': 0.653</em>, <em class="sig-param">'LI': 0.013</em>, <em class="sig-param">'LK': 0.893</em>, <em class="sig-param">'LL': 0.0</em>, <em class="sig-param">'LM': 0.062</em>, <em class="sig-param">'LN': 0.452</em>, <em class="sig-param">'LP': 0.309</em>, <em class="sig-param">'LQ': 0.376</em>, <em class="sig-param">'LR': 1.0</em>, <em class="sig-param">'LS': 0.443</em>, <em class="sig-param">'LT': 0.397</em>, <em class="sig-param">'LV': 0.133</em>, <em class="sig-param">'LW': 0.341</em>, <em class="sig-param">'LY': 0.205</em>, <em class="sig-param">'MA': 0.383</em>, <em class="sig-param">'MC': 0.276</em>, <em class="sig-param">'MD': 0.932</em>, <em class="sig-param">'ME': 0.879</em>, <em class="sig-param">'MF': 0.182</em>, <em class="sig-param">'MG': 0.569</em>, <em class="sig-param">'MH': 0.648</em>, <em class="sig-param">'MI': 0.058</em>, <em class="sig-param">'MK': 0.884</em>, <em class="sig-param">'ML': 0.062</em>, <em class="sig-param">'MM': 0.0</em>, <em class="sig-param">'MN': 0.447</em>, <em class="sig-param">'MP': 0.285</em>, <em class="sig-param">'MQ': 0.372</em>, <em class="sig-param">'MR': 1.0</em>, <em class="sig-param">'MS': 0.417</em>, <em class="sig-param">'MT': 0.358</em>, <em class="sig-param">'MV': 0.12</em>, <em class="sig-param">'MW': 0.391</em>, <em class="sig-param">'MY': 0.255</em>, <em class="sig-param">'NA': 0.424</em>, <em class="sig-param">'NC': 0.425</em>, <em class="sig-param">'ND': 0.838</em>, <em class="sig-param">'NE': 0.835</em>, <em class="sig-param">'NF': 0.766</em>, <em class="sig-param">'NG': 0.512</em>, <em class="sig-param">'NH': 0.78</em>, <em class="sig-param">'NI': 0.615</em>, <em class="sig-param">'NK': 0.891</em>, <em class="sig-param">'NL': 0.603</em>, <em class="sig-param">'NM': 0.588</em>, <em class="sig-param">'NN': 0.0</em>, <em class="sig-param">'NP': 0.266</em>, <em class="sig-param">'NQ': 0.175</em>, <em class="sig-param">'NR': 1.0</em>, <em class="sig-param">'NS': 0.361</em>, <em class="sig-param">'NT': 0.368</em>, <em class="sig-param">'NV': 0.503</em>, <em class="sig-param">'NW': 0.945</em>, <em class="sig-param">'NY': 0.641</em>, <em class="sig-param">'PA': 0.22</em>, <em class="sig-param">'PC': 0.179</em>, <em class="sig-param">'PD': 0.852</em>, <em class="sig-param">'PE': 0.831</em>, <em class="sig-param">'PF': 0.515</em>, <em class="sig-param">'PG': 0.376</em>, <em class="sig-param">'PH': 0.696</em>, <em class="sig-param">'PI': 0.363</em>, <em class="sig-param">'PK': 0.875</em>, <em class="sig-param">'PL': 0.357</em>, <em class="sig-param">'PM': 0.326</em>, <em class="sig-param">'PN': 0.231</em>, <em class="sig-param">'PP': 0.0</em>, <em class="sig-param">'PQ': 0.228</em>, <em class="sig-param">'PR': 1.0</em>, <em class="sig-param">'PS': 0.196</em>, <em class="sig-param">'PT': 0.161</em>, <em class="sig-param">'PV': 0.244</em>, <em class="sig-param">'PW': 0.72</em>, <em class="sig-param">'PY': 0.481</em>, <em class="sig-param">'QA': 0.512</em>, <em class="sig-param">'QC': 0.462</em>, <em class="sig-param">'QD': 0.903</em>, <em class="sig-param">'QE': 0.861</em>, <em class="sig-param">'QF': 0.671</em>, <em class="sig-param">'QG': 0.648</em>, <em class="sig-param">'QH': 0.765</em>, <em class="sig-param">'QI': 0.532</em>, <em class="sig-param">'QK': 0.881</em>, <em class="sig-param">'QL': 0.518</em>, <em class="sig-param">'QM': 0.505</em>, <em class="sig-param">'QN': 0.181</em>, <em class="sig-param">'QP': 0.272</em>, <em class="sig-param">'QQ': 0.0</em>, <em class="sig-param">'QR': 1.0</em>, <em class="sig-param">'QS': 0.461</em>, <em class="sig-param">'QT': 0.389</em>, <em class="sig-param">'QV': 0.464</em>, <em class="sig-param">'QW': 0.831</em>, <em class="sig-param">'QY': 0.522</em>, <em class="sig-param">'RA': 0.919</em>, <em class="sig-param">'RC': 0.905</em>, <em class="sig-param">'RD': 0.305</em>, <em class="sig-param">'RE': 0.225</em>, <em class="sig-param">'RF': 0.977</em>, <em class="sig-param">'RG': 0.928</em>, <em class="sig-param">'RH': 0.498</em>, <em class="sig-param">'RI': 0.929</em>, <em class="sig-param">'RK': 0.141</em>, <em class="sig-param">'RL': 0.92</em>, <em class="sig-param">'RM': 0.908</em>, <em class="sig-param">'RN': 0.69</em>, <em class="sig-param">'RP': 0.796</em>, <em class="sig-param">'RQ': 0.668</em>, <em class="sig-param">'RR': 0.0</em>, <em class="sig-param">'RS': 0.86</em>, <em class="sig-param">'RT': 0.808</em>, <em class="sig-param">'RV': 0.914</em>, <em class="sig-param">'RW': 1.0</em>, <em class="sig-param">'RY': 0.859</em>, <em class="sig-param">'SA': 0.1</em>, <em class="sig-param">'SC': 0.185</em>, <em class="sig-param">'SD': 0.801</em>, <em class="sig-param">'SE': 0.812</em>, <em class="sig-param">'SF': 0.622</em>, <em class="sig-param">'SG': 0.17</em>, <em class="sig-param">'SH': 0.718</em>, <em class="sig-param">'SI': 0.478</em>, <em class="sig-param">'SK': 0.883</em>, <em class="sig-param">'SL': 0.474</em>, <em class="sig-param">'SM': 0.44</em>, <em class="sig-param">'SN': 0.289</em>, <em class="sig-param">'SP': 0.181</em>, <em class="sig-param">'SQ': 0.358</em>, <em class="sig-param">'SR': 1.0</em>, <em class="sig-param">'SS': 0.0</em>, <em class="sig-param">'ST': 0.174</em>, <em class="sig-param">'SV': 0.342</em>, <em class="sig-param">'SW': 0.827</em>, <em class="sig-param">'SY': 0.615</em>, <em class="sig-param">'TA': 0.251</em>, <em class="sig-param">'TC': 0.261</em>, <em class="sig-param">'TD': 0.83</em>, <em class="sig-param">'TE': 0.812</em>, <em class="sig-param">'TF': 0.604</em>, <em class="sig-param">'TG': 0.312</em>, <em class="sig-param">'TH': 0.737</em>, <em class="sig-param">'TI': 0.455</em>, <em class="sig-param">'TK': 0.866</em>, <em class="sig-param">'TL': 0.453</em>, <em class="sig-param">'TM': 0.403</em>, <em class="sig-param">'TN': 0.315</em>, <em class="sig-param">'TP': 0.159</em>, <em class="sig-param">'TQ': 0.322</em>, <em class="sig-param">'TR': 1.0</em>, <em class="sig-param">'TS': 0.185</em>, <em class="sig-param">'TT': 0.0</em>, <em class="sig-param">'TV': 0.345</em>, <em class="sig-param">'TW': 0.816</em>, <em class="sig-param">'TY': 0.596</em>, <em class="sig-param">'VA': 0.275</em>, <em class="sig-param">'VC': 0.165</em>, <em class="sig-param">'VD': 0.9</em>, <em class="sig-param">'VE': 0.867</em>, <em class="sig-param">'VF': 0.269</em>, <em class="sig-param">'VG': 0.471</em>, <em class="sig-param">'VH': 0.649</em>, <em class="sig-param">'VI': 0.135</em>, <em class="sig-param">'VK': 0.889</em>, <em class="sig-param">'VL': 0.134</em>, <em class="sig-param">'VM': 0.12</em>, <em class="sig-param">'VN': 0.38</em>, <em class="sig-param">'VP': 0.212</em>, <em class="sig-param">'VQ': 0.339</em>, <em class="sig-param">'VR': 1.0</em>, <em class="sig-param">'VS': 0.322</em>, <em class="sig-param">'VT': 0.305</em>, <em class="sig-param">'VV': 0.0</em>, <em class="sig-param">'VW': 0.472</em>, <em class="sig-param">'VY': 0.31</em>, <em class="sig-param">'WA': 0.658</em>, <em class="sig-param">'WC': 0.56</em>, <em class="sig-param">'WD': 1.0</em>, <em class="sig-param">'WE': 0.931</em>, <em class="sig-param">'WF': 0.196</em>, <em class="sig-param">'WG': 0.829</em>, <em class="sig-param">'WH': 0.678</em>, <em class="sig-param">'WI': 0.305</em>, <em class="sig-param">'WK': 0.892</em>, <em class="sig-param">'WL': 0.304</em>, <em class="sig-param">'WM': 0.344</em>, <em class="sig-param">'WN': 0.631</em>, <em class="sig-param">'WP': 0.555</em>, <em class="sig-param">'WQ': 0.538</em>, <em class="sig-param">'WR': 0.968</em>, <em class="sig-param">'WS': 0.689</em>, <em class="sig-param">'WT': 0.638</em>, <em class="sig-param">'WV': 0.418</em>, <em class="sig-param">'WW': 0.0</em>, <em class="sig-param">'WY': 0.204</em>, <em class="sig-param">'YA': 0.587</em>, <em class="sig-param">'YC': 0.478</em>, <em class="sig-param">'YD': 1.0</em>, <em class="sig-param">'YE': 0.932</em>, <em class="sig-param">'YF': 0.202</em>, <em class="sig-param">'YG': 0.782</em>, <em class="sig-param">'YH': 0.678</em>, <em class="sig-param">'YI': 0.23</em>, <em class="sig-param">'YK': 0.904</em>, <em class="sig-param">'YL': 0.219</em>, <em class="sig-param">'YM': 0.268</em>, <em class="sig-param">'YN': 0.512</em>, <em class="sig-param">'YP': 0.444</em>, <em class="sig-param">'YQ': 0.404</em>, <em class="sig-param">'YR': 0.995</em>, <em class="sig-param">'YS': 0.612</em>, <em class="sig-param">'YT': 0.557</em>, <em class="sig-param">'YV': 0.328</em>, <em class="sig-param">'YW': 0.244</em>, <em class="sig-param">'YY': 0.0}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_sequence_order_coupling_number_sw"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_sequence_order_coupling_number_sw" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the sequence order coupling numbers from 1 to maxlag
for a given protein sequence based on the Schneider-Wrede physicochemical
distance matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>maxlag</strong> – maxlag is the maximum lag and the length of the protein should be larger than maxlag. default is 30.</p></li>
<li><p><strong>distancematrix</strong> – dict form containing Schneider-Wrede physicochemical     distance matrix. omitted!</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing all sequence order coupling numbers based on the Schneider-Wrede physicochemical distance matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_sequence_order_coupling_number_total">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_sequence_order_coupling_number_total</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_sequence_order_coupling_number_total"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_sequence_order_coupling_number_total" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the sequence order coupling numbers from 1 to maxlag for a given protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – rotein is a pure protein sequence</p></li>
<li><p><strong>maxlag</strong> – maxlag is the maximum lag and the length of the protein should be larger than maxlag. default is 30.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing all sequence order coupling numbers</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="propythia.adjuv_functions.features_functions.quasi_sequence_order.get_sequence_order_coupling_numberp">
<code class="sig-prename descclassname">propythia.adjuv_functions.features_functions.quasi_sequence_order.</code><code class="sig-name descname">get_sequence_order_coupling_numberp</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em>, <em class="sig-param">maxlag=30</em>, <em class="sig-param">distancematrix={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/adjuv_functions/features_functions/quasi_sequence_order.html#get_sequence_order_coupling_numberp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.adjuv_functions.features_functions.quasi_sequence_order.get_sequence_order_coupling_numberp" title="Permalink to this definition">¶</a></dt>
<dd><p>Computing the sequence order coupling numbers from 1 to maxlag
for a given protein sequence based on the user-defined property.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_sequence</strong> – protein is a pure protein sequence</p></li>
<li><p><strong>maxlag</strong> – maxlag is the maximum lag and the length of the protein should be larger than maxlag. default is 30.</p></li>
<li><p><strong>distancematrix</strong> – distancematrix is the a dict form containing 400 distance values</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict form containing all sequence order coupling numbers based on the given property</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-propythia.adjuv_functions.features_functions">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-propythia.adjuv_functions.features_functions" title="Permalink to this headline">¶</a></h2>
<p>construct a package!</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="propythia.adjuv_functions.scoring.html" class="btn btn-neutral float-right" title="Adjuv functions: Scoring" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="propythia.adjuv_functions.sequence.html" class="btn btn-neutral float-left" title="Adjuv functions: Sequence" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Ana Marta Sequeira

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>