

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>ProPythia modules &mdash; ProPythia 0.04 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="propythia.adjuv_functions package" href="propythia.adjuv_functions.html" />
    <link rel="prev" title="Welcome to ProPythia’s documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> ProPythia
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">ProPythia modules</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-propythia.clustering">clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-propythia.descriptors">descriptors</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-propythia.feature_reduction">feature_reduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-propythia.feature_selection">feature_selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-propythia.machine_learning">machine_learning</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-propythia.preprocess">preprocess</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-propythia.sequence">sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-propythia">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="propythia.adjuv_functions.html">propythia.adjuv_functions package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ProPythia</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>ProPythia modules</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/propythia.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="propythia-modules">
<h1>ProPythia modules<a class="headerlink" href="#propythia-modules" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-propythia.clustering">
<span id="clustering"></span><h2>clustering<a class="headerlink" href="#module-propythia.clustering" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>File containing a class intend to facilitate clustering analysis.
The functions are based on the package scikit learn.</p>
<p>Authors: Ana Marta Sequeira</p>
<p>Date: 06/2019</p>
<p>Email:</p>
<hr class="docutils" />
<dl class="class">
<dt id="propythia.clustering.Cluster">
<em class="property">class </em><code class="sig-prename descclassname">propythia.clustering.</code><code class="sig-name descname">Cluster</code><span class="sig-paren">(</span><em class="sig-param">sklearn_load</em>, <em class="sig-param">target</em>, <em class="sig-param">test_size=0.3</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/clustering.html#Cluster"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.clustering.Cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The cluster class aims to perform and plot clustering analysis
Based on scikit learn.</p>
<dl class="method">
<dt id="propythia.clustering.Cluster.classify">
<code class="sig-name descname">classify</code><span class="sig-paren">(</span><em class="sig-param">model=SVC(C=1.0</em>, <em class="sig-param">cache_size=200</em>, <em class="sig-param">class_weight=None</em>, <em class="sig-param">coef0=0.0</em>, <em class="sig-param">decision_function_shape='ovr'</em>, <em class="sig-param">degree=3</em>, <em class="sig-param">gamma='auto_deprecated'</em>, <em class="sig-param">kernel='rbf'</em>, <em class="sig-param">max_iter=-1</em>, <em class="sig-param">probability=False</em>, <em class="sig-param">random_state=42</em>, <em class="sig-param">shrinking=True</em>, <em class="sig-param">tol=0.001</em>, <em class="sig-param">verbose=False)</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/clustering.html#Cluster.classify"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.clustering.Cluster.classify" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that fits the model in train datasets and predict on the tests dataset, returning the accuracy
:param model: model to make prediction (SVC by default)
:return: the accuracy of the prediction</p>
</dd></dl>

<dl class="method">
<dt id="propythia.clustering.Cluster.hierarchical">
<code class="sig-name descname">hierarchical</code><span class="sig-paren">(</span><em class="sig-param">metric='correlation'</em>, <em class="sig-param">method='complete'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/clustering.html#Cluster.hierarchical"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.clustering.Cluster.hierarchical" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform hierarchical clustering
:param metric: distance metric to use in the case that y is a collection of observation vectors. eg. ‘correlation’, ‘euclidean’
see (<a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html#scipy.spatial.distance.pdist">https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html#scipy.spatial.distance.pdist</a>)
:param method: method to be used. exemples: ‘complete’, ‘single’, ‘average’, ‘ward’
see (<a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html">https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html</a>)
:return: dendogram of the clustered data</p>
</dd></dl>

<dl class="method">
<dt id="propythia.clustering.Cluster.kmeans">
<code class="sig-name descname">kmeans</code><span class="sig-paren">(</span><em class="sig-param">max_iter=300</em>, <em class="sig-param">n_clusters=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/clustering.html#Cluster.kmeans"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.clustering.Cluster.kmeans" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that performs K means cluster.
:param max_iter: number of max terations for cluster (300 by default)
:param n_clusters: if None, it will define the number of clusters as the number of existing labels
:return: cross table with counts for labels vs classification in clusters</p>
</dd></dl>

<dl class="method">
<dt id="propythia.clustering.Cluster.kmeans_predict">
<code class="sig-name descname">kmeans_predict</code><span class="sig-paren">(</span><em class="sig-param">output='add'</em>, <em class="sig-param">max_iter=300</em>, <em class="sig-param">n_clusters=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/clustering.html#Cluster.kmeans_predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.clustering.Cluster.kmeans_predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform the kmeans to train data and predict the tests set
:param output: ‘add’ or ‘replace’.
If add, the labels produced by clustering will be added as features
If replace, labels produced will be replace the old labels
:param max_iter: max number of iterations of cluster
:param n_clusters: if None, it will define the number of clusters as the number of existing labels
:return: values of X datasets altered (if add) or the Y datasets replaced</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-propythia.descriptors">
<span id="descriptors"></span><h2>descriptors<a class="headerlink" href="#module-propythia.descriptors" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>A class  used for computing different types of protein descriptors.
It contains descriptors from packages pydpi, biopython, pfeature and modlamp.</p>
<p>Authors: Ana Marta Sequeira</p>
<p>Date: 05/2019</p>
<p>Email:</p>
<hr class="docutils" />
<dl class="class">
<dt id="propythia.descriptors.Descriptor">
<em class="property">class </em><code class="sig-prename descclassname">propythia.descriptors.</code><code class="sig-name descname">Descriptor</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The Descriptor class collects all descriptor calculation functions into a simple class.
Some of the descriptors functions are based on code from pydpi (altered to python3), modlamp, pfature and biopython.
It returns the features in a dictionary object</p>
<dl class="attribute">
<dt id="propythia.descriptors.Descriptor.AALetter">
<code class="sig-name descname">AALetter</code><em class="property"> = ['A', 'R', 'N', 'D', 'C', 'E', 'Q', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V']</em><a class="headerlink" href="#propythia.descriptors.Descriptor.AALetter" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="propythia.descriptors.Descriptor.Version">
<code class="sig-name descname">Version</code><em class="property"> = 1.0</em><a class="headerlink" href="#propythia.descriptors.Descriptor.Version" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.adaptable">
<code class="sig-name descname">adaptable</code><span class="sig-paren">(</span><em class="sig-param">list_of_functions=[]</em>, <em class="sig-param">ph=7</em>, <em class="sig-param">amide=False</em>, <em class="sig-param">tricomp=False</em>, <em class="sig-param">lamda_paac=10</em>, <em class="sig-param">weight_paac=0.05</em>, <em class="sig-param">lamda_apaac=10</em>, <em class="sig-param">weight_apaac=0.05</em>, <em class="sig-param">aap=[]</em>, <em class="sig-param">maxlag_socn=45</em>, <em class="sig-param">maxlag_qso=30</em>, <em class="sig-param">weight_qso=0.1</em>, <em class="sig-param">distancematrix={}</em>, <em class="sig-param">window=7</em>, <em class="sig-param">scalename='Eisenberg'</em>, <em class="sig-param">scalename_arc='peparc'</em>, <em class="sig-param">angle=100</em>, <em class="sig-param">modality='max'</em>, <em class="sig-param">prof_type='uH'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.adaptable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.adaptable" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to calculate user selected descriptors
:param ph:parameters for geral descriptors
:param amide:parameters for geral descriptors
:param tricomp: true or false to calculate or not tri-peptide on get_all
:param lamda_paac: parameters for PAAC: lamdaPAAC=10
:param weight_paac: parameters for PAAC weightPAAC=0.05
:param lamda_apaac: parmeters for APAAC lamdaAPAAC=5 IT SHOULD NOT BE LARGER THAN LENGHT SEQUENCE
:param weight_apaac:parmeters for APAAC weightAPAAC=0.05
:param aap:
:param maxlag_socn: parameters for SOCN: maxlagSOCN=45
:param maxlag_qso:parameters for QSO maxlagQSO=30
:param weight_qso:parameters for  weightQSO=0.1
:param distancematrix:
:param window:parameters for base class descriptors
:param scalename:parameters for base class descriptors
:param scalename_arc:parameters for base class descriptors
:param angle:parameters for base class descriptors
:param modality:parameters for base class descriptors
:param prof_type:parameters for base class descriptors
:return:dictionary with all features (value is variable)
:param list_of_functions:</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.calculate_arc">
<code class="sig-name descname">calculate_arc</code><span class="sig-paren">(</span><em class="sig-param">modality='max'</em>, <em class="sig-param">scalename='peparc'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.calculate_arc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.calculate_arc" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates arcs as seen in the helical wheel plot. Use for binary amino acid scales only ( 5 values) from modlamp
:param modality: maximum or mean
:param scalename: binary amino acid scales only
:return: dictionary with 5 values</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.calculate_autocorr">
<code class="sig-name descname">calculate_autocorr</code><span class="sig-paren">(</span><em class="sig-param">window</em>, <em class="sig-param">scalename='Eisenberg'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.calculate_autocorr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.calculate_autocorr" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates autocorrelation of amino acid values for a given descriptor scale ( variable &gt;&gt;&gt;&gt;&gt;&gt;values) from modlamp
:param window: correlation window for descriptor calculation in a sliding window approach
:param scalename:
:return: dictionary with values of autocorrelation</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.calculate_crosscorr">
<code class="sig-name descname">calculate_crosscorr</code><span class="sig-paren">(</span><em class="sig-param">window</em>, <em class="sig-param">scalename='Eisenberg'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.calculate_crosscorr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.calculate_crosscorr" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates cross correlation of amino acid values for a given descriptor scale ( variable &gt;&gt;&gt;&gt;&gt;&gt;values) from modlamp
:param window:correlation window for descriptor calculation in a sliding window approach
:param scalename:
:return: dictionary with values of crosscorrelation</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.calculate_global">
<code class="sig-name descname">calculate_global</code><span class="sig-paren">(</span><em class="sig-param">window=1000</em>, <em class="sig-param">modality='max'</em>, <em class="sig-param">scalename='Eisenberg'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.calculate_global"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.calculate_global" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates a global / window averaging descriptor value of a given AA scale of sequence (1 value) from modlamp
:param window: amino acid window. If the sequence is shorter than the window, the length of the sequence is taken
:param modality: maximum or mean hydrophobic moment
:param scalename:
:return: dictionary with one value</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.calculate_moment">
<code class="sig-name descname">calculate_moment</code><span class="sig-paren">(</span><em class="sig-param">window=1000</em>, <em class="sig-param">angle=100</em>, <em class="sig-param">modality='max'</em>, <em class="sig-param">scalename='Eisenberg'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.calculate_moment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.calculate_moment" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates moment of sequence (1 value) from modlamp
:param window: amino acid window in which to calculate the moment. If the sequence is shorter than the window, the length of the sequence is taken
:param angle: angle in which to calculate the moment. 100 for alpha helices, 180 for beta sheets
:param modality: maximum or mean hydrophobic moment
:param scalename:
:return: dictionary with one value of moment</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.calculate_profile">
<code class="sig-name descname">calculate_profile</code><span class="sig-paren">(</span><em class="sig-param">prof_type='uH'</em>, <em class="sig-param">window=7</em>, <em class="sig-param">scalename='Eisenberg'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.calculate_profile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.calculate_profile" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates hydrophobicity or hydrophobic moment profiles for given sequences and fitting for slope and intercep
(2 values) from modlamp
:param prof_type: prof_type of profile, ‘H’ for hydrophobicity or ‘uH’ for hydrophobic moment
:param window: size of sliding window used (odd-numbered)
:param scalename:
:return: dictionary with two value</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_aa_comp">
<code class="sig-name descname">get_aa_comp</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_aa_comp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_aa_comp" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates amino acid compositon (20 values)  from pydpi
:return: dictionary with the fractions of all 20 aa(keys are the aa)</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_aa_index1">
<code class="sig-name descname">get_aa_index1</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">path='.'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_aa_index1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_aa_index1" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the amino acid property values from aaindex1
(function from pydpi)
:param name: name is the name of amino acid property (e.g., KRIW790103)
:param path: path to get aa index. by default is the one in the package.
:return: result is a dict form containing the properties of 20 amino acids</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_aa_index23">
<code class="sig-name descname">get_aa_index23</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">path='.'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_aa_index23"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_aa_index23" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the amino acid property values from aaindex2 and aaindex3
(function from from pydpi)
:param name: name is the name of amino acid property (e.g.,TANS760101,GRAR740104)
:param path: path to get aa index. by default is the one in the package.
:return: result is a dict form containing the properties of 400 amino acid pairs</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_aliphatic_index">
<code class="sig-name descname">get_aliphatic_index</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_aliphatic_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_aliphatic_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates aliphatic index of sequence (1 value) from modlamp
:return: dictionary with the value of aliphatic index</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_all">
<code class="sig-name descname">get_all</code><span class="sig-paren">(</span><em class="sig-param">ph=7</em>, <em class="sig-param">amide=False</em>, <em class="sig-param">tricomp=False</em>, <em class="sig-param">bin_aa=False</em>, <em class="sig-param">bin_prop=False</em>, <em class="sig-param">lamda_paac=10</em>, <em class="sig-param">weight_paac=0.05</em>, <em class="sig-param">lamda_apaac=10</em>, <em class="sig-param">weight_apaac=0.05</em>, <em class="sig-param">maxlag_socn=45</em>, <em class="sig-param">maxlag_qso=30</em>, <em class="sig-param">weight_qso=0.1</em>, <em class="sig-param">window=7</em>, <em class="sig-param">scalename='Eisenberg'</em>, <em class="sig-param">scalename_arc='peparc'</em>, <em class="sig-param">angle=100</em>, <em class="sig-param">modality='max'</em>, <em class="sig-param">prof_type='uH'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_all"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_all" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all descriptors from pydpi_py3 except tri-peptide pydpi_py3 and binary profiles
:param ph:parameters for geral descriptors
:param amide:parameters for geral descriptors
:param tricomp: true or false to calculate or not tri-peptide pydpi_py3
:param bin_aa: true or false to calculate or not binary composition of aa
:param bin_prop: true or false to calculate or not binary composition of properties of resides
:param lamda_paac: parameters for PAAC: lamdaPAAC=10 should not be larger than len(sequence)
:param weight_paac: parameters for PAAC weightPAAC=0.05
:param AAP: list with
:param lamda_apaac: parmeters for APAAC lamdaAPAAC=10
:param weight_apaac:parmeters for APAAC weightAPAAC=0.05
:param maxlag_socn: parameters for SOCN: maxlagSOCN=45
:param maxlag_qso:parameters for QSO maxlagQSO=30
:param weight_qso:parameters for  weightQSO=0.1
:param window:parameters for base class descriptors
:param scalename:parameters for base class descriptors
:param scalename_arc:parameters for base class descriptors
:param angle:parameters for base class descriptors
:param modality:parameters for base class descriptors
:param prof_type:parameters for base class descriptors
:return:dictionary with all features (value is variable)</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_all_aac">
<code class="sig-name descname">get_all_aac</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_all_aac"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_all_aac" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all descriptors from Amino Acid Composition
:return: dictionary with values from AAC, DPC and TPC</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_all_base_class">
<code class="sig-name descname">get_all_base_class</code><span class="sig-paren">(</span><em class="sig-param">window=7</em>, <em class="sig-param">scalename='Eisenberg'</em>, <em class="sig-param">scalename_arc='peparc'</em>, <em class="sig-param">angle=100</em>, <em class="sig-param">modality='max'</em>, <em class="sig-param">prof_type='uH'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_all_base_class"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_all_base_class" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all functions from Base class
:param window:
:param scalename:
:param scalename_arc:
:param angle:
:param modality:
:param prof_type:
:return: dictionary with all 6 base class peptide descriptors (the value is variable)</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_all_correlation">
<code class="sig-name descname">get_all_correlation</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_all_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_all_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all descriptors from Autocorrelation
:return: values for the funtions Moreau Broto, Moran and Geary autocorrelation</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_all_paac">
<code class="sig-name descname">get_all_paac</code><span class="sig-paren">(</span><em class="sig-param">lamda_paac=10</em>, <em class="sig-param">weight_paac=0.05</em>, <em class="sig-param">lamda_apaac=10</em>, <em class="sig-param">weight_apaac=0.05</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_all_paac"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_all_paac" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all descriptors from Pseudo Amino Acid Composition
:param lamda_paac: parameter for PAAC default 10
:param weight_paac: parameter for PAAC default 0.05
:param lamda_apaac: parameter for APAAC default 10
:param weight_apaac: parameter for APAAC default 0.05
:return: dictionary with values from PAAC and APAAC</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_all_physicochemical">
<code class="sig-name descname">get_all_physicochemical</code><span class="sig-paren">(</span><em class="sig-param">ph=7</em>, <em class="sig-param">amide=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_all_physicochemical"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_all_physicochemical" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all 15 geral descriptors functions derived from biopython and modlpam
:param ph: for functions Charge, charge density and formula
:param amide: for functions Charge, charge density and formula
:return: dictionary with variable number of descriptors</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_all_sequenceorder">
<code class="sig-name descname">get_all_sequenceorder</code><span class="sig-paren">(</span><em class="sig-param">maxlag_socn=45</em>, <em class="sig-param">maxlag_qso=30</em>, <em class="sig-param">weight_qso=0.1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_all_sequenceorder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_all_sequenceorder" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all values for sequence order descriptors
:param maxlag_socn: parameter for SOCN default 45
:param maxlag_qso: parameter for QSO default 30
:param weight_qso: parameter for QSO default 0.1
:return: dictionary with values for quasi sequence order and sequence order couplig numbers</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_apaac">
<code class="sig-name descname">get_apaac</code><span class="sig-paren">(</span><em class="sig-param">lamda=10</em>, <em class="sig-param">weight=0.5</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_apaac"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_apaac" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates Type II Pseudo amino acid composition - Amphiphilic (default is 30, depends on lamda) from pydpi
:param lamda: reflects the rank of correlation and is a non-Negative integer, such as 10.</p>
<blockquote>
<div><p>should NOT be larger than the length of input protein sequence
when lamda =0, theoutput of PseAA server is the 20-D amino acid composition</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>weight</strong> – weight on the additional PseAA components. with respect to the conventional AA components.
The user can select any value within the region from 0.05 to 0.7 for the weight factor.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary with the fractions of all PAAC(keys are the PAAC). Number of keys depends on lamda</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_aromacity">
<code class="sig-name descname">get_aromacity</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_aromacity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_aromacity" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates Aromacity from sequence (1 value) from biopython
:return: dictionary with the value of aromacity</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_bin_aa">
<code class="sig-name descname">get_bin_aa</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_bin_aa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_bin_aa" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div><p>binary profile of aminoacid composition</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dictionary containing binary profile</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_bin_resi_prop">
<code class="sig-name descname">get_bin_resi_prop</code><span class="sig-paren">(</span><em class="sig-param">list_descriptors=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_bin_resi_prop"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_bin_resi_prop" title="Permalink to this definition">¶</a></dt>
<dd><p>Binary profile of residues for 25 phychem feature
:param list_descriptors: ist containing feature numbers (valid number, 0-24) (e.g. [5,4,8,24]</p>
<blockquote>
<div><p>FEATURE NAME                    FEATURE NUMBER</p>
<blockquote>
<div><p>‘Positively charged’ –                  0
‘Negatively charged’ –                  1
‘Neutral charged’ –                     2
‘Polarity’ –                            3
‘Non polarity’ –                        4
‘Aliphaticity’ –                        5
‘Cyclic’ –                              6
‘Aromaticity’ –                         7
‘Acidicity’–                            8
‘Basicity’–                             9
‘Neutral (ph)’ –                       10
‘Hydrophobicity’ –                     11
‘Hydrophilicity’ –                     12
‘Neutral’ –                            13
‘Hydroxylic’ –                         14
‘Sulphur content’ -                     15
‘Secondary Structure(Helix)’            16
‘Secondary Structure(Strands)’,         17
‘Secondary Structure(Coil)’,            18
‘Solvent Accessibilty (Buried)’,        19
‘Solvent Accesibilty(Exposed)’,         20
‘Solvent Accesibilty(Intermediate)’,    21
‘Tiny’,                                 22
‘Small’,                                23
‘Large’                                 24</p>
</div></blockquote>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict form with binary values</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_boman_index">
<code class="sig-name descname">get_boman_index</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_boman_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_boman_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates boman index of sequence (1 value) from modlamp
:return: dictionary with the value of boman index</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_bond">
<code class="sig-name descname">get_bond</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_bond"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_bond" title="Permalink to this definition">¶</a></dt>
<dd><p>This function gives the sum of the bond composition for each type of bond
For bond composition four types of bonds are considered
total number of bonds (including aromatic), hydrogen bond, single bond and double bond.
:return: dictionary with 4 values</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_charge">
<code class="sig-name descname">get_charge</code><span class="sig-paren">(</span><em class="sig-param">ph=7.4</em>, <em class="sig-param">amide=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_charge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_charge" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates charge of sequence (1 value) from modlamp
:param ph: ph considered to calculate. 7.4 by default
:param amide: by default is not considered an amide protein sequence.
:return: dictionary with the value of charge</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_charge_density">
<code class="sig-name descname">get_charge_density</code><span class="sig-paren">(</span><em class="sig-param">ph=7.0</em>, <em class="sig-param">amide=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_charge_density"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_charge_density" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates charge density of sequence (1 value) from modlamp
:param ph: ph considered to calculate. 7 by default
:param amide: by default is not considered an amide protein sequence.
:return: dictionary with the value of charge density</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_conj_t">
<code class="sig-name descname">get_conj_t</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_conj_t"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_conj_t" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the Conjoint Triad descriptors (343 descriptors) from pydpi
:return: dictionary with the 343 descriptors</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_ctd">
<code class="sig-name descname">get_ctd</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_ctd"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_ctd" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the Composition Transition Distribution descriptors (147 values) from pydpi
:return: dictionary with the 147 descriptors</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_dp_comp">
<code class="sig-name descname">get_dp_comp</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_dp_comp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_dp_comp" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates dipeptide composition (400 values) from pydpi
:return: dictionary with the fractions of all 400 possible combiinations of 2 aa</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_flexibility">
<code class="sig-name descname">get_flexibility</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_flexibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_flexibility" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the flexibility according to Vihinen, 1994 (return proteinsequencelenght-9 values ) from biopython
:return: dictionary with proteinsequencelenght-9 values of flexiblity</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_formula">
<code class="sig-name descname">get_formula</code><span class="sig-paren">(</span><em class="sig-param">amide=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_formula"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_formula" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates number of C,H,N,O and S of the aa of sequence (5 values) from modlamp
:param amide: by default is not considered an amide protein sequence.
:return: dictionary with the 5 values of C,H,N,O and S</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_geary_auto">
<code class="sig-name descname">get_geary_auto</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_geary_auto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_geary_auto" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates  Geary autocorrelation (240 values) from pydpi
:return: dictionary with the 240 descriptors</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_gravy">
<code class="sig-name descname">get_gravy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_gravy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_gravy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates Gravy from sequence (1 value) from biopython
:return: dictionary with the value of gravy</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_hydrophobic_ratio">
<code class="sig-name descname">get_hydrophobic_ratio</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_hydrophobic_ratio"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_hydrophobic_ratio" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates hydrophobic ratio of sequence (1 value) from modlamp
:return: dictionary with the value of hydrophobic ratio</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_instability_index">
<code class="sig-name descname">get_instability_index</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_instability_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_instability_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates Instability index from sequence (1 value) from biopython
:return: dictionary with the value of Instability index</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_isoelectric_point">
<code class="sig-name descname">get_isoelectric_point</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_isoelectric_point"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_isoelectric_point" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates Isolectric Point from sequence (1 value) from biopython
:return: dictionary with the value of Isolectric Point</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_lenght">
<code class="sig-name descname">get_lenght</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_lenght"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_lenght" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates lenght of sequence (number of aa)
:return: dictionary with the value of lenght</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_molar_extinction_coefficient">
<code class="sig-name descname">get_molar_extinction_coefficient</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_molar_extinction_coefficient"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_molar_extinction_coefficient" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the molar extinction coefficient (2 values) from biopython
:return: dictionary with the value of reduced cysteins and oxidized (with disulfid bridges)</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_moran_auto">
<code class="sig-name descname">get_moran_auto</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_moran_auto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_moran_auto" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates  Moran autocorrelation (240 values) from pydpi
:return: dictionary with the 240 descriptors</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_moreau_broto_auto">
<code class="sig-name descname">get_moreau_broto_auto</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_moreau_broto_auto"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_moreau_broto_auto" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates Normalized Moreau-Broto autocorrelation (240 values) from pydpi
:return: dictionary with the 240 descriptors</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_mw">
<code class="sig-name descname">get_mw</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_mw"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_mw" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates molecular weight of sequence (1 value) from modlamp
:return: dictionary with the value of molecular weight</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_paac">
<code class="sig-name descname">get_paac</code><span class="sig-paren">(</span><em class="sig-param">lamda=10</em>, <em class="sig-param">weight=0.05</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_paac"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_paac" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates Type I Pseudo amino acid composition (default is 30, depends on lamda) from pydpi
:param lamda: reflects the rank of correlation and is a non-Negative integer, such as 10.</p>
<blockquote>
<div><p>should NOT be larger than the length of input protein sequence
when lamda =0, the output of PseAA server is the 20-D amino acid composition</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>weight</strong> – weight on the additional PseAA components. with respect to the conventional AA components.
The user can select any value within the region from 0.05 to 0.7 for the weight factor.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary with the fractions of all PAAC (keys are the PAAC). Number of keys depends on lamda</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_paac_p">
<code class="sig-name descname">get_paac_p</code><span class="sig-paren">(</span><em class="sig-param">lamda=10</em>, <em class="sig-param">weight=0.05</em>, <em class="sig-param">AAP=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_paac_p"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_paac_p" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates Type I Pseudo amino acid composition for a given property (default is 30, depends on lamda) from pydpi
:param lamda: reflects the rank of correlation and is a non-Negative integer, such as 10.</p>
<blockquote>
<div><p>should NOT be larger than the length of input protein sequence
when lamda =0, theoutput of PseAA server is the 20-D amino acid composition</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>weight</strong> – weight on the additional PseAA components. with respect to the conventional AA components.
The user can select any value within the region from 0.05 to 0.7 for the weight factor.</p></li>
<li><p><strong>AAP</strong> – list of properties. each of which is a dict form.
PseudoAAC._Hydrophobicity,PseudoAAC._hydrophilicity, PseudoAAC._residuemass,PseudoAAC._pK1,PseudoAAC._pK2,PseudoAAC._pI</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary with the fractions of all PAAC(keys are the PAAC). Number of keys depends on lamda</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_qso">
<code class="sig-name descname">get_qso</code><span class="sig-paren">(</span><em class="sig-param">maxlag=30</em>, <em class="sig-param">weight=0.1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_qso"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_qso" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the Quasi sequence order  (retrieves 100 values by default) from pydpi
:param maxlag: maximum lag. Smaller than length of the protein
:param weight:
:return: dictionary with the descriptors (100 descriptors)</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_qso_p">
<code class="sig-name descname">get_qso_p</code><span class="sig-paren">(</span><em class="sig-param">maxlag=30</em>, <em class="sig-param">weight=0.1</em>, <em class="sig-param">distancematrix={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_qso_p"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_qso_p" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the Quasi sequence order  (retrieves 100 values by default) from pydpi
:param maxlag: maximum lag. Smaller than length of the protein
:param weight:
:param distancematrix: dict form containing 400 distance values
:return: dictionary with the descriptors (100 descriptors)</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_sec_struct">
<code class="sig-name descname">get_sec_struct</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_sec_struct"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_sec_struct" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the fraction of amino acids which tend to be in helix, turn or sheet (3 value) from biopython
:return: dictionary with the 3 value of helix, turn, sheet</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_socn">
<code class="sig-name descname">get_socn</code><span class="sig-paren">(</span><em class="sig-param">maxlag=45</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_socn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_socn" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the Sequence order coupling numbers  (retrieves 90 values by default) from pydpi
:param maxlag: maximum lag. Smaller than length of the protein
:return: dictionary with the descriptors (90 descriptors)</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_socn_p">
<code class="sig-name descname">get_socn_p</code><span class="sig-paren">(</span><em class="sig-param">maxlag=45</em>, <em class="sig-param">distancematrix={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_socn_p"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_socn_p" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the Sequence order coupling numbers  (retrieves 90 values by default) from pydpi
:param maxlag: maximum lag. Smaller than length of the protein
:param distancematrix: dict form containing 400 distance values
:return: dictionary with the descriptors (90 descriptors)</p>
</dd></dl>

<dl class="method">
<dt id="propythia.descriptors.Descriptor.get_tp_comp">
<code class="sig-name descname">get_tp_comp</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/descriptors.html#Descriptor.get_tp_comp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.descriptors.Descriptor.get_tp_comp" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates tripeptide composition (8000 values) from pydpi
:return: dictionary with the fractions of all 8000 possible combinations of 3 aa</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-propythia.feature_reduction">
<span id="feature-reduction"></span><h2>feature_reduction<a class="headerlink" href="#module-propythia.feature_reduction" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>File containing a class used for reducing the number of features on a dataset based
on unsupervised techniques.</p>
<p>Authors: Ana Marta Sequeira</p>
<p>Date: 06/2019</p>
<p>Email:</p>
<hr class="docutils" />
<dl class="class">
<dt id="propythia.feature_reduction.FeatureReduction">
<em class="property">class </em><code class="sig-prename descclassname">propythia.feature_reduction.</code><code class="sig-name descname">FeatureReduction</code><a class="reference internal" href="_modules/propythia/feature_reduction.html#FeatureReduction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_reduction.FeatureReduction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The Feature Reduction class aims to Reduce the number of features on a dataset based on unsupervised techniques.
pca statistical procedure that orthogonally transforms the original n coordinates of a data set into a new set of
n coordinates called principal components. Principal components are a combination of features that capture well the
variance of the original features.
Based on scikit learn</p>
<dl class="method">
<dt id="propythia.feature_reduction.FeatureReduction.contribution_of_features_to_component">
<code class="sig-name descname">contribution_of_features_to_component</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">pca</em>, <em class="sig-param">x_pca</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_reduction.html#FeatureReduction.contribution_of_features_to_component"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_reduction.FeatureReduction.contribution_of_features_to_component" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that retrieves a dataframe containing the contribution of each feature (rows) for component
As unsupervised learning does not represent the importance of features but representing the directions
of maximum variance in the data.
:param data: dataset as dataframe
:param pca: dataset fit to pca
:param x_pca: dataset transformed to pca
:return: dataframe containing the contribution of each feature (rows) for component</p>
</dd></dl>

<dl class="method">
<dt id="propythia.feature_reduction.FeatureReduction.pca">
<code class="sig-name descname">pca</code><span class="sig-paren">(</span><em class="sig-param">dataset</em>, <em class="sig-param">scaler=StandardScaler(copy=True</em>, <em class="sig-param">with_mean=True</em>, <em class="sig-param">with_std=True)</em>, <em class="sig-param">n_components=None</em>, <em class="sig-param">copy=True</em>, <em class="sig-param">whiten=False</em>, <em class="sig-param">svd_solver='auto'</em>, <em class="sig-param">tol=0.0</em>, <em class="sig-param">iterated_power='auto'</em>, <em class="sig-param">random_state=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_reduction.html#FeatureReduction.pca"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_reduction.FeatureReduction.pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that realizes the pca analysis
:param dataset: data on to perform pca
:param scaler: scaler to scale data. standard scaler by default
:param n_components: Number of components to keep. if n_components is not set all components are kept
:param copy:
:param whiten:
:param svd_solver:  string {‘auto’, ‘full’, ‘arpack’, ‘randomized’}
:param tol:
:param iterated_power:
:param random_state:
:return:
For more information: <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html">https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html</a></p>
</dd></dl>

<dl class="method">
<dt id="propythia.feature_reduction.FeatureReduction.pca_bar_plot">
<code class="sig-name descname">pca_bar_plot</code><span class="sig-paren">(</span><em class="sig-param">pca</em>, <em class="sig-param">height=1</em>, <em class="sig-param">width=1</em>, <em class="sig-param">data=None</em>, <em class="sig-param">color='b'</em>, <em class="sig-param">edgecolor='k'</em>, <em class="sig-param">linewidth=0</em>, <em class="sig-param">tick_label=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_reduction.html#FeatureReduction.pca_bar_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_reduction.FeatureReduction.pca_bar_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>function that derives a bar plot representing the percentage of explained variance ratio by pca
:param pca: dataset fit to pca
:param height: scalar or sequence of scalars. The height(s) of the bars.
:param width:  scalar or array-like, optional. The width(s) of the bars
:param data:
:param color: scalar or array-like, optional. The colors of the bar faces.
:param edgecolor:  scalar or array-like, optional. The colors of the bar edges.
:param linewidth: scalar or array-like, optional. Width of the bar edge(s). If 0, don’t draw edges.
:param tick_label: string or array-like, optional. The tick labels of the bars. Default: None
:return: bar plot representing the percentage of explained variance ratio by pca
For more information please see <a class="reference external" href="https://matplotlib.org/api/_as_gen/matplotlib.pyplot.bar.html">https://matplotlib.org/api/_as_gen/matplotlib.pyplot.bar.html</a></p>
</dd></dl>

<dl class="method">
<dt id="propythia.feature_reduction.FeatureReduction.pca_scatter_plot">
<code class="sig-name descname">pca_scatter_plot</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">pca</em>, <em class="sig-param">x_pca</em>, <em class="sig-param">target</em>, <em class="sig-param">pca1=0</em>, <em class="sig-param">pca2=1</em>, <em class="sig-param">title='pca'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_reduction.html#FeatureReduction.pca_scatter_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_reduction.FeatureReduction.pca_scatter_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Scatter plot of the labels based on two components (by default the first ones)
:param title: string. title of the scatter plot
:param data: dataset. dataframe
:param pca: dataset fit to pca
:param x_pca: dataset transformed to pca
:param target: labels of dataset
:param pca1: first pca to be considered. default PCA1
:param pca2: second pca to be considered. default PCA2
:return: graph showing the positions of labels according of the two chosen components</p>
</dd></dl>

<dl class="method">
<dt id="propythia.feature_reduction.FeatureReduction.variance_ratio_components">
<code class="sig-name descname">variance_ratio_components</code><span class="sig-paren">(</span><em class="sig-param">x_pca</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_reduction.html#FeatureReduction.variance_ratio_components"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_reduction.FeatureReduction.variance_ratio_components" title="Permalink to this definition">¶</a></dt>
<dd><p>measures the variance ratio of the principal components
:param x_pca:
:return: variance ratio of principal components</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-propythia.feature_selection">
<span id="feature-selection"></span><h2>feature_selection<a class="headerlink" href="#module-propythia.feature_selection" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>File containing a class used for feature selection. The FeatureSelection class aims to select features based on supervised algorithms in order to improve
estimators’ accuracy scores or to boost their performance on very high-dimensional datasets.</p>
<p>Authors: Ana Marta Sequeira</p>
<p>Date:06/2019</p>
<p>Email:</p>
<hr class="docutils" />
<dl class="class">
<dt id="propythia.feature_selection.FeatureSelection">
<em class="property">class </em><code class="sig-prename descclassname">propythia.feature_selection.</code><code class="sig-name descname">FeatureSelection</code><span class="sig-paren">(</span><em class="sig-param">dataset</em>, <em class="sig-param">x_original</em>, <em class="sig-param">target</em>, <em class="sig-param">test_size=0.3</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_selection.html#FeatureSelection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_selection.FeatureSelection" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The FeatureSelection class aims to select features to improve estimators’ accuracy scores or to boost
their performance on very high-dimensional datasets.
It implements sklearn functions</p>
<dl class="method">
<dt id="propythia.feature_selection.FeatureSelection.features_scores">
<code class="sig-name descname">features_scores</code><span class="sig-paren">(</span><em class="sig-param">x_original</em>, <em class="sig-param">scores</em>, <em class="sig-param">column_selected</em>, <em class="sig-param">all=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_selection.html#FeatureSelection.features_scores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_selection.FeatureSelection.features_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieves a dataframe with features names and scores of importance resulting of the univariate tests
:param dataset: original dataset with columns names
:param scores: list of scores of the features (can be obtained by the function univariate)
:param column_selected: list containing the indexes of the selected columns (can be obtained by the function univariate)
:param all: to return all the features and scores or only the selected ones (by default)
:return: a dataframe containing the names of features and the scores of univariate tests by descending importance</p>
</dd></dl>

<dl class="method">
<dt id="propythia.feature_selection.FeatureSelection.recursive_feature_elimination">
<code class="sig-name descname">recursive_feature_elimination</code><span class="sig-paren">(</span><em class="sig-param">cross_validation=False</em>, <em class="sig-param">estimator=SVC(C=1.0</em>, <em class="sig-param">cache_size=200</em>, <em class="sig-param">class_weight=None</em>, <em class="sig-param">coef0=0.0</em>, <em class="sig-param">decision_function_shape='ovr'</em>, <em class="sig-param">degree=3</em>, <em class="sig-param">gamma='auto_deprecated'</em>, <em class="sig-param">kernel='linear'</em>, <em class="sig-param">max_iter=-1</em>, <em class="sig-param">probability=False</em>, <em class="sig-param">random_state=None</em>, <em class="sig-param">shrinking=True</em>, <em class="sig-param">tol=0.001</em>, <em class="sig-param">verbose=False)</em>, <em class="sig-param">n_features_to_select=None</em>, <em class="sig-param">min_features_to_select=1</em>, <em class="sig-param">cv=None</em>, <em class="sig-param">scoring=None</em>, <em class="sig-param">n_jobs=None</em>, <em class="sig-param">step=1</em>, <em class="sig-param">verbose=0</em>, <em class="sig-param">scaler=StandardScaler(copy=True</em>, <em class="sig-param">with_mean=True</em>, <em class="sig-param">with_std=True)</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_selection.html#FeatureSelection.recursive_feature_elimination"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_selection.FeatureSelection.recursive_feature_elimination" title="Permalink to this definition">¶</a></dt>
<dd><p>Given an external estimator that assigns weights to features (e.g., the coefficients of a linear model), recursive feature elimination (RFE)
is to select features by recursively considering smaller and smaller sets of features.
First, the estimator is trained on the initial set of features and the importance of each feature is obtained either
through a <a href="#id1"><span class="problematic" id="id2">coef_</span></a> attribute or through a <a href="#id3"><span class="problematic" id="id4">feature_importances_</span></a> attribute.
Then, the least important features are pruned from current set of features.That procedure is recursively repeated on the pruned set until
the desired number of features to select is eventually reached.</p>
<p>RFECV performs RFE in a cross-validation loop to find the optimal number of features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>scaler</strong> – scaler function to use to datasets before apply univariate tests.</p>
</dd>
</dl>
<p>It can be None or any function supported by SKlearn like StandardScaler()
:param cross_validation: if yes: RFECV . if not: RFE
:param estimator: estimator that assign wights to features
:param n_features_to_select: to RFE
:param min_features_to_select: to RFECV
:param cv: number of folds in cross validation
:param scoring: for RFECV
:param n_jobs:
:param step:If greater than or equal to 1, then step corresponds to the (integer) number of features to remove
at each iteration. If within (0.0, 1.0), then step corresponds to the percentage (rounded down) of features to
remove at each iteration
:param verbose:
:return: rfe fit, rfe transformed, original dataset with features selected, columns names and the features ranking</p>
</dd></dl>

<dl class="method">
<dt id="propythia.feature_selection.FeatureSelection.rfe_ranking">
<code class="sig-name descname">rfe_ranking</code><span class="sig-paren">(</span><em class="sig-param">features_ranking</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_selection.html#FeatureSelection.rfe_ranking"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_selection.FeatureSelection.rfe_ranking" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieves a dataframe with features names and its ranking position ordered. Positions 1 are the selected ones.
:param dataset: dataset used to performed te fre
:param features_ranking: array containing the features ranking obtained with rfe
:return: dataframe containing the features names and its position ranking ordered by ranking</p>
</dd></dl>

<dl class="method">
<dt id="propythia.feature_selection.FeatureSelection.select_from_model_feature_elimination">
<code class="sig-name descname">select_from_model_feature_elimination</code><span class="sig-paren">(</span><em class="sig-param">model=LinearSVC(C=0.1</em>, <em class="sig-param">class_weight=None</em>, <em class="sig-param">dual=False</em>, <em class="sig-param">fit_intercept=True</em>, <em class="sig-param">intercept_scaling=1</em>, <em class="sig-param">loss='squared_hinge'</em>, <em class="sig-param">max_iter=1000</em>, <em class="sig-param">multi_class='ovr'</em>, <em class="sig-param">penalty='l1'</em>, <em class="sig-param">random_state=None</em>, <em class="sig-param">tol=0.0001</em>, <em class="sig-param">verbose=0)</em>, <em class="sig-param">scaler=StandardScaler(copy=True</em>, <em class="sig-param">with_mean=True</em>, <em class="sig-param">with_std=True)</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_selection.html#FeatureSelection.select_from_model_feature_elimination"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_selection.FeatureSelection.select_from_model_feature_elimination" title="Permalink to this definition">¶</a></dt>
<dd><p>SelectFromModel is a meta-transformer that can be used along with any estimator that has a <a href="#id5"><span class="problematic" id="id6">coef_</span></a> or <a href="#id7"><span class="problematic" id="id8">feature_importances_</span></a> attribute
after fitting.
The features are considered unimportant and removed, if the corresponding <a href="#id9"><span class="problematic" id="id10">coef_</span></a> or <a href="#id11"><span class="problematic" id="id12">feature_importances_</span></a> values are
below the provided threshold parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>scaler</strong> – scaler function to use to datasets before apply univariate tests.</p>
</dd>
</dl>
<p>It can be None or any function supported by SKlearn like StandardScaler()
:param data: dataset to perform the feature selection
:param model:</p>
<blockquote>
<div><p>examples:
ExtraTreesClassifier(n_estimators=50)
LinearSVC(C=0.01, penalty=”l1”, dual=False)
LogisticRegression(C=0.1, penalty=”l1”, dual=False)</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.feature_selection.FeatureSelection.univariate">
<code class="sig-name descname">univariate</code><span class="sig-paren">(</span><em class="sig-param">score_func=&lt;function mutual_info_classif&gt;</em>, <em class="sig-param">mode='percentile'</em>, <em class="sig-param">param=50</em>, <em class="sig-param">scaler=StandardScaler(copy=True</em>, <em class="sig-param">with_mean=True</em>, <em class="sig-param">with_std=True)</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/feature_selection.html#FeatureSelection.univariate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.feature_selection.FeatureSelection.univariate" title="Permalink to this definition">¶</a></dt>
<dd><p>Univariate feature selector, it selects works the best features based on univariate statistical tests.
It can select the k highest scoring features or a user specified percentage of features.
Scoring functions for classification problems can be chi2, f_classif or mutual_info_classif</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>scaler</strong> – scaler function to use to datasets before apply univariate tests.</p>
</dd>
</dl>
<p>It can be None or any function supported by SKlearn like StandardScaler()
:param score_func: function that returns univariate scores and p-values (or only scores for SelectKBest and SelectPercentile)</p>
<blockquote>
<div><p>( for classification: chi2, f_classif, mutual_info_classif</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mode</strong> – feature selection mode (‘percentile’, ‘k_best’, ‘fpr’, ‘fdr’, ‘fwe’)</p>
</dd>
</dl>
<p>:param param:parameter of corresponding mode
:return: univariate scores and p-values</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-propythia.machine_learning">
<span id="machine-learning"></span><h2>machine_learning<a class="headerlink" href="#module-propythia.machine_learning" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>File containing a class intend to facilitate machine learning with peptides.
The functions are based on the package scikit learn.
The organization of the code is based on the mddlamp package.</p>
<p>Authors: Ana Marta Sequeira</p>
<p>Date:06/2019</p>
<p>Email:</p>
<hr class="docutils" />
<dl class="class">
<dt id="propythia.machine_learning.MachineLearning">
<em class="property">class </em><code class="sig-prename descclassname">propythia.machine_learning.</code><code class="sig-name descname">MachineLearning</code><span class="sig-paren">(</span><em class="sig-param">sklearn_load, target, test_size=0.3, classes=['pos', 'neg']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/machine_learning.html#MachineLearning"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.machine_learning.MachineLearning" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The MachineLearning class aims to process different models of machine learning with peptides.
Based on scikit learn.</p>
<dl class="method">
<dt id="propythia.machine_learning.MachineLearning.add_features">
<code class="sig-name descname">add_features</code><span class="sig-paren">(</span><em class="sig-param">list_of_sequences</em>, <em class="sig-param">list_functions</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/machine_learning.html#MachineLearning.add_features"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.machine_learning.MachineLearning.add_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate all features available in package or lis of descriptors given
:param list_functions: list of features to be calculated with function adaptable from module Descriptors
:param list_of_sequences: list of sequences to calculate features
:return: dataframe with sequences and its features</p>
</dd></dl>

<dl class="method">
<dt id="propythia.machine_learning.MachineLearning.features_importances">
<code class="sig-name descname">features_importances</code><span class="sig-paren">(</span><em class="sig-param">classifier</em>, <em class="sig-param">model_name</em>, <em class="sig-param">top_features=20</em>, <em class="sig-param">color='b'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/machine_learning.html#MachineLearning.features_importances"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.machine_learning.MachineLearning.features_importances" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that given a classifier retrieves the features importances as a dataset and represent as barplot.
:param classifier: classifier
:param model_name: model used in classifier. Choose between ‘svm’, ‘sgd’, ‘lr’, ‘gboosting’ or ‘rf’
:param x_train: descriptor values for training data.
:param top_features: number of features to display on plot
:param color: color of plot bard
:return: bar plot of features and table with features names and importance for the model</p>
</dd></dl>

<dl class="method">
<dt id="propythia.machine_learning.MachineLearning.plot_learning_curve">
<code class="sig-name descname">plot_learning_curve</code><span class="sig-paren">(</span><em class="sig-param">estimator</em>, <em class="sig-param">title='Learning curve {}'</em>, <em class="sig-param">ylim=None</em>, <em class="sig-param">cv=ShuffleSplit(n_splits=100</em>, <em class="sig-param">random_state=42</em>, <em class="sig-param">test_size=0.3</em>, <em class="sig-param">train_size=None)</em>, <em class="sig-param">n_jobs=1</em>, <em class="sig-param">train_sizes=array([0.1</em>, <em class="sig-param">0.325</em>, <em class="sig-param">0.55</em>, <em class="sig-param">0.775</em>, <em class="sig-param">1.   ])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/machine_learning.html#MachineLearning.plot_learning_curve"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.machine_learning.MachineLearning.plot_learning_curve" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot a learning curve to determine cross validated training and tests scores for different training set sizes
:param estimator: classifier/ model to use
:param title: title of the plot
:param ylim:
:param cv: cross validation to use
:param n_jobs:  number of parallel jobs to use for calculation. if <code class="docutils literal notranslate"><span class="pre">-1</span></code>, all available cores are used.
:param train_sizes: train sizes to tests
:return: graphic representing learning curves, numbers of trainig examples, scores on training sets, and scores on tests set</p>
</dd></dl>

<dl class="method">
<dt id="propythia.machine_learning.MachineLearning.plot_roc_curve">
<code class="sig-name descname">plot_roc_curve</code><span class="sig-paren">(</span><em class="sig-param">classifier, ylim=[0.0, 1.0], xlim=[0.0, 1.0], title='Receiver operating characteristic'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/machine_learning.html#MachineLearning.plot_roc_curve"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.machine_learning.MachineLearning.plot_roc_curve" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Function to plot a ROC curve</dt><dd><p>On the y axis, true positive rate and false positive rate on the X axis.
The top left corner of the plot is the ‘ideal’ point - a false positive rate of zero, and a true positive rate of one,
meaning a larger area under the curve (AUC) is usually better.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>classifier</strong> – {classifier instance} pre-trained classifier used for predictions.</p></li>
<li><p><strong>x_test</strong> – {array} descriptor values of the tests data.</p></li>
</ul>
</dd>
</dl>
<p>:param y_test:{array} true class values of the tests data.
:param ylim: y-axis limits
:param xlim: x- axis limits
:param title: title of plot
:return: plot ROC curve</p>
</dd></dl>

<dl class="method">
<dt id="propythia.machine_learning.MachineLearning.plot_validation_curve">
<code class="sig-name descname">plot_validation_curve</code><span class="sig-paren">(</span><em class="sig-param">classifier</em>, <em class="sig-param">param_name</em>, <em class="sig-param">param_range</em>, <em class="sig-param">cv=ShuffleSplit(n_splits=100</em>, <em class="sig-param">random_state=42</em>, <em class="sig-param">test_size=0.3</em>, <em class="sig-param">train_size=None)</em>, <em class="sig-param">score=make_scorer(matthews_corrcoef)</em>, <em class="sig-param">title='Validation Curve'</em>, <em class="sig-param">xlab='parameter range'</em>, <em class="sig-param">ylab='MCC'</em>, <em class="sig-param">n_jobs=-1</em>, <em class="sig-param">filename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/machine_learning.html#MachineLearning.plot_validation_curve"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.machine_learning.MachineLearning.plot_validation_curve" title="Permalink to this definition">¶</a></dt>
<dd><p>This function plots a cross-validation curve for the specified classifier on all tested parameters given in the
option ‘param_range’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>classifier</strong> – {classifier instance} classifier or validation curve (e.g. sklearn.svm.SVC).</p></li>
<li><p><strong>x_train</strong> – {array} descriptor values for training data.</p></li>
<li><p><strong>y_train</strong> – {array} class values for training data.</p></li>
<li><p><strong>param_name</strong> – {string} parameter to assess in the validation curve plot. For example,</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>For SVM,</dt><dd><p>“clf__C” (C parameter), “clf__gamma” (gamma parameter).</p>
</dd>
<dt>For Random Forest,</dt><dd><p>“clf__n_estimators” (number of trees),”clf__max_depth” (max num of branches per tree, “clf__min_samples_split” (min number of samples required to
split an internal tree node), “clf__min_samples_leaf” (min number of samples in newly created leaf).</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>param_range</strong> – {list} parameter range for the validation curve.</p></li>
<li><p><strong>cv</strong> – {int} number of folds for cross-validation.</p></li>
<li><p><strong>score</strong> – {metrics instance} scoring function built from make_scorer() or a predefined value in string form
<a class="reference external" href="http://scikit-learn.org/stable/modules/model_evaluation.html#scoring-parameter">sklearn.model_evaluation.scoring-parameter</a>.</p></li>
<li><p><strong>title</strong> – {str} graph title</p></li>
<li><p><strong>xlab</strong> – {str} x axis label.</p></li>
<li><p><strong>ylab</strong> – {str} y axis label.</p></li>
<li><p><strong>n_jobs</strong> – {int} number of parallel jobs to use for calculation. if <code class="docutils literal notranslate"><span class="pre">-1</span></code>, all available cores are used.</p></li>
<li><p><strong>filename</strong> – {str} if filename given the figure is stored in the specified path.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>plot of the validation curve.</p>
</dd>
</dl>
<p>Based on a function from moodlamp
Müller A. T. et al. (2017) modlAMP: Python for anitmicrobial peptides, Bioinformatics 33, (17), 2753-2755,
DOI:10.1093/bioinformatics/btx285</p>
</dd></dl>

<dl class="method">
<dt id="propythia.machine_learning.MachineLearning.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">classifier</em>, <em class="sig-param">x</em>, <em class="sig-param">seqs</em>, <em class="sig-param">names=None</em>, <em class="sig-param">y=None</em>, <em class="sig-param">filename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/machine_learning.html#MachineLearning.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.machine_learning.MachineLearning.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>This function can be used to predict novel peptides with a trained classifier model. The function returns a
‘pandas.DataFrame’ with predictions using the specified estimator and tests data. If true class is provided,
it returns the scoring value for the tests data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>classifier</strong> – {classifier instance} classifier used for predictions.</p></li>
<li><p><strong>x</strong> – {array} descriptor values of the peptides to be predicted.</p></li>
<li><p><strong>seqs</strong> – {list} sequences of the peptides in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>names</strong> – {list} (optional) names of the peptides in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>y</strong> – {array} (optional) true (known) classes of the peptides.</p></li>
<li><p><strong>filename</strong> – {string} (optional) output filename to store the predictions to (<code class="docutils literal notranslate"><span class="pre">.csv</span></code> format); if <code class="docutils literal notranslate"><span class="pre">None</span></code>:
not saved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> containing predictions for <code class="docutils literal notranslate"><span class="pre">x</span></code>. <code class="docutils literal notranslate"><span class="pre">P_class0</span></code> and <code class="docutils literal notranslate"><span class="pre">P_class1</span></code>
are the predicted probability of the peptide belonging to class 0 and class 1, respectively.</p>
</dd>
</dl>
<p>Based on a function from moodlamp
Müller A. T. et al. (2017) modlAMP: Python for anitmicrobial peptides, Bioinformatics 33, (17), 2753-2755,
DOI:10.1093/bioinformatics/btx285</p>
</dd></dl>

<dl class="method">
<dt id="propythia.machine_learning.MachineLearning.predict_window">
<code class="sig-name descname">predict_window</code><span class="sig-paren">(</span><em class="sig-param">classifier</em>, <em class="sig-param">seq</em>, <em class="sig-param">x=None</em>, <em class="sig-param">window_size=20</em>, <em class="sig-param">gap=1</em>, <em class="sig-param">features=[]</em>, <em class="sig-param">features_names=None</em>, <em class="sig-param">features_dataframe=None</em>, <em class="sig-param">names=None</em>, <em class="sig-param">y=None</em>, <em class="sig-param">filename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/machine_learning.html#MachineLearning.predict_window"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.machine_learning.MachineLearning.predict_window" title="Permalink to this definition">¶</a></dt>
<dd><p>Scan a protein in a sliding window approach to predict novel peptides with a trained classifier model.
The function returns a
‘pandas.DataFrame’ with predictions using the specified estimator and tests data. If true class is provided,
it returns the scoring value for the tests data.
The user can provide a features_dataframe if does not want to use the descriptors supported. Otherwise, the user
can also provide a list with the numbers of the descriptors that want to be calculated (descriptors module adaptable())
if none is provided, the function will calculate all the descriptors available.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>classifier</strong> – {classifier instance} classifier used for predictions.</p></li>
<li><p><strong>x</strong> – {array} descriptor values of the peptides to be predicted.</p></li>
<li><p><strong>seq</strong> – sequence of the peptides in <code class="docutils literal notranslate"><span class="pre">x</span></code>.</p></li>
<li><p><strong>window_size</strong> – number of aminoacids to considerer in each seq to check. for default 20</p></li>
<li><p><strong>gap</strong> – gap size of the search of windows in sequence. default 1</p></li>
<li><p><strong>features</strong> – list containing the list features to be calculated under de adaptable function.
if list is empty will calculate all the descriptors available in the package</p></li>
<li><p><strong>features_dataframe</strong> – dataframe with sequences and its features</p></li>
</ul>
</dd>
</dl>
<p>:param features_names:names of features. If none will match the features with the ones with the model given
:param names: {list} (optional) names of the peptides in <code class="docutils literal notranslate"><span class="pre">x</span></code>.
:param y: {array} (optional) true (known) classes of the peptides.
:param filename: {string} (optional) output filename to store the predictions to (<code class="docutils literal notranslate"><span class="pre">.csv</span></code> format); if <code class="docutils literal notranslate"><span class="pre">None</span></code>:</p>
<blockquote>
<div><p>not saved.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> containing predictions for subsequences generated, <code class="docutils literal notranslate"><span class="pre">P_class0</span></code> and <code class="docutils literal notranslate"><span class="pre">P_class1</span></code>
are the predicted probability of the peptide belonging to class 0 and class 1, respectively. The previsions
are divided in probability classes &lt;0.99, &gt;0.95, &gt;0.9, &lt;0.8, &gt;0.7, &gt;0.6 and 0</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.machine_learning.MachineLearning.score_testset">
<code class="sig-name descname">score_testset</code><span class="sig-paren">(</span><em class="sig-param">classifier</em>, <em class="sig-param">sample_weights=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/machine_learning.html#MachineLearning.score_testset"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.machine_learning.MachineLearning.score_testset" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the tests set scores for the specified scoring metrics in a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>. The calculated metrics
are Matthews correlation coefficient, accuracy, precision, recall, f1 and area under the Receiver-Operator Curve
(roc_auc). See <a class="reference external" href="http://scikit-learn.org/stable/modules/classes.html#sklearn-metrics-metrics">sklearn.metrics</a>
for more information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>classifier</strong> – {classifier instance} pre-trained classifier used for predictions.</p></li>
<li><p><strong>x_test</strong> – {array} descriptor values of the tests data.</p></li>
<li><p><strong>y_test</strong> – {array} true class values of the tests data.</p></li>
<li><p><strong>sample_weights</strong> – {array} weights for the tests data.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> containing the cross validation scores for the specified metrics.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.machine_learning.MachineLearning.train_best_model">
<code class="sig-name descname">train_best_model</code><span class="sig-paren">(</span><em class="sig-param">model</em>, <em class="sig-param">sample_weights=None</em>, <em class="sig-param">scaler=StandardScaler(copy=True</em>, <em class="sig-param">with_mean=True</em>, <em class="sig-param">with_std=True)</em>, <em class="sig-param">score=make_scorer(matthews_corrcoef)</em>, <em class="sig-param">param_grid=None</em>, <em class="sig-param">n_jobs=-1</em>, <em class="sig-param">cv=10</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/machine_learning.html#MachineLearning.train_best_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.machine_learning.MachineLearning.train_best_model" title="Permalink to this definition">¶</a></dt>
<dd><p>This function performs a parameter grid search on a selected classifier model and peptide training data set.
It returns a scikit-learn pipeline that performs standard scaling and contains the best model found by the
grid search according to the Matthews correlation coefficient.
:param model: {str} model to train. Choose between ‘svm’, ‘knn’, ‘sgd’, ‘lr’,’rf’, ‘gnb’, ‘nn’,’gboosting’
:param x_train: {array} descriptor values for training data.
:param y_train: {array} class values for training data.
:param sample_weights: {array} sample weights for training data.
:param scaler: {scaler} scaler to use in the pipe to scale data prior to training. Choose from</p>
<blockquote>
<div><p><code class="docutils literal notranslate"><span class="pre">sklearn.preprocessing</span></code>, e.g. ‘StandardScaler()’, ‘MinMaxScaler()’, ‘Normalizer()’.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>score</strong> – {metrics instance} scoring function built from make_scorer() or a predefined value in string form
(choose from the scikit-learn`scoring-parameters &lt;<a class="reference external" href="http://scikit-learn.org/stable/modules/model_evaluation.html#scoring-parameter">http://scikit-learn.org/stable/modules/model_evaluation.html#scoring-parameter</a>&gt;`_).</p></li>
<li><p><strong>param_grid</strong> – {dict} parameter grid for the gridsearch (see`sklearn.grid_search &lt;<a class="reference external" href="http://scikit-learn.org/stable/modules/model_evaluation.html">http://scikit-learn.org/stable/modules/model_evaluation.html</a>&gt;`_).</p></li>
<li><p><strong>n_jobs</strong> – {int} number of parallel jobs to use for calculation. if <code class="docutils literal notranslate"><span class="pre">-1</span></code>, all available cores are used.</p></li>
<li><p><strong>cv</strong> – {int} number of folds for cross-validation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>best estimator pipeline.</p>
</dd>
</dl>
<p><strong>Default parameter grids:</strong></p>
<p>Based on a function from moodlamp
Müller A. T. et al. (2017) modlAMP: Python for anitmicrobial peptides, Bioinformatics 33, (17), 2753-2755,
DOI:10.1093/bioinformatics/btx285</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-propythia.preprocess">
<span id="preprocess"></span><h2>preprocess<a class="headerlink" href="#module-propythia.preprocess" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>A class used for utility functions and transformer classes to change raw feature
vectors into a representation that is more suitable for the downstream estimators
All the functions are imported from sklearn.preprocessing</p>
<p>Authors: Ana Marta Sequeira</p>
<p>Date: 05/2019</p>
<p>Email:</p>
<hr class="docutils" />
<dl class="class">
<dt id="propythia.preprocess.Preprocess">
<em class="property">class </em><code class="sig-prename descclassname">propythia.preprocess.</code><code class="sig-name descname">Preprocess</code><a class="reference internal" href="_modules/propythia/preprocess.html#Preprocess"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.preprocess.Preprocess" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The Preprocess class aims to transform the feature vectors into a representation suitable for downstream estimators.
Clean the dataset from features redundants and deal with Nans.
All the functions are imported from the module sklearn.preprocessing</p>
<dl class="method">
<dt id="propythia.preprocess.Preprocess.missing_data">
<code class="sig-name descname">missing_data</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/preprocess.html#Preprocess.missing_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.preprocess.Preprocess.missing_data" title="Permalink to this definition">¶</a></dt>
<dd><p>check nans. if yes: returns a error message
:param data: dataframe
:return: error message or 0 nans error</p>
</dd></dl>

<dl class="method">
<dt id="propythia.preprocess.Preprocess.preprocess">
<code class="sig-name descname">preprocess</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">columns_names=True</em>, <em class="sig-param">threshold=0</em>, <em class="sig-param">standard=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/preprocess.html#Preprocess.preprocess"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.preprocess.Preprocess.preprocess" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes columns that have all values as zero, duplicated and low variance columns</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> – dataset of input</p></li>
<li><p><strong>columns_names</strong> – if True retrieves the names of columns deleted</p></li>
<li><p><strong>threshold</strong> – the threshold of variance to drop columns</p></li>
<li><p><strong>standard</strong> – if true minmaxscaler will be applied</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dataset original without deleted columns</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.preprocess.Preprocess.remove_columns_all_zeros">
<code class="sig-name descname">remove_columns_all_zeros</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">columns_names=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/preprocess.html#Preprocess.remove_columns_all_zeros"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.preprocess.Preprocess.remove_columns_all_zeros" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes columns that have all values as zero.
:param data: dataframe
:param columns_names: if True retrieves the names of columns with only zeros
:return: dataset without columns with all values=zero</p>
</dd></dl>

<dl class="method">
<dt id="propythia.preprocess.Preprocess.remove_duplicate_columns">
<code class="sig-name descname">remove_duplicate_columns</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">columns_names=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/preprocess.html#Preprocess.remove_duplicate_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.preprocess.Preprocess.remove_duplicate_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes columns duplicated.
:param data: dataframe
:param columns_names: if True retrieves the names of columns duplicates
:return: dataset without duplicated columns</p>
</dd></dl>

<dl class="method">
<dt id="propythia.preprocess.Preprocess.remove_low_variance">
<code class="sig-name descname">remove_low_variance</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">threshold=0</em>, <em class="sig-param">standard=True</em>, <em class="sig-param">columns_names=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/preprocess.html#Preprocess.remove_low_variance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.preprocess.Preprocess.remove_low_variance" title="Permalink to this definition">¶</a></dt>
<dd><p>Based on scikit learn
VarianceThreshold is a simple baseline approach to feature selection.
It removes all features whose variance doesn’t meet some threshold
:param data: dataframe
:param threshold: value. The threshold of variance to drop columns (eg 0.8)
:param standard: if , in the case of threshold &gt;0, the user wants to standardize features before apply variance
threshold. minmaxscaler will be applied
:param columns_names:
:return: dataset without low variance columns (not scaled)</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-propythia.sequence">
<span id="sequence"></span><h2>sequence<a class="headerlink" href="#module-propythia.sequence" title="Permalink to this headline">¶</a></h2>
<hr class="docutils" />
<p>A class used for reading sequences or change sequences.
The main objective is to create sequence objects to calculate descriptors
The class allows to:</p>
<blockquote>
<div><p>1)Read sequences from string or from uniprot ID (is also possible retrieve sequences fom txt with uniprot IDs)
2)Check if the protein sequence is a valid sequence
3)Obtain a sized sequence of list of sequences, adding or cutting from both n and c terminals
4)From one sequence generate list of subsequences based on sliding window approach, from specific aa, from the
the terminals or divide the sequence in parts</p>
</div></blockquote>
<p>Authors:Ana Marta Sequeira</p>
<p>Date: 01/2019</p>
<p>Email:</p>
<hr class="docutils" />
<dl class="class">
<dt id="propythia.sequence.ReadSequence">
<em class="property">class </em><code class="sig-prename descclassname">propythia.sequence.</code><code class="sig-name descname">ReadSequence</code><a class="reference internal" href="_modules/propythia/sequence.html#ReadSequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.sequence.ReadSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The ReadSequence class aims to read the input and transform it into a sequence that can be used to calculate Descriptors.
It accepts uniprotID, txt with uniprot ID and string with aminoacid sequence. The functions to read protein are based on the package pydpi.
It allows to get sized sequences or a variety of subsequences from just one sequence.</p>
<dl class="method">
<dt id="propythia.sequence.ReadSequence.checkprotein">
<code class="sig-name descname">checkprotein</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/sequence.html#ReadSequence.checkprotein"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.sequence.ReadSequence.checkprotein" title="Permalink to this definition">¶</a></dt>
<dd><p>Check whether the protein sequence is a valid amino acid sequence or not.
Just check. Not assign to the object.
:param protein_sequence: protein sequence
:return: error or valid message.</p>
</dd></dl>

<dl class="method">
<dt id="propythia.sequence.ReadSequence.get_protein_sequence_from_id">
<code class="sig-name descname">get_protein_sequence_from_id</code><span class="sig-paren">(</span><em class="sig-param">uniprotid=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/sequence.html#ReadSequence.get_protein_sequence_from_id"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.sequence.ReadSequence.get_protein_sequence_from_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloading a protein sequence by uniprot id.
:param uniprotid: String with UniprotID
:return: String with aa sequence</p>
</dd></dl>

<dl class="method">
<dt id="propythia.sequence.ReadSequence.get_protein_sequence_from_txt">
<code class="sig-name descname">get_protein_sequence_from_txt</code><span class="sig-paren">(</span><em class="sig-param">path</em>, <em class="sig-param">openfile</em>, <em class="sig-param">savefile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/sequence.html#ReadSequence.get_protein_sequence_from_txt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.sequence.ReadSequence.get_protein_sequence_from_txt" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to retrieve sequences from a txt with uniprot ID. Does not retrieve a sequence object
:param path: directory path containing ID file
:param openfile:ID file (‘name.txt’)
:param savefile: saved file with obtained protein sequences (‘name2.txt’)
:return:File containing string of sequences</p>
</dd></dl>

<dl class="method">
<dt id="propythia.sequence.ReadSequence.get_sized_seq">
<code class="sig-name descname">get_sized_seq</code><span class="sig-paren">(</span><em class="sig-param">sequences=[]</em>, <em class="sig-param">n_terminal=10</em>, <em class="sig-param">c_terminal=10</em>, <em class="sig-param">terminal=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/sequence.html#ReadSequence.get_sized_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.sequence.ReadSequence.get_sized_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>cut or add aminoacids to obtain sequences with equal lenght.
:param sequences: list containing protein sequences (string) or just a protein sequence
:param n_terminal: number of aa to consider in the n terminal (left side of sequence)
:param c_terminal: number of aa to consider in the c terminal (right side of sequence)
:param terminal: in case of the need to add dummy aa and no terminal as already been chosen, it decides where to add</p>
<blockquote>
<div><p>0 to add to the right (consider N terminal)
-1 to add to the left (consider C terminal)
2 to add in the middle (N and C terminal will be both present and repeated with dummy in middle</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of sequences containing all the same lenght. if just one sequence given it will return a string</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.sequence.ReadSequence.get_sub_seq_sliding_window">
<code class="sig-name descname">get_sub_seq_sliding_window</code><span class="sig-paren">(</span><em class="sig-param">seq</em>, <em class="sig-param">window_size=20</em>, <em class="sig-param">gap=1</em>, <em class="sig-param">index=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/sequence.html#ReadSequence.get_sub_seq_sliding_window"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.sequence.ReadSequence.get_sub_seq_sliding_window" title="Permalink to this definition">¶</a></dt>
<dd><p>sliding window of the protein given. It will generate a list of n sequences with
lenght equal to the value of window and spaced a gap value. It can or not retrieve
the indices of location of the subsequence in the original sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – protein sequence</p></li>
<li><p><strong>window_size</strong> – number of aminoacids to considerer, lenght of the subsequence. for default 20</p></li>
<li><p><strong>gap</strong> – gap size of the search of windows in sequence. default 1</p></li>
<li><p><strong>index</strong> – if true, return the indices of location of the subsequence in the original sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list with subsequences generated with or without a list of tuples with location of subsequences
in original sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.sequence.ReadSequence.get_sub_seq_split">
<code class="sig-name descname">get_sub_seq_split</code><span class="sig-paren">(</span><em class="sig-param">seq</em>, <em class="sig-param">number_of_subseq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/sequence.html#ReadSequence.get_sub_seq_split"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.sequence.ReadSequence.get_sub_seq_split" title="Permalink to this definition">¶</a></dt>
<dd><p>Split the originl seq in n number of subsequences.
:param seq: protein sequence
:param number_of_subseq: number of subsequences to divide the original seq
:return: list with number_of_sequences sequences.</p>
</dd></dl>

<dl class="method">
<dt id="propythia.sequence.ReadSequence.get_sub_seq_terminals">
<code class="sig-name descname">get_sub_seq_terminals</code><span class="sig-paren">(</span><em class="sig-param">seq</em>, <em class="sig-param">N_terminal=5</em>, <em class="sig-param">C_terminal=5</em>, <em class="sig-param">rest=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/sequence.html#ReadSequence.get_sub_seq_terminals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.sequence.ReadSequence.get_sub_seq_terminals" title="Permalink to this definition">¶</a></dt>
<dd><p>Divide the sequence in the N terminal and  C terminal with sizes defined by the user. It returns
a list with N and C terminal and rest of the sequence.
By default the N terminal is considerer to be the beggining of the sequence (left)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – protein sequence</p></li>
<li><p><strong>N_terminal</strong> – size of the N terminal to consider. If zero will not return</p></li>
<li><p><strong>C_terminal</strong> – size of the C terminal to consider. If zero will not return</p></li>
<li><p><strong>rest</strong> – If true will return the restant subsequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list with N, C and rest of the sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.sequence.ReadSequence.get_sub_seq_to_aa">
<code class="sig-name descname">get_sub_seq_to_aa</code><span class="sig-paren">(</span><em class="sig-param">seq</em>, <em class="sig-param">ToAA='S'</em>, <em class="sig-param">window=5</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/sequence.html#ReadSequence.get_sub_seq_to_aa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.sequence.ReadSequence.get_sub_seq_to_aa" title="Permalink to this definition">¶</a></dt>
<dd><p>Get all 2*window+1 sub-sequences whose center is ToAA in a protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – protein sequence</p></li>
<li><p><strong>ToAA</strong> – central (query point) amino acid in the sub-sequence</p></li>
<li><p><strong>window</strong> – span (number of amnoacids to go front and back from the ToAA</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list form containing sub-sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="propythia.sequence.ReadSequence.read_protein_sequence">
<code class="sig-name descname">read_protein_sequence</code><span class="sig-paren">(</span><em class="sig-param">protein_sequence=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/propythia/sequence.html#ReadSequence.read_protein_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#propythia.sequence.ReadSequence.read_protein_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Read a protein sequence.
:param protein_sequence: String with sequence
:return: String with aa sequence</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-propythia">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-propythia" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="propythia.adjuv_functions.html" class="btn btn-neutral float-right" title="propythia.adjuv_functions package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to ProPythia’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Ana Marta Sequeira

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>